<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Other multivariate analyses: cluster analysis</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; } /* Alert */
code span.an { color: #008000; } /* Annotation */
code span.at { } /* Attribute */
code span.bu { } /* BuiltIn */
code span.cf { color: #0000ff; } /* ControlFlow */
code span.ch { color: #008080; } /* Char */
code span.cn { } /* Constant */
code span.co { color: #008000; } /* Comment */
code span.cv { color: #008000; } /* CommentVar */
code span.do { color: #008000; } /* Documentation */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.im { } /* Import */
code span.in { color: #008000; } /* Information */
code span.kw { color: #0000ff; } /* Keyword */
code span.op { } /* Operator */
code span.ot { color: #ff4000; } /* Other */
code span.pp { color: #ff4000; } /* Preprocessor */
code span.sc { color: #008080; } /* SpecialChar */
code span.ss { color: #008080; } /* SpecialString */
code span.st { color: #008080; } /* String */
code span.va { } /* Variable */
code span.vs { color: #008080; } /* VerbatimString */
code span.wa { color: #008000; font-weight: bold; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link rel="stylesheet" href="html-md-01.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">R for Earth-System Science</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="overview.html">Overview</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Topics
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Lecture topics</li>
    <li>
      <a href="intro.html">Introduction: R, RStudio, and other infrastructure</a>
    </li>
    <li>
      <a href="usingR.html">Using R</a>
    </li>
    <li>
      <a href="ESSdata.html">Earth-system science data</a>
    </li>
    <li>
      <a href="netCDF.html">netCDF in R</a>
    </li>
    <li>
      <a href="raster_intro.html">raster and netCDF in R</a>
    </li>
    <li>
      <a href="hdf5_intro.html">HDF in R</a>
    </li>
    <li>
      <a href="geospatial.html">Geospatial analyses</a>
    </li>
    <li>
      <a href="RMaps.html">Maps in R</a>
    </li>
    <li>
      <a href="RMaps2.html">Maps in R (2)</a>
    </li>
    <li>
      <a href="rasterVis01.html">Visualization:  raster and rasterVis</a>
    </li>
    <li>
      <a href="vis_ggplot01.html">Visualization for explanation</a>
    </li>
    <li>
      <a href="highdim.html">Visualization of high-dimensional data</a>
    </li>
    <li>
      <a href="PCA.html">Principal components analysis</a>
    </li>
    <li>
      <a href="multivariate.html">Other multivariate analyses</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tasks
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="geog495_ex01.html">Install R and RStudio</a>
    </li>
    <li>
      <a href="packages-and-data.html">Install packages and data</a>
    </li>
    <li>
      <a href="GitHub.html">Git and GitHub</a>
    </li>
    <li>
      <a href="Ex_UsingR.html">Using R exercise</a>
    </li>
    <li>
      <a href="markdown.html">Markdown and RMarkdown</a>
    </li>
    <li>
      <a href="install_netCDF.html">Install netCDF</a>
    </li>
    <li>
      <a href="install_Panoply.html">Install Panoply</a>
    </li>
    <li>
      <a href="transfer.html">File transfer</a>
    </li>
    <li>
      <a href="project.html">GitHub project repository and web page</a>
    </li>
    <li>
      <a href="ltms-and-anomalies.html">Long-term means and anomalies</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Resources
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="readings.html">Readings</a>
    </li>
    <li>
      <a href="websites.html">Old course web pages</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="about.html">About</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Other multivariate analyses: cluster
analysis</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#introduction" id="toc-introduction"><span
class="toc-section-number">1</span> Introduction</a></li>
<li><a href="#a-simple-cluster-analysis"
id="toc-a-simple-cluster-analysis"><span
class="toc-section-number">2</span> A simple cluster analysis</a>
<ul>
<li><a href="#the-cluster-analysis" id="toc-the-cluster-analysis"><span
class="toc-section-number">2.1</span> The cluster analysis</a></li>
<li><a href="#examining-the-cluster-analysis"
id="toc-examining-the-cluster-analysis"><span
class="toc-section-number">2.2</span> Examining the cluster
analysis</a></li>
</ul></li>
<li><a href="#distances-among-observations-and-among-variables"
id="toc-distances-among-observations-and-among-variables"><span
class="toc-section-number">3</span> Distances: among observations and
among variables</a></li>
<li><a href="#heatmaps-2-way-time-and-space-clustering"
id="toc-heatmaps-2-way-time-and-space-clustering"><span
class="toc-section-number">4</span> Heatmaps (2-way, time and space
clustering)</a>
<ul>
<li><a href="#a-simple-example-with-hovmöller-matrix-data"
id="toc-a-simple-example-with-hovmöller-matrix-data"><span
class="toc-section-number">4.1</span> A simple example with
Hovmöller-matrix data</a>
<ul>
<li><a href="#set-up" id="toc-set-up"><span
class="toc-section-number">4.1.1</span> Set up</a></li>
<li><a href="#hovmöller-matrix-plot-decimated-decadal-data"
id="toc-hovmöller-matrix-plot-decimated-decadal-data"><span
class="toc-section-number">4.1.2</span> Hovmöller-matrix plot (decimated
decadal data)</a></li>
<li><a href="#hovmöller-matrix-heatmap-decimated-decadal-data"
id="toc-hovmöller-matrix-heatmap-decimated-decadal-data"><span
class="toc-section-number">4.1.3</span> Hovmöller-matrix heatmap
(decimated decadal data)</a></li>
</ul></li>
<li><a href="#a-second-example-with-decadal-data"
id="toc-a-second-example-with-decadal-data"><span
class="toc-section-number">4.2</span> A second example with decadal
data</a>
<ul>
<li><a href="#hovmöller-matrix-plot"
id="toc-hovmöller-matrix-plot"><span
class="toc-section-number">4.2.1</span> Hovmöller-matrix plot</a></li>
<li><a href="#hovmöller-matrix-heatmap"
id="toc-hovmöller-matrix-heatmap"><span
class="toc-section-number">4.2.2</span> Hovmöller-matrix
heatmap</a></li>
</ul></li>
<li><a href="#full-grid-example" id="toc-full-grid-example"><span
class="toc-section-number">4.3</span> Full-grid example</a>
<ul>
<li><a href="#set-up-1" id="toc-set-up-1"><span
class="toc-section-number">4.3.1</span> Set up</a></li>
<li><a href="#matrix-plot" id="toc-matrix-plot"><span
class="toc-section-number">4.3.2</span> Matrix plot</a></li>
<li><a href="#heatmap" id="toc-heatmap"><span
class="toc-section-number">4.3.3</span> Heatmap</a></li>
</ul></li>
</ul></li>
</ul>
</div>

<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('right', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<div id="introduction" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>There is a broad group of multivariate analyses that have as their
objective the organization of individual observations (objects, sites,
grid points, individuals), and these analyses are built upon the concept
of multivariate distances (expressed either as similarities or
dissimilarities) among the objects.</p>
<p>The organization generally takes two forms:</p>
<ul>
<li>the arrangement of the objects in a lower-dimensional space than the
data were originally observed on;</li>
<li>the development of “natural” clusters or the classifications of the
objects.</li>
</ul>
<p>These analyses share many concepts and techniques (both numerical and
practical) with other procedures such as principal components analysis,
numerical taxonomy, discriminant analysis and so on.</p>
<p>The analyses generally begin with the construction of an <em>n</em> x
<em>n</em> matrix <strong>D</strong> of the distances between objects.
For example, in a two-dimensional space, the elements
<em>d<sub>ij</sub></em> of <strong>D</strong> could be the Euclidean
distances between points,</p>
<p><em>d<sub>ij</sub></em> = [(<em>x<sub>i1</sub></em>-
<em>x<sub>j1</sub></em>)<sup>2</sup> + (<em>x<sub>i1</sub></em>-
<em>x<sub>j1</sub></em>)<sup>2</sup>]<sup>½</sup></p>
<p>The Euclidean distance, and related measures are easily generalized
to more than two dimensions.</p>
</div>
<div id="a-simple-cluster-analysis" class="section level1" number="2">
<h1><span class="header-section-number">2</span> A simple cluster
analysis</h1>
<p>The basic idea of cluster analysis can be illustrated using the
Oregon climate-station dataset, with the goal being to identify groups
of stations with similar climates, which also regionalizes the climate
of the state. Begin by reading a shape file for mapping the data, and
the data itself.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># read county outlines</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(sp)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(rgdal)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>otl_path <span class="ot">&lt;-</span> <span class="st">&quot;/Users/bartlein/Projects/ESSD/data/shp_files/OR/&quot;</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>otl_name <span class="ot">&lt;-</span> <span class="st">&quot;orotl.shp&quot;</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>otl_file <span class="ot">&lt;-</span> <span class="fu">paste</span>(otl_path, otl_name, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>orotl_shp <span class="ot">&lt;-</span> <span class="fu">readOGR</span>(otl_file)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># read stations</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>orstationc_path <span class="ot">&lt;-</span> <span class="st">&quot;/Users/bartlein/Projects/ESSD/data/csv_files/&quot;</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>orstationc_name <span class="ot">&lt;-</span> <span class="st">&quot;orstationc.csv&quot;</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>orstationc_file <span class="ot">&lt;-</span> <span class="fu">paste</span>(orstationc_path, orstationc_name, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>orstationc <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(orstationc_file)</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="fu">summary</span>(orstationc)</span></code></pre></div>
<pre><code>##    station               lat             lon              elev              tjan             tjul      
##  Length:92          Min.   :42.05   Min.   :-124.6   Min.   :   2.00   Min.   :-6.000   Min.   :12.20  
##  Class :character   1st Qu.:43.60   1st Qu.:-123.2   1st Qu.:  94.25   1st Qu.:-1.225   1st Qu.:17.30  
##  Mode  :character   Median :44.46   Median :-121.7   Median : 462.00   Median : 0.750   Median :19.20  
##                     Mean   :44.32   Mean   :-121.3   Mean   : 556.99   Mean   : 1.417   Mean   :19.01  
##                     3rd Qu.:45.28   3rd Qu.:-119.6   3rd Qu.: 863.75   3rd Qu.: 4.150   3rd Qu.:20.20  
##                     Max.   :46.15   Max.   :-117.0   Max.   :1974.00   Max.   : 8.400   Max.   :26.40  
##       tann             pjan             pjul            pann            idnum            Name          
##  Min.   : 3.200   Min.   : 14.00   Min.   : 4.00   Min.   : 198.0   Min.   :350197   Length:92         
##  1st Qu.: 8.700   1st Qu.: 39.75   1st Qu.: 7.00   1st Qu.: 299.0   1st Qu.:352308   Class :character  
##  Median :10.400   Median : 87.50   Median : 9.00   Median : 531.0   Median :354564   Mode  :character  
##  Mean   : 9.885   Mean   :141.67   Mean   :10.84   Mean   : 868.6   Mean   :354455                     
##  3rd Qu.:11.200   3rd Qu.:239.25   3rd Qu.:12.25   3rd Qu.:1355.8   3rd Qu.:356663                     
##  Max.   :12.600   Max.   :414.00   Max.   :34.00   Max.   :2517.0   Max.   :359316</code></pre>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">attach</span>(orstationc)</span></code></pre></div>
<pre><code>## The following objects are masked _by_ .GlobalEnv:
## 
##     lat, lon</code></pre>
<p>Plot the station locations.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># plot stations</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">plot</span>(orotl_shp)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="fu">points</span>(lon, lat, <span class="at">xlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">125</span>, <span class="sc">-</span><span class="dv">116</span>), <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">41</span>,<span class="dv">47</span>),</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>   <span class="at">xlab=</span><span class="st">&quot;Latitude&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;Longitude&quot;</span>, <span class="at">type=</span><span class="st">&quot;n&quot;</span>)</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="fu">text</span>(lon, lat, <span class="at">labels=</span>station, <span class="at">cex=</span><span class="fl">0.6</span>)</span></code></pre></div>
<div id="the-cluster-analysis" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> The cluster
analysis</h2>
<p>In this example, the groups of stations (or the climate regions of
Oregon, as expressed in the climate-station data) will be defined via an
“agglomerative-hierarchical” cluster analysis. In such an analysis,
individual objects (observations) are combined according to their
(dis)similarity, with the two <em>most- similar</em> objects being
combined first (to create a new composite object), then the next two
most similar objects (including the composites), and so on until a
single object is defined. The sequence of combinations is illustrated
using a “dendrogram” (cluster tree), and this can be examined to
determine the number of clusters. The function cutree determines the
cluster membership of each observation, which can be listed or
plotted.</p>
<p>The first step in an agglomerative cluster analysis (although not
always expicitly displayed), is the calculation of a dissimilarity
matrix, which is a square symmetrical matrix that illustrates the
dissimilarities (typically Euclidean distances) between individual cases
or observations, using, in this case the six climate variables to define
the space that distance is calculated in.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># get dissimilarities (distances) and cluster</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="fu">library</span>(lattice)</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(orstationc[,<span class="dv">5</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>X_dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(<span class="fu">scale</span>(X))</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>grid <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">obs1 =</span> <span class="fu">seq</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">92</span>), <span class="at">obs2 =</span> <span class="fu">seq</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">92</span>))</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="fu">levelplot</span>(<span class="fu">as.matrix</span>(X_dist)<span class="sc">/</span><span class="fu">max</span>(X_dist) <span class="sc">~</span> obs1<span class="sc">*</span>obs2, <span class="at">data=</span>grid, </span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>    <span class="at">col.regions =</span> <span class="fu">gray</span>(<span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">by=</span><span class="fl">0.0625</span>)), <span class="at">xlab=</span><span class="st">&quot;observation&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;observation&quot;</span>)</span></code></pre></div>
<p>In the code above, for convience, a matrix <strong>X</strong> was
created to omit the variables we don’t want to include in the clustering
(i.e. the non-climatic information). The levelplot() function plots the
dissimilarity matrix that the cluster analysis works with. The “plaid”
as opposed to completely random appearance of the plot indicates that it
is likely that natural groups will exist in the data. The dark diagonal
line is formed from the Eucldean distance values of 0.0 from the
comparison of each observation with itself.</p>
<p>Now do the cluster analysis:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># cluster analysis</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>X_clusters <span class="ot">&lt;-</span> <span class="fu">hclust</span>(X_dist, <span class="at">method =</span> <span class="st">&quot;ward.D2&quot;</span>)</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="fu">plot</span>(X_clusters, <span class="at">labels=</span>station, <span class="at">cex=</span><span class="fl">0.5</span>)</span></code></pre></div>
<p>The function <code>hclust()</code> performs a “Wards’s
single-linkage” analysis, and the <code>plot()</code> function plots the
dendrogram. Notice that the individual “leaves” of the plot are labeled
by the climate-station names.</p>
<p>The dendrogram can be inspected to get an idea of how many distinct
clusters there are (and it can also identify unusual points–look for
Crater Lake <code>CLK</code>). As a first guess, let’s say there were
three distinct clusters evident in the dendrogram. The
<code>cuttree()</code> function can be used to “cut” the dendrogram at a
particular level:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># cut dendrogram and plot points in each cluster </span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>nclust <span class="ot">&lt;-</span> <span class="dv">3</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>clusternum <span class="ot">&lt;-</span> <span class="fu">cutree</span>(X_clusters, <span class="at">k=</span>nclust)</span></code></pre></div>
<p>And then the cluster membership (i.e. which cluster?) of each station
can be mapped.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># plot cluster membership of each station</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">plot</span>(lon, lat, <span class="at">xlim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">125</span>, <span class="sc">-</span><span class="dv">116</span>), <span class="at">ylim=</span><span class="fu">c</span>(<span class="dv">41</span>,<span class="dv">47</span>), <span class="at">xlab=</span><span class="st">&quot;Latitude&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;Longitude&quot;</span>,</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>     <span class="at">main=</span><span class="fu">paste</span>(<span class="fu">as.character</span>(nclust), <span class="st">&quot;Clusters&quot;</span>), <span class="at">type=</span><span class="st">&quot;n&quot;</span>)</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="fu">plot</span>(orotl_shp, <span class="at">add=</span>T)</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="fu">text</span>(lon, lat, <span class="at">labels=</span>clusternum, <span class="at">col=</span><span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p>(A listing of the cluster number that each station belongs to can be
gotten by the following code.)</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">cbind</span>(station, clusternum, <span class="fu">as.character</span>(Name)))</span></code></pre></div>
</div>
<div id="examining-the-cluster-analysis" class="section level2"
number="2.2">
<h2><span class="header-section-number">2.2</span> Examining the cluster
analysis</h2>
<p>The efficacy of the analyis in regionalizing or grouping the stations
can be gauged usigns plots and a analysis that looks at the “separation”
of the groups of stations.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co"># examine clusters -- simple plots </span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="fu">tapply</span>(tjan, clusternum, mean)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="fu">histogram</span>(<span class="sc">~</span> tjan <span class="sc">|</span> <span class="fu">factor</span>(clusternum), <span class="at">nint=</span><span class="dv">20</span>, <span class="at">layout=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span></code></pre></div>
<p>Note how many of the histograms seem multimodal–this is a clear sign
of inhomogeneity of the clusters. Let’s see how the clusters differ.
This can be done via discriminant analysis, which (like PCA) defines
some new variables with desireable properties, in this case, new
variables that maximally “spread out” the groups along the new variable
axes:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co"># discriminant analysis (chunk 3)</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="fu">library</span>(MASS)</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>cluster_lda1 <span class="ot">&lt;-</span> <span class="fu">lda</span>(clusternum <span class="sc">~</span> tjan <span class="sc">+</span> tjul <span class="sc">+</span> tann <span class="sc">+</span> pjan <span class="sc">+</span> pjul <span class="sc">+</span> pann)</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>cluster_lda1</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="fu">plot</span>(cluster_lda1)</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>cluster_dscore <span class="ot">&lt;-</span> <span class="fu">predict</span>(cluster_lda1, <span class="at">dimen=</span>nclust)<span class="sc">$</span>x</span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a><span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(nclust<span class="dv">-1</span>)) {</span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a>  <span class="fu">boxplot</span>(cluster_dscore[,j] <span class="sc">~</span> clusternum, <span class="at">xlab=</span><span class="fu">paste</span>(<span class="st">&quot;LD&quot;</span>, <span class="fu">as.character</span>(j), <span class="at">sep=</span><span class="st">&quot;&quot;</span>))</span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a>  }</span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a><span class="fu">cor</span>(orstationc[<span class="dv">5</span><span class="sc">:</span><span class="dv">10</span>],cluster_dscore)</span></code></pre></div>
<p>The discriminant analysis here is used to answer the question “how do
the clusters of stations differ climatically?” In this case, it looks
like <code>pann</code> and <code>tann</code> are the variables most
closely correlated with each discriminant function, and because each of
these variables are more-or-less averages of the seasonal extreme
variables, that might explain why the clusters seem inhomogeneous.</p>
<p>The analysis can be repeated, extracting different numbers of
clusters. The easiest way to do this is to copy the three chunks of code
above into a text editor, changing the assignment to nclust and
including or excluding before executing each chunk, one at a time.</p>
</div>
</div>
<div id="distances-among-observations-and-among-variables"
class="section level1" number="3">
<h1><span class="header-section-number">3</span> Distances: among
observations and among variables</h1>
<p>As noted above, distance matrices are interesting in their own right,
as well as being input to a cluster analysis. In a typical rectangular
data set, there are two distance matrices that could be calculated, one
comparing variables (i.e. the columns in a data frame), and the other
comparing observations (or rows). In the examples below the “TraCE 21ka
decadal data are used. There are (96 x 48) 4608 grid points and 2204
observations. Read the raster brick in the netCDF file
(<code>Trace21_TREFHT_anm2.nc</code>), and reshape it the usual way:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># reshape the 3-d array</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>var_vec_long <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(var_array)</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="fu">length</span>(var_vec_long)</span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">matrix</span>(var_vec_long, <span class="at">nrow =</span> nlon <span class="sc">*</span> nlat, <span class="at">ncol =</span> nt))</span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="fu">dim</span>(X)</span></code></pre></div>
<p>Notice the use of the transposition function (<code>t()</code>) to
get the data in the proper “tidy” arrangement for a data frame
(variables in columns, observations in rows). Distance matrix
calcualation can take a while, but the <code>parallelDist</code> package
can be used to greatly speed things up, by handing off the creation of
different blocks of the matrix to differnt processors in multi-processor
machines.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="co"># load packages</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="co"># distance matrices</span></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a><span class="fu">library</span>(parallelDist)</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a><span class="fu">library</span>(raster)</span></code></pre></div>
<p>The <code>parDist()</code> function gets the dissimilarity
martrix:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="co"># disatances among observations (time)</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>dist_obs <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">parDist</span>(X)) </span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a><span class="fu">dim</span>(dist_obs)</span></code></pre></div>
<p>The dissimilarity martrix can be conveniently visualized by turning
it into a raster, and then using the raster <code>plot()</code> function
to visualize it.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>raster_obs <span class="ot">&lt;-</span> <span class="fu">flip</span>(<span class="fu">raster</span>(dist_obs, <span class="at">xmn=</span><span class="dv">0</span>, <span class="at">ymn=</span><span class="dv">0</span>, <span class="at">xmx=</span><span class="fu">dim</span>(dist_obs)[<span class="dv">1</span>], <span class="at">ymx=</span><span class="fu">dim</span>(dist_obs)[<span class="dv">2</span>]), <span class="at">direction=</span><span class="dv">1</span>)</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>raster_obs</span></code></pre></div>
<p>Note the use of the <code>maxpixels</code> argument to avoid
smoothing in the image. The rendering of the image is faster if the
image is written to a file:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">png</span>(dist_obs_pngfile, <span class="at">width=</span><span class="dv">720</span>, <span class="at">height=</span><span class="dv">720</span>) <span class="co"># open the file</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a><span class="fu">plot</span>(raster_obs, <span class="at">maxpixels=</span><span class="fu">dim</span>(dist_obs)[<span class="dv">1</span>]<span class="sc">*</span><span class="fu">dim</span>(dist_obs)[<span class="dv">2</span>], <span class="at">col=</span><span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">9</span>,<span class="st">&quot;RdPu&quot;</span>))) </span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p>And here’s the image:</p>
<p><img src="TraCE_decadal_dist_var_v1.png" /></p>
<p>The image is sort of a map: The grid points are arranged in the data
frame with longitudes varying rapidly and the sqrare dark area (low
dissimilarities) in the middle is the tropics, where the spatial
variability of climate is low, while the more spatially variable
higher-latitude regions (at all longitudes) frame the low-variability
region.</p>
<p>The dissimilarity matrix for variables is gotten in a similar
fashion, but not the use of the transpose function <code>t()</code> to
get the variables in rows, and observations in columns:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="co"># variables (grid points)</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>dist_var <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">parDist</span>(<span class="fu">t</span>(X))) </span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a><span class="fu">dim</span>(dist_var)</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a></span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>raster_var <span class="ot">&lt;-</span> <span class="fu">flip</span>(<span class="fu">raster</span>(dist_var, <span class="at">xmn=</span><span class="dv">0</span>, <span class="at">ymn=</span><span class="dv">0</span>, <span class="at">xmx=</span><span class="fu">dim</span>(dist_var)[<span class="dv">1</span>], <span class="at">ymx=</span><span class="fu">dim</span>(dist_var)[<span class="dv">2</span>]), <span class="at">direction=</span><span class="dv">1</span>)</span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a>raster_var</span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a><span class="fu">png</span>(dist_var_pngfile, <span class="at">width=</span><span class="dv">720</span>, <span class="at">height=</span><span class="dv">720</span>) <span class="co"># open the file</span></span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a><span class="fu">plot</span>(raster_var, <span class="at">maxpixels=</span><span class="fu">dim</span>(dist_var)[<span class="dv">1</span>]<span class="sc">*</span><span class="fu">dim</span>(dist_var)[<span class="dv">2</span>], <span class="at">col=</span><span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">9</span>,<span class="st">&quot;RdPu&quot;</span>))) </span>
<span id="cb19-10"><a href="#cb19-10" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p><img src="TraCE_decadal_dist_obs_v1.png" /></p>
<p>Here the temporal variability of is emphasized.</p>
</div>
<div id="heatmaps-2-way-time-and-space-clustering"
class="section level1" number="4">
<h1><span class="header-section-number">4</span> Heatmaps (2-way, time
and space clustering)</h1>
<p>“Heatmaps” are a visualization approach that is implemented in the
<code>base</code> package of R. A heatmap basically consists of the
raster-image type plot of the data frame, with the rows and columns of
the matrix reaggranged using cluster analyses of the varables and
observations. (Note that the cluster analyses are carried out
individually, as oppsosed to simultaneously, an approach known as
“biclustering”.) Typically, for small data sets, the matrix plot is
augmented by dendrograms, and by “side bars” which plot some additional
attribute of the data (location or time, for example) using a scatter
plot, gradient plot, etc.</p>
<div id="a-simple-example-with-hovmöller-matrix-data"
class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> A simple example with
Hovmöller-matrix data</h2>
<p>A simple heatmap can be generated for a “Hovmöller matrix” calculated
using the gridded data. The Hovmöller matrix contains longitudinal
averages and is laid out with one row for each latitude and one column
for each time.</p>
<div id="set-up" class="section level3" number="4.1.1">
<h3><span class="header-section-number">4.1.1</span> Set up</h3>
<p>Create the Hovmöller matrix:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="co"># Hovmöller matrix </span></span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, <span class="at">nrow=</span>nlat, <span class="at">ncol=</span>nt)</span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a><span class="cf">for</span> (n <span class="cf">in</span> (<span class="dv">1</span><span class="sc">:</span>nt)) {</span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a>  <span class="cf">for</span> (k <span class="cf">in</span> (<span class="dv">1</span><span class="sc">:</span>nlat)) {</span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a>    <span class="cf">for</span> (j <span class="cf">in</span> (<span class="dv">1</span><span class="sc">:</span>nlon)) {</span>
<span id="cb20-6"><a href="#cb20-6" tabindex="-1"></a>      X[k, n] <span class="ot">&lt;-</span> X[k, n] <span class="sc">+</span> var_array[j, k, n]</span>
<span id="cb20-7"><a href="#cb20-7" tabindex="-1"></a>    }</span>
<span id="cb20-8"><a href="#cb20-8" tabindex="-1"></a>    X[k, n] <span class="ot">&lt;-</span> X[k,n ]<span class="sc">/</span>nlon</span>
<span id="cb20-9"><a href="#cb20-9" tabindex="-1"></a>  }</span>
<span id="cb20-10"><a href="#cb20-10" tabindex="-1"></a>}</span>
<span id="cb20-11"><a href="#cb20-11" tabindex="-1"></a><span class="fu">dim</span>(X)</span>
<span id="cb20-12"><a href="#cb20-12" tabindex="-1"></a>ncols <span class="ot">&lt;-</span> <span class="fu">dim</span>(X)[<span class="dv">2</span>]</span></code></pre></div>
<p>Set row and column names for the matrix.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="co"># set row and column names</span></span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a><span class="fu">rownames</span>(X) <span class="ot">&lt;-</span> <span class="fu">str_pad</span>(<span class="fu">sprintf</span>(<span class="st">&quot;%.3f&quot;</span>, lat), <span class="dv">5</span>, <span class="at">pad=</span><span class="st">&quot;0&quot;</span>)</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rownames</span>(X));<span class="fu">tail</span>(<span class="fu">rownames</span>(X)); <span class="fu">length</span>(<span class="fu">rownames</span>(X))</span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a><span class="fu">colnames</span>(X) <span class="ot">&lt;-</span> <span class="fu">str_pad</span>(<span class="fu">sprintf</span>(<span class="st">&quot;%.3f&quot;</span>, plt_xvals), <span class="dv">5</span>, <span class="at">pad=</span><span class="st">&quot;0&quot;</span>)</span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colnames</span>(X));<span class="fu">tail</span>(<span class="fu">colnames</span>(X)); <span class="fu">length</span>(<span class="fu">colnames</span>(X))</span></code></pre></div>
<p>A matrix plot of the Hovmöller matrix will illustrate the data (and
is an actual Hovmöller plot). To get high (northern) latitudes at the
top of the plot, and high (southern) latitudes at the bottom, flip the
data:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="co"># flip X</span></span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>X <span class="ot">&lt;-</span> X[(nlat<span class="sc">:</span><span class="dv">1</span>),]</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rownames</span>(X)); <span class="fu">tail</span>(<span class="fu">rownames</span>(X)); <span class="fu">length</span>(<span class="fu">rownames</span>(X))</span></code></pre></div>
<p>For this example, the data are decimated, by taking every 20th
observation (to allow the dendrograms to be visualized):</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="co"># subset</span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>X <span class="ot">&lt;-</span> X[,<span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">2201</span>, <span class="at">by=</span><span class="dv">20</span>)]</span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a><span class="fu">dim</span>(X)</span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a>ncols <span class="ot">&lt;-</span> <span class="fu">dim</span>(X)[<span class="dv">2</span>]</span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a><span class="fu">rownames</span>(X)</span></code></pre></div>
<p>More set up, Generate colors for the matrix pot and the side bars,
which here will illustrate the latitude and time of each data point.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="co"># generate colors for heatmap and side bars</span></span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>rcol <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>,<span class="st">&quot;PRGn&quot;</span>))(nlat)</span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a>ccol <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">9</span>,<span class="st">&quot;Purples&quot;</span>))(ncols)</span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a>ncols <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb24-5"><a href="#cb24-5" tabindex="-1"></a>zcol <span class="ot">&lt;-</span> (<span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>,<span class="st">&quot;RdBu&quot;</span>))) <span class="co">#(ncols)</span></span>
<span id="cb24-6"><a href="#cb24-6" tabindex="-1"></a>breaks <span class="ot">=</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">20</span>,<span class="sc">-</span><span class="dv">10</span>,<span class="sc">-</span><span class="dv">5</span>,<span class="sc">-</span><span class="dv">2</span>,<span class="sc">-</span><span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">5</span>,<span class="dv">10</span>,<span class="dv">20</span>)</span>
<span id="cb24-7"><a href="#cb24-7" tabindex="-1"></a><span class="fu">length</span>(breaks)</span></code></pre></div>
</div>
<div id="hovmöller-matrix-plot-decimated-decadal-data"
class="section level3" number="4.1.2">
<h3><span class="header-section-number">4.1.2</span> Hovmöller-matrix
plot (decimated decadal data)</h3>
<p>Produce the heatmap for the Hovmöller matrix:</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a>png_file <span class="ot">&lt;-</span> <span class="fu">paste</span>(file_label, <span class="st">&quot;_heatmap_hov_01&quot;</span>, <span class="st">&quot;.png&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a><span class="fu">png</span>(png_file, <span class="at">width=</span><span class="dv">960</span>, <span class="at">height=</span><span class="dv">960</span>)</span>
<span id="cb25-3"><a href="#cb25-3" tabindex="-1"></a><span class="fu">system.time</span>( X_heatmap <span class="ot">&lt;-</span> <span class="fu">heatmap.2</span>(X, <span class="at">cexRow =</span> <span class="fl">0.8</span>, <span class="at">cexCol =</span> <span class="fl">0.8</span>, <span class="at">scale=</span><span class="st">&quot;none&quot;</span>, </span>
<span id="cb25-4"><a href="#cb25-4" tabindex="-1"></a>        <span class="at">Rowv=</span><span class="cn">NA</span>, <span class="at">Colv=</span><span class="cn">NA</span>, <span class="at">dendrogram =</span> <span class="st">&quot;none&quot;</span>, <span class="co"># just plot data, don&#39;t do clustering</span></span>
<span id="cb25-5"><a href="#cb25-5" tabindex="-1"></a>        <span class="at">RowSideColors=</span>rcol, <span class="at">ColSideColors=</span>ccol, <span class="at">col=</span>zcol, <span class="at">breaks=</span>breaks, <span class="at">trace=</span><span class="st">&quot;none&quot;</span>, <span class="at">density.info =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb25-6"><a href="#cb25-6" tabindex="-1"></a>        <span class="at">lmat=</span><span class="fu">rbind</span>(<span class="fu">c</span>(<span class="dv">6</span>, <span class="dv">0</span>, <span class="dv">5</span>), <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">2</span>), <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">3</span>)), <span class="at">lwid=</span><span class="fu">c</span>(<span class="fl">1.0</span>, <span class="fl">0.2</span>, <span class="fl">5.0</span>), <span class="at">lhei =</span> <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="fl">0.2</span>, <span class="fl">4.0</span>), </span>
<span id="cb25-7"><a href="#cb25-7" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;ka&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Latitude&quot;</span>, <span class="at">key =</span> <span class="cn">TRUE</span>, <span class="at">key.title =</span> <span class="cn">NA</span>) </span>
<span id="cb25-8"><a href="#cb25-8" tabindex="-1"></a>)</span>
<span id="cb25-9"><a href="#cb25-9" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p><img src="TraCE_decadal_heatmap_hov_01.png" /></p>
<p>Notice the way the side bars work in illustrating the location and
time of each data value.</p>
</div>
<div id="hovmöller-matrix-heatmap-decimated-decadal-data"
class="section level3" number="4.1.3">
<h3><span class="header-section-number">4.1.3</span> Hovmöller-matrix
heatmap (decimated decadal data)</h3>
<p>Now here’s the heatmap:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a>png_file <span class="ot">&lt;-</span> <span class="fu">paste</span>(file_label, <span class="st">&quot;_heatmap_01&quot;</span>, <span class="st">&quot;.png&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a><span class="fu">png</span>(png_file, <span class="at">width=</span><span class="dv">960</span>, <span class="at">height=</span><span class="dv">960</span>)</span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a><span class="fu">system.time</span>( X_heatmap <span class="ot">&lt;-</span> <span class="fu">heatmap.2</span>(X, <span class="at">cexRow =</span> <span class="fl">0.8</span>, <span class="at">cexCol =</span> <span class="fl">0.8</span>, <span class="at">scale=</span><span class="st">&quot;none&quot;</span>, </span>
<span id="cb26-4"><a href="#cb26-4" tabindex="-1"></a>        <span class="at">RowSideColors=</span>rcol, <span class="at">ColSideColors=</span>ccol, <span class="at">col=</span>zcol, <span class="at">breaks=</span>breaks, <span class="at">trace=</span><span class="st">&quot;none&quot;</span>, <span class="at">density.info =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb26-5"><a href="#cb26-5" tabindex="-1"></a>        <span class="at">lmat=</span><span class="fu">rbind</span>(<span class="fu">c</span>(<span class="dv">6</span>, <span class="dv">0</span>, <span class="dv">5</span>), <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">2</span>), <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">3</span>)), <span class="at">lwid=</span><span class="fu">c</span>(<span class="fl">1.0</span>, <span class="fl">0.2</span>, <span class="fl">5.0</span>), <span class="at">lhei =</span> <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="fl">0.2</span>, <span class="fl">4.0</span>), </span>
<span id="cb26-6"><a href="#cb26-6" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;ka&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Latitude&quot;</span>, <span class="at">key =</span> <span class="cn">TRUE</span>, <span class="at">key.title =</span> <span class="cn">NA</span>) </span>
<span id="cb26-7"><a href="#cb26-7" tabindex="-1"></a>)</span>
<span id="cb26-8"><a href="#cb26-8" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p><img src="TraCE_decadal_heatmap_01.png" /></p>
<p>The heatmap has rearranged the data according to the cluster analyses
of the observations and variables, placing more similar data points next
to one another. Notice how the side bars allow one to figure out the
location and time of individual points. Here, and not surprisingly,
there appear to be four distinct time-space clusters (or “regimes”) in
the data): glacial vs. interglacial (Holocene) and high latitudes (in
both hemispheres) vs. low latitudes.</p>
</div>
</div>
<div id="a-second-example-with-decadal-data" class="section level2"
number="4.2">
<h2><span class="header-section-number">4.2</span> A second example with
decadal data</h2>
<p>A second example is given using the decadal data directly. Here,
there are too many points to make the dendrograms interpretable in any
way, and so they are suppressed.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="co"># generate colors for horizontal side bar</span></span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a>ccol <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">9</span>,<span class="st">&quot;Purples&quot;</span>))(ncols)</span></code></pre></div>
<div id="hovmöller-matrix-plot" class="section level3" number="4.2.1">
<h3><span class="header-section-number">4.2.1</span> Hovmöller-matrix
plot</h3>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a>png_file <span class="ot">&lt;-</span> <span class="fu">paste</span>(file_label, <span class="st">&quot;_heatmap_hov_01&quot;</span>, <span class="st">&quot;.png&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a><span class="fu">png</span>(png_file, <span class="at">width=</span><span class="dv">960</span>, <span class="at">height=</span><span class="dv">960</span>)</span>
<span id="cb28-3"><a href="#cb28-3" tabindex="-1"></a><span class="fu">system.time</span>( X_heatmap <span class="ot">&lt;-</span> <span class="fu">heatmap.2</span>(X, <span class="at">cexRow =</span> <span class="fl">0.8</span>, <span class="at">cexCol =</span> <span class="fl">0.8</span>, <span class="at">scale=</span><span class="st">&quot;none&quot;</span>, <span class="at">Rowv=</span><span class="cn">NA</span>, <span class="at">Colv=</span><span class="cn">NA</span>, <span class="at">dendrogram =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb28-4"><a href="#cb28-4" tabindex="-1"></a>        <span class="at">RowSideColors=</span>rcol, <span class="at">ColSideColors=</span>ccol, <span class="at">col=</span>zcol, <span class="at">breaks=</span>breaks, <span class="at">trace=</span><span class="st">&quot;none&quot;</span>, <span class="at">density.info =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb28-5"><a href="#cb28-5" tabindex="-1"></a>        <span class="at">lmat=</span><span class="fu">rbind</span>(<span class="fu">c</span>(<span class="dv">6</span>, <span class="dv">0</span>, <span class="dv">5</span>), <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">2</span>), <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">3</span>)), <span class="at">lwid=</span><span class="fu">c</span>(<span class="fl">1.0</span>, <span class="fl">0.2</span>, <span class="fl">5.0</span>), <span class="at">lhei =</span> <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="fl">0.2</span>, <span class="fl">4.0</span>), </span>
<span id="cb28-6"><a href="#cb28-6" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;ka&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Latitude&quot;</span>, <span class="at">key =</span> <span class="cn">TRUE</span>, <span class="at">key.title =</span> <span class="cn">NA</span>) </span>
<span id="cb28-7"><a href="#cb28-7" tabindex="-1"></a>)</span>
<span id="cb28-8"><a href="#cb28-8" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p><img src="TraCE_decadal_heatmap_hov_02.png" /></p>
</div>
<div id="hovmöller-matrix-heatmap" class="section level3"
number="4.2.2">
<h3><span class="header-section-number">4.2.2</span> Hovmöller-matrix
heatmap</h3>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a>png_file <span class="ot">&lt;-</span> <span class="fu">paste</span>(file_label, <span class="st">&quot;_heatmap_01&quot;</span>, <span class="st">&quot;.png&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a><span class="fu">png</span>(png_file, <span class="at">width=</span><span class="dv">960</span>, <span class="at">height=</span><span class="dv">960</span>)</span>
<span id="cb29-3"><a href="#cb29-3" tabindex="-1"></a><span class="fu">system.time</span>( X_heatmap <span class="ot">&lt;-</span> <span class="fu">heatmap.2</span>(X, <span class="at">cexRow =</span> <span class="fl">0.8</span>, <span class="at">cexCol =</span> <span class="fl">0.8</span>, <span class="at">scale=</span><span class="st">&quot;none&quot;</span>, <span class="co"># Rowv=NA, Colv=NA, dendrogram = &quot;none&quot;,</span></span>
<span id="cb29-4"><a href="#cb29-4" tabindex="-1"></a>        <span class="at">RowSideColors=</span>rcol, <span class="at">ColSideColors=</span>ccol, <span class="at">col=</span>zcol, <span class="at">breaks=</span>breaks, <span class="at">trace=</span><span class="st">&quot;none&quot;</span>, <span class="at">density.info =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb29-5"><a href="#cb29-5" tabindex="-1"></a>        <span class="at">lmat=</span><span class="fu">rbind</span>(<span class="fu">c</span>(<span class="dv">6</span>, <span class="dv">0</span>, <span class="dv">5</span>), <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">2</span>), <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">3</span>)), <span class="at">lwid=</span><span class="fu">c</span>(<span class="fl">1.0</span>, <span class="fl">0.2</span>, <span class="fl">5.0</span>), <span class="at">lhei =</span> <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="fl">0.2</span>, <span class="fl">4.0</span>), </span>
<span id="cb29-6"><a href="#cb29-6" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;ka&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Latitude&quot;</span>, <span class="at">key =</span> <span class="cn">TRUE</span>, <span class="at">key.title =</span> <span class="cn">NA</span>) </span>
<span id="cb29-7"><a href="#cb29-7" tabindex="-1"></a>)</span>
<span id="cb29-8"><a href="#cb29-8" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p><img src="TraCE_decadal_heatmap_02.png" /></p>
</div>
</div>
<div id="full-grid-example" class="section level2" number="4.3">
<h2><span class="header-section-number">4.3</span> Full-grid
example</h2>
<p>Finally, a heatmap of the gridded data (as opposed to the Hovmöller
matrix data) can be obtained</p>
<div id="set-up-1" class="section level3" number="4.3.1">
<h3><span class="header-section-number">4.3.1</span> Set up</h3>
<p>There are several set up steps necessary. Reshape the array, and
generate row and column names:</p>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" tabindex="-1"></a><span class="co"># reshape the 3-d array</span></span>
<span id="cb30-2"><a href="#cb30-2" tabindex="-1"></a>var_vec_long <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(var_array)</span>
<span id="cb30-3"><a href="#cb30-3" tabindex="-1"></a><span class="fu">length</span>(var_vec_long)</span>
<span id="cb30-4"><a href="#cb30-4" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">matrix</span>(var_vec_long, <span class="at">nrow =</span> nlon <span class="sc">*</span> nlat, <span class="at">ncol =</span> nt) <span class="co"># Don&#39;t transform as usual..</span></span>
<span id="cb30-5"><a href="#cb30-5" tabindex="-1"></a><span class="fu">dim</span>(X)</span>
<span id="cb30-6"><a href="#cb30-6" tabindex="-1"></a>ncols <span class="ot">&lt;-</span> <span class="fu">dim</span>(x)[<span class="dv">3</span>]</span></code></pre></div>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="co"># generaate row and column names</span></span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a><span class="fu">rownames</span>(X) <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;E&quot;</span>, <span class="fu">str_pad</span>(<span class="fu">sprintf</span>(<span class="st">&quot;%.2f&quot;</span>, <span class="fu">round</span>(grid<span class="sc">$</span>lon, <span class="dv">2</span>)), <span class="dv">5</span>, <span class="at">pad=</span><span class="st">&quot;0&quot;</span>),</span>
<span id="cb31-3"><a href="#cb31-3" tabindex="-1"></a>  <span class="st">&quot;N&quot;</span>, <span class="fu">str_pad</span>(<span class="fu">sprintf</span>(<span class="st">&quot;%.2f&quot;</span>, <span class="fu">round</span>(grid<span class="sc">$</span>lat, <span class="dv">2</span>)), <span class="dv">5</span>, <span class="at">pad=</span><span class="st">&quot;0&quot;</span>), <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb31-4"><a href="#cb31-4" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rownames</span>(X), <span class="dv">20</span>); <span class="fu">tail</span>(<span class="fu">rownames</span>(X), <span class="dv">20</span>); <span class="fu">length</span>(<span class="fu">rownames</span>(X))</span>
<span id="cb31-5"><a href="#cb31-5" tabindex="-1"></a></span>
<span id="cb31-6"><a href="#cb31-6" tabindex="-1"></a><span class="fu">colnames</span>(X) <span class="ot">&lt;-</span> <span class="fu">str_pad</span>(<span class="fu">sprintf</span>(<span class="st">&quot;%.3f&quot;</span>, t), <span class="dv">5</span>, <span class="at">pad=</span><span class="st">&quot;0&quot;</span>)</span>
<span id="cb31-7"><a href="#cb31-7" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colnames</span>(X)); <span class="fu">tail</span>(<span class="fu">colnames</span>(X)); <span class="fu">length</span>(<span class="fu">colnames</span>(X))</span></code></pre></div>
<p>Flip the data…</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="co"># flip X</span></span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a>X <span class="ot">&lt;-</span> X[((nlon<span class="sc">*</span>nlat)<span class="sc">:</span><span class="dv">1</span>),]</span>
<span id="cb32-3"><a href="#cb32-3" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rownames</span>(X)); <span class="fu">tail</span>(<span class="fu">rownames</span>(X)); <span class="fu">length</span>(<span class="fu">rownames</span>(X))</span></code></pre></div>
<p>… and generate colors:</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a><span class="co"># generate colors for vertical side bar</span></span>
<span id="cb33-2"><a href="#cb33-2" tabindex="-1"></a>icol <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>,<span class="st">&quot;PRGn&quot;</span>))(nlat)</span>
<span id="cb33-3"><a href="#cb33-3" tabindex="-1"></a>idx <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">rep</span>((<span class="dv">1</span><span class="sc">:</span>nlat), nlon))</span>
<span id="cb33-4"><a href="#cb33-4" tabindex="-1"></a>rcol <span class="ot">&lt;-</span> icol[idx]</span>
<span id="cb33-5"><a href="#cb33-5" tabindex="-1"></a>ccol <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">9</span>,<span class="st">&quot;Purples&quot;</span>))(nt)</span></code></pre></div>
</div>
<div id="matrix-plot" class="section level3" number="4.3.2">
<h3><span class="header-section-number">4.3.2</span> Matrix plot</h3>
<p>Here’s a matrix plot of the data:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a>png_file <span class="ot">&lt;-</span> <span class="fu">paste</span>(file_label, <span class="st">&quot;_heatmap_hov_03&quot;</span>, <span class="st">&quot;.png&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb34-2"><a href="#cb34-2" tabindex="-1"></a><span class="fu">png</span>(png_file, <span class="at">width=</span><span class="dv">960</span>, <span class="at">height=</span><span class="dv">960</span>)</span>
<span id="cb34-3"><a href="#cb34-3" tabindex="-1"></a><span class="fu">system.time</span>( X_heatmap <span class="ot">&lt;-</span> <span class="fu">heatmap.2</span>(X, <span class="at">cexRow =</span> <span class="fl">0.8</span>, <span class="at">cexCol =</span> <span class="fl">0.8</span>, <span class="at">scale=</span><span class="st">&quot;none&quot;</span>, </span>
<span id="cb34-4"><a href="#cb34-4" tabindex="-1"></a>        <span class="at">Rowv=</span><span class="cn">NA</span>, <span class="at">Colv=</span><span class="cn">NA</span>, <span class="at">dendrogram =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb34-5"><a href="#cb34-5" tabindex="-1"></a>        <span class="at">RowSideColors=</span>rcol, <span class="at">ColSideColors=</span>ccol, <span class="at">col=</span>zcol, <span class="at">breaks=</span>breaks, <span class="at">trace=</span><span class="st">&quot;none&quot;</span>, <span class="at">density.info =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb34-6"><a href="#cb34-6" tabindex="-1"></a>        <span class="at">lmat=</span><span class="fu">rbind</span>(<span class="fu">c</span>(<span class="dv">6</span>, <span class="dv">0</span>, <span class="dv">5</span>), <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">2</span>), <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">3</span>)), <span class="at">lwid=</span><span class="fu">c</span>(<span class="fl">1.0</span>, <span class="fl">0.2</span>, <span class="fl">5.0</span>), <span class="at">lhei =</span> <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="fl">0.2</span>, <span class="fl">4.0</span>), </span>
<span id="cb34-7"><a href="#cb34-7" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;ka&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Locations&quot;</span>, <span class="at">key =</span> <span class="cn">TRUE</span>, <span class="at">key.title =</span> <span class="cn">NA</span>, <span class="at">labRow =</span> <span class="cn">FALSE</span>, <span class="at">labCol =</span> <span class="cn">FALSE</span>) </span>
<span id="cb34-8"><a href="#cb34-8" tabindex="-1"></a>)</span>
<span id="cb34-9"><a href="#cb34-9" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>##    user  system elapsed 
## 1700.08   42.51 1744.53</code></pre>
<p><img src="TraCE_decadal_heatmap_hov_03.png" /></p>
</div>
<div id="heatmap" class="section level3" number="4.3.3">
<h3><span class="header-section-number">4.3.3</span> Heatmap</h3>
<p>Heatmap of the grid-point data</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a>png_file <span class="ot">&lt;-</span> <span class="fu">paste</span>(file_label, <span class="st">&quot;_heatmap_03&quot;</span>, <span class="st">&quot;.png&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb36-2"><a href="#cb36-2" tabindex="-1"></a><span class="fu">png</span>(png_file, <span class="at">width=</span><span class="dv">960</span>, <span class="at">height=</span><span class="dv">960</span>)</span>
<span id="cb36-3"><a href="#cb36-3" tabindex="-1"></a></span>
<span id="cb36-4"><a href="#cb36-4" tabindex="-1"></a><span class="fu">system.time</span>( X_heatmap <span class="ot">&lt;-</span> <span class="fu">heatmap.2</span>(X, <span class="at">cexRow =</span> <span class="fl">0.8</span>, <span class="at">cexCol =</span> <span class="fl">0.8</span>, <span class="at">scale=</span><span class="st">&quot;none&quot;</span>, <span class="at">dendrogram =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb36-5"><a href="#cb36-5" tabindex="-1"></a>        <span class="at">RowSideColors=</span>rcol, <span class="at">ColSideColors=</span>ccol, <span class="at">col=</span>zcol, <span class="at">breaks=</span>breaks, <span class="at">trace=</span><span class="st">&quot;none&quot;</span>, <span class="at">density.info =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb36-6"><a href="#cb36-6" tabindex="-1"></a>        <span class="at">lmat=</span><span class="fu">rbind</span>(<span class="fu">c</span>(<span class="dv">6</span>, <span class="dv">0</span>, <span class="dv">5</span>), <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">2</span>), <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">3</span>)), <span class="at">lwid=</span><span class="fu">c</span>(<span class="fl">1.0</span>, <span class="fl">0.2</span>, <span class="fl">5.0</span>), <span class="at">lhei =</span> <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="fl">0.2</span>, <span class="fl">4.0</span>), </span>
<span id="cb36-7"><a href="#cb36-7" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;ka&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Locations&quot;</span>, <span class="at">key =</span> <span class="cn">TRUE</span>, <span class="at">key.title =</span> <span class="cn">NA</span>, <span class="at">labRow =</span> <span class="cn">FALSE</span>, <span class="at">labCol =</span> <span class="cn">FALSE</span>) </span>
<span id="cb36-8"><a href="#cb36-8" tabindex="-1"></a>)</span></code></pre></div>
<pre><code>##    user  system elapsed 
##  643.89   42.26  688.10 </code></pre>
<p><img src="TraCE_decadal_heatmap_03.png" /></p>
</div>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
