<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Cluster analysis</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; } /* Alert */
code span.an { color: #008000; } /* Annotation */
code span.at { } /* Attribute */
code span.bu { } /* BuiltIn */
code span.cf { color: #0000ff; } /* ControlFlow */
code span.ch { color: #008080; } /* Char */
code span.cn { } /* Constant */
code span.co { color: #008000; } /* Comment */
code span.cv { color: #008000; } /* CommentVar */
code span.do { color: #008000; } /* Documentation */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.im { } /* Import */
code span.in { color: #008000; } /* Information */
code span.kw { color: #0000ff; } /* Keyword */
code span.op { } /* Operator */
code span.ot { color: #ff4000; } /* Other */
code span.pp { color: #ff4000; } /* Preprocessor */
code span.sc { color: #008080; } /* SpecialChar */
code span.ss { color: #008080; } /* SpecialString */
code span.st { color: #008080; } /* String */
code span.va { } /* Variable */
code span.vs { color: #008080; } /* VerbatimString */
code span.wa { color: #008000; font-weight: bold; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link rel="stylesheet" href="html-md-01.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">R for Earth-System Science</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="overview.html">Overview</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Topics
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Lecture topics</li>
    <li>
      <a href="intro.html">Introduction: R, RStudio, and other infrastructure</a>
    </li>
    <li>
      <a href="usingR.html">Using R</a>
    </li>
    <li>
      <a href="ESSdata.html">Earth-system science data</a>
    </li>
    <li>
      <a href="netCDF.html">netCDF in R</a>
    </li>
    <li>
      <a href="netcdf_terra.html">netCDF and terra</a>
    </li>
    <li>
      <a href="plots1.html">Plots (1)</a>
    </li>
    <li>
      <a href="plots2.html">Plots (2)</a>
    </li>
    <li>
      <a href="maps1.html">Maps (1)</a>
    </li>
    <li>
      <a href="maps2.html">Maps (2)</a>
    </li>
    <li>
      <a href="geospatial.html">Geospatial analyses</a>
    </li>
    <li>
      <a href="regression1.html">Regression 1</a>
    </li>
    <li>
      <a href="regression2.html">Regression 2</a>
    </li>
    <li>
      <a href="pca.html">PCA</a>
    </li>
    <li>
      <a href="highdim.html">High-resolution &amp; high-dimensional data</a>
    </li>
    <li>
      <a href="multivariate.html">Multivariate methods</a>
    </li>
    <li>
      <a href="timeseries.html">Time-series analysis</a>
    </li>
    <li class="dropdown-header">Other topics</li>
    <li>
      <a href="sf_stars_sftime.html">sf / stars / sftime</a>
    </li>
    <li>
      <a href="hdf5_intro.html">HDF in R</a>
    </li>
    <li>
      <a href="anomalies.html">Long-term means and anomalies</a>
    </li>
    <li>
      <a href="tsplots.html">Time-series plots</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tasks
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Tasks</li>
    <li>
      <a href="installR.html">Install R and RStudio</a>
    </li>
    <li>
      <a href="Ex_UsingR.html">Using R exercise</a>
    </li>
    <li>
      <a href="install_netCDF.html">Install netCDF &amp; Panoply</a>
    </li>
    <li>
      <a href="transfer.html">File transfer</a>
    </li>
    <li>
      <a href="texteditor.html">Install a text editor</a>
    </li>
    <li>
      <a href="dataselection.html">Project data set selection</a>
    </li>
    <li>
      <a href="markdown.html">Markdown and RMarkdown</a>
    </li>
    <li>
      <a href="scripts_and_md.html">R Scripts and R Markdown</a>
    </li>
    <li>
      <a href="github.html">GitHub</a>
    </li>
    <li>
      <a href="pages.html">pages.uoregon.edu</a>
    </li>
    <li>
      <a href="localweb.html">Local websites</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Other
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="readings.html">Readings</a>
    </li>
    <li>
      <a href="websites.html">Old course web pages</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="about.html">About</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Cluster analysis</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#introduction" id="toc-introduction"><span
class="toc-section-number">1</span> Introduction</a></li>
<li><a href="#a-simple-cluster-analysis"
id="toc-a-simple-cluster-analysis"><span
class="toc-section-number">2</span> A simple cluster analysis</a>
<ul>
<li><a href="#distancedissimilarity-matrices"
id="toc-distancedissimilarity-matrices"><span
class="toc-section-number">2.1</span> Distance/Dissimilarity
matrices</a></li>
<li><a href="#the-cluster-analysis" id="toc-the-cluster-analysis"><span
class="toc-section-number">2.2</span> The cluster analysis</a></li>
<li><a href="#examining-the-cluster-analysis"
id="toc-examining-the-cluster-analysis"><span
class="toc-section-number">2.3</span> Examining the cluster
analysis</a></li>
</ul></li>
<li><a href="#distances-among-observations-and-among-variables"
id="toc-distances-among-observations-and-among-variables"><span
class="toc-section-number">3</span> Distances: among observations and
among variables</a>
<ul>
<li><a href="#distances-among-times"
id="toc-distances-among-times"><span
class="toc-section-number">3.1</span> Distances among times</a></li>
<li><a href="#distance-among-locations"
id="toc-distance-among-locations"><span
class="toc-section-number">3.2</span> Distance among locations</a></li>
</ul></li>
<li><a href="#heatmaps-2-way-time-and-space-clustering"
id="toc-heatmaps-2-way-time-and-space-clustering"><span
class="toc-section-number">4</span> Heatmaps (2-way, time-and-space
clustering)</a>
<ul>
<li><a href="#a-simple-example-with-hovmöller-matrix-data"
id="toc-a-simple-example-with-hovmöller-matrix-data"><span
class="toc-section-number">4.1</span> A simple example with
Hovmöller-matrix data</a>
<ul>
<li><a href="#set-up" id="toc-set-up"><span
class="toc-section-number">4.1.1</span> Set up</a></li>
<li><a href="#hovmöller-matrix-plot-subsetted-decadal-data"
id="toc-hovmöller-matrix-plot-subsetted-decadal-data"><span
class="toc-section-number">4.1.2</span> Hovmöller-matrix plot (subsetted
decadal data)</a></li>
<li><a href="#hovmöller-matrix-heatmap-subsetted-decadal-data"
id="toc-hovmöller-matrix-heatmap-subsetted-decadal-data"><span
class="toc-section-number">4.1.3</span> Hovmöller-matrix heatmap
(subsetted decadal data)</a></li>
</ul></li>
<li><a href="#a-second-example-with-decadal-data"
id="toc-a-second-example-with-decadal-data"><span
class="toc-section-number">4.2</span> A second example with decadal
data</a>
<ul>
<li><a href="#hovmöller-matrix-plot"
id="toc-hovmöller-matrix-plot"><span
class="toc-section-number">4.2.1</span> Hovmöller-matrix plot</a></li>
<li><a href="#hovmöller-matrix-heatmap"
id="toc-hovmöller-matrix-heatmap"><span
class="toc-section-number">4.2.2</span> Hovmöller-matrix
heatmap</a></li>
</ul></li>
<li><a href="#full-grid-example" id="toc-full-grid-example"><span
class="toc-section-number">4.3</span> Full-grid example</a>
<ul>
<li><a href="#set-up-1" id="toc-set-up-1"><span
class="toc-section-number">4.3.1</span> Set up</a></li>
<li><a href="#matrix-plot" id="toc-matrix-plot"><span
class="toc-section-number">4.3.2</span> Matrix plot</a></li>
<li><a href="#heatmap" id="toc-heatmap"><span
class="toc-section-number">4.3.3</span> Heatmap</a></li>
</ul></li>
</ul></li>
</ul>
</div>

<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('right', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<p>
<span style="color: #00cc00;">NOTE: This page has been revised for
Winter 2024, but may undergo further edits.</span>
</p>
<div id="introduction" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>“Cluster analysis” is a broad group of multivariate analyses that
have as their objective the organization of individual observations
(objects, sites, grid points, individuals), and these analyses are built
upon the concept of multivariate distances (expressed either as
similarities or dissimilarities) among the objects.</p>
<p>The organization generally takes two forms:</p>
<ul>
<li>the arrangement of the objects in a lower-dimensional space than the
data were originally observed on;</li>
<li>the development of “natural” clusters or the classifications of the
objects.</li>
</ul>
<p>These analyses share many concepts and techniques (both numerical and
practical) with other procedures such as principal components analysis,
numerical taxonomy, discriminant analysis and so on, and cluster
analysis is one of the few machine learning techniques that uses its
traditional name in multivariate analysis as its modern Data
Science/Machine Learning name.</p>
<p>The analyses generally begin with the construction of an <em>n</em> x
<em>n</em> matrix <strong>D</strong> of the distances between objects.
For example, in a two-dimensional space, the elements
<em>d<sub>ij</sub></em> of <strong>D</strong> could be the Euclidean
distances between points,</p>
<p><em>d<sub>ij</sub></em> = [(<em>x<sub>i1</sub></em>-
<em>x<sub>j1</sub></em>)<sup>2</sup> + (<em>x<sub>i1</sub></em>-
<em>x<sub>j1</sub></em>)<sup>2</sup>]<sup>½</sup></p>
<p>The Euclidean distance, and related measures are easily generalized
to more than two dimensions.</p>
<p>To begin load some packages:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(sf)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">library</span>(lattice)</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="fu">library</span>(GGally)</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="fu">library</span>(MASS)</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="fu">library</span>(ncdf4)</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="fu">library</span>(terra)</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="fu">library</span>(tidyterra)</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="fu">library</span>(parallelDist)</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="fu">library</span>(gplots)</span></code></pre></div>
<p>Some of these packages are familiar, but <code>{GGally}</code>
provides some enhanced <code>{ggplot2}</code> plots,
<code>{stringr}</code> manages the manipulation of strings,
<code>{parallelDist}</code> exploits multiple processors/cores to
efficiently calculate large dissimilarlty/distance matrices, and
’{gplots}` calculates heat maps.</p>
</div>
<div id="a-simple-cluster-analysis" class="section level1" number="2">
<h1><span class="header-section-number">2</span> A simple cluster
analysis</h1>
<p>The basic idea of cluster analysis can be illustrated using the
Oregon climate-station dataset, with the goal being to identify groups
of stations with similar climates, which also regionalizes the climate
of the state. Begin by reading a shape file for mapping the data, and
the data itself.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># read county outlines</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="co"># orotl_sf</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>shapefile <span class="ot">&lt;-</span> <span class="st">&quot;/Users/bartlein/Projects/RESS/data/shp_files/OR/orotl.shp&quot;</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>orotl_sf <span class="ot">&lt;-</span> <span class="fu">st_read</span>(shapefile)</span></code></pre></div>
<pre><code>## Reading layer `orotl&#39; from data source `/Users/bartlein/Projects/RESS/data/shp_files/OR/orotl.shp&#39; using driver `ESRI Shapefile&#39;
## Simple feature collection with 36 features and 1 field
## Geometry type: POLYGON
## Dimension:     XY
## Bounding box:  xmin: -124.5584 ymin: 41.98779 xmax: -116.4694 ymax: 46.23626
## CRS:           NA</code></pre>
<p>Add a coordinate reference system and plot:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># add a CRS</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">st_crs</span>(orotl_sf) <span class="ot">&lt;-</span> <span class="fu">st_crs</span>(<span class="st">&quot;+proj=longlat&quot;</span>)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>or_otl_sf <span class="ot">&lt;-</span> <span class="fu">st_geometry</span>(orotl_sf)</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="fu">plot</span>(or_otl_sf) </span></code></pre></div>
<p><img src="multivariate_files/figure-html/multivariate-4-1.png" width="75%" height="75%" /></p>
<p>Now read a *.csv file, and attach it.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># read stations</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>orstationc_path <span class="ot">&lt;-</span> <span class="st">&quot;/Users/bartlein/Projects/RESS/data/csv_files/&quot;</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>orstationc_name <span class="ot">&lt;-</span> <span class="st">&quot;orstationc.csv&quot;</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>orstationc_file <span class="ot">&lt;-</span> <span class="fu">paste</span>(orstationc_path, orstationc_name, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>orstationc <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(orstationc_file)</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="fu">summary</span>(orstationc)</span></code></pre></div>
<pre><code>##    station               lat             lon              elev              tjan             tjul      
##  Length:92          Min.   :42.05   Min.   :-124.6   Min.   :   2.00   Min.   :-6.000   Min.   :12.20  
##  Class :character   1st Qu.:43.60   1st Qu.:-123.2   1st Qu.:  94.25   1st Qu.:-1.225   1st Qu.:17.30  
##  Mode  :character   Median :44.46   Median :-121.7   Median : 462.00   Median : 0.750   Median :19.20  
##                     Mean   :44.32   Mean   :-121.3   Mean   : 556.99   Mean   : 1.417   Mean   :19.01  
##                     3rd Qu.:45.28   3rd Qu.:-119.6   3rd Qu.: 863.75   3rd Qu.: 4.150   3rd Qu.:20.20  
##                     Max.   :46.15   Max.   :-117.0   Max.   :1974.00   Max.   : 8.400   Max.   :26.40  
##       tann             pjan             pjul            pann            idnum            Name          
##  Min.   : 3.200   Min.   : 14.00   Min.   : 4.00   Min.   : 198.0   Min.   :350197   Length:92         
##  1st Qu.: 8.700   1st Qu.: 39.75   1st Qu.: 7.00   1st Qu.: 299.0   1st Qu.:352308   Class :character  
##  Median :10.400   Median : 87.50   Median : 9.00   Median : 531.0   Median :354564   Mode  :character  
##  Mean   : 9.885   Mean   :141.67   Mean   :10.84   Mean   : 868.6   Mean   :354455                     
##  3rd Qu.:11.200   3rd Qu.:239.25   3rd Qu.:12.25   3rd Qu.:1355.8   3rd Qu.:356663                     
##  Max.   :12.600   Max.   :414.00   Max.   :34.00   Max.   :2517.0   Max.   :359316</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">attach</span>(orstationc)</span></code></pre></div>
<p>Plot the station locations, and then the three-letter station name
abbreviations. Notice that the use of the <code>coord_sf()</code>
function in the first plot to explicity specify the geographical extent
of the map, while in the second case <code>ggplot2()</code> determines
the extent.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># plot stations</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data =</span> or_otl_sf, <span class="at">fill =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data =</span> orstationc, <span class="fu">aes</span>(<span class="at">x=</span>lon, <span class="at">y=</span>lat), <span class="at">shape=</span><span class="dv">21</span>, <span class="at">color=</span><span class="st">&quot;black&quot;</span>, <span class="at">size =</span> <span class="dv">3</span>, <span class="at">fill=</span><span class="st">&quot;lightblue&quot;</span>) <span class="sc">+</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>  <span class="fu">coord_sf</span>(<span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">125</span>, <span class="sc">-</span><span class="dv">116</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">41</span>, <span class="dv">47</span>), <span class="at">expand =</span> <span class="cn">FALSE</span>) <span class="sc">+</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Oregon Climate Stations&quot;</span>, <span class="at">x =</span> <span class="st">&quot;Longitude&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Latitude&quot;</span>) <span class="sc">+</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="multivariate_files/figure-html/multivariate-6-1.png" width="75%" height="75%" /></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># plot station names</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data =</span> or_otl_sf, <span class="at">fill =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data =</span> orstationc, <span class="fu">aes</span>(<span class="at">x =</span>lon, <span class="at">y =</span> lat, <span class="at">label =</span> <span class="fu">as.character</span>(station)), </span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>            <span class="at">check_overlap =</span> <span class="cn">TRUE</span>, <span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Oregon Climate Stations&quot;</span>, <span class="at">x =</span> <span class="st">&quot;Longitude&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Latitude&quot;</span>) <span class="sc">+</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="multivariate_files/figure-html/multivariate-7-1.png" width="75%" height="75%" /></p>
<p>Take a quick look at the correlations among variables.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># matrix scatter plot and correlations</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">ggpairs</span>(orstationc[<span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>])</span></code></pre></div>
<p><img src="multivariate_files/figure-html/multivariate-8-1.png" width="75%" height="75%" /></p>
<p>The off-diagonal cells for the climate variables indicate that there
are some strong correlations among the climate variables (the asterisks
indicate significance in the usual way).</p>
<div id="distancedissimilarity-matrices" class="section level2"
number="2.1">
<h2><span class="header-section-number">2.1</span>
Distance/Dissimilarity matrices</h2>
<p>In this example, the groups of stations (or the climate regions of
Oregon, as expressed in the climate-station data) will be defined via an
“agglomerative-hierarchical” cluster analysis. In such an analysis,
individual objects (observations) are combined according to their
(dis)similarity, with the two <em>most-similar</em> objects being
combined first (to create a new composite object), then the next two
most-similar objects (including the composites), and so on until a
single object is defined. The sequence of combinations is illustrated
using a “dendrogram” (cluster tree), and this can be examined to
determine the number of clusters. The function <code>cutree()</code>
determines the cluster membership of each observation, which can be
listed or plotted.</p>
<p>The first step in an agglomerative cluster analysis (although not
always expiclitly displayed), is the calculation of a dissimilarity
matrix, which is a square symmetrical matrix that illustrates the
dissimilarities (typically Euclidean distances) between individual cases
or observations, using in this case the six climate variables to define
the space that distance is calculated in. The <code>levelplot()</code>
funcction from the <code>{lattice}</code> package is used to make the
plot.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co"># get dissimilarities (distances) and cluster</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(orstationc[,<span class="dv">5</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>X_dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(<span class="fu">scale</span>(X))</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>grid <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">obs1 =</span> <span class="fu">seq</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">92</span>), <span class="at">obs2 =</span> <span class="fu">seq</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">92</span>))</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a><span class="fu">levelplot</span>(<span class="fu">as.matrix</span>(X_dist)<span class="sc">/</span><span class="fu">max</span>(X_dist) <span class="sc">~</span> obs1<span class="sc">*</span>obs2, <span class="at">data=</span>grid, </span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>    <span class="at">col.regions =</span> <span class="fu">gray</span>(<span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">by=</span><span class="fl">0.0625</span>)), <span class="at">xlab=</span><span class="st">&quot;observation&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;observation&quot;</span>)</span></code></pre></div>
<p><img src="multivariate_files/figure-html/multivariate-9-1.png" width="75%" height="75%" /></p>
<p>In the code above, for convenience, a matrix <strong>X</strong> was
created to omit the variables we don’t want to include in the clustering
(i.e. the non-climatic information). The levelplot() function plots the
dissimilarity matrix that the cluster analysis works with. The dark
diagonal line is formed from the Eucldean distance values of 0.0 from
the comparison of each observation with itself.</p>
<p>There is a range of different styles of patterns that may show up</p>
<ul>
<li>If the climates at the individual stations were completely unrelated
to on another, or if they were identical, the plot would appear light
gray (with black pixels along the diagonal).The “plaid” as opposed to
completely random appearance of the plot indicates that it is likely
that natural groups will exist in the data.<br />
</li>
<li>If stations near to one another had similar climates then one would
see a plaid pattern.</li>
</ul>
<p>The “plaidness” of the image could be enhanced by sorting the
observations in some way that makes sense. (Right now, they’re in
alphabetical order.) The plot can be redone, this time sorting in to
longitudinal order, West to East
(i.e. <code>orstationc[order(orstationc$lon), 5:10]</code>):</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co"># get dissimilarities (distances) and cluster</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(orstationc[<span class="fu">order</span>(orstationc<span class="sc">$</span>lon), <span class="dv">5</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>X_dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(<span class="fu">scale</span>(X))</span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>grid <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">obs1 =</span> <span class="fu">seq</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">92</span>), <span class="at">obs2 =</span> <span class="fu">seq</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">92</span>))</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="fu">levelplot</span>(<span class="fu">as.matrix</span>(X_dist)<span class="sc">/</span><span class="fu">max</span>(X_dist) <span class="sc">~</span> obs1<span class="sc">*</span>obs2, <span class="at">data=</span>grid, </span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>    <span class="at">col.regions =</span> <span class="fu">gray</span>(<span class="fu">seq</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="at">by=</span><span class="fl">0.0625</span>)), <span class="at">xlab=</span><span class="st">&quot;observation&quot;</span>, <span class="at">ylab=</span><span class="st">&quot;observation&quot;</span>)</span></code></pre></div>
<p><img src="multivariate_files/figure-html/multivariate-10-1.png" width="75%" height="75%" /></p>
<p>The “plaid and blocks” pattern suggests that, as we might have
expected, there is a longitudinal trend in Oregon climates.</p>
</div>
<div id="the-cluster-analysis" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> The cluster
analysis</h2>
<p>Get the orignal distance matrix again, and do the cluster
analysis:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co"># distance matrix</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(orstationc[,<span class="dv">5</span><span class="sc">:</span><span class="dv">10</span>])</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>X_dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(<span class="fu">scale</span>(X))</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a><span class="co"># cluster analysis</span></span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>X_clusters <span class="ot">&lt;-</span> <span class="fu">hclust</span>(X_dist, <span class="at">method =</span> <span class="st">&quot;ward.D2&quot;</span>)</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a><span class="fu">plot</span>(X_clusters, <span class="at">labels=</span>station, <span class="at">cex=</span><span class="fl">0.4</span>)</span></code></pre></div>
<p><img src="multivariate_files/figure-html/multivariate-11-1.png" width="75%" height="75%" /></p>
<p>The function <code>hclust()</code> performs a “Wards’s
single-linkage” analysis, and the <code>plot()</code> function plots the
dendrogram. Notice that the individual “leaves” of the plot are labeled
by the climate-station names.</p>
<p>The dendrogram can be inspected to get an idea of how many distinct
clusters there are (and it can also identify unusual points–look for
Crater Lake <code>CLK</code>). As a first guess, let’s say there were
four distinct clusters evident in the dendrogram. The
<code>cuttree()</code> function can be used to “cut” the dendrogram at a
particular level:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># cut dendrogram and plot points in each cluster </span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>nclust <span class="ot">&lt;-</span> <span class="dv">4</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a>clusternum <span class="ot">&lt;-</span> <span class="fu">cutree</span>(X_clusters, <span class="at">k=</span>nclust)</span></code></pre></div>
<p>And then the cluster membership (i.e. which cluster?) of each station
can be mapped.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="co"># plot cluster memberships of each station</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data =</span> or_otl_sf, <span class="at">fill =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="at">data =</span> orstationc, <span class="fu">aes</span>(<span class="at">x =</span>lon, <span class="at">y =</span> lat, <span class="at">label =</span> <span class="fu">as.character</span>(clusternum)), </span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a>            <span class="at">check_overlap =</span> <span class="cn">TRUE</span>, <span class="at">color =</span> <span class="st">&quot;red&quot;</span>, <span class="at">size =</span> <span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Oregon Climate Stations -- Cluster Numbers&quot;</span>, <span class="at">x =</span> <span class="st">&quot;Longitude&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Latitude&quot;</span>) <span class="sc">+</span></span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<p><img src="multivariate_files/figure-html/multivariate-13-1.png" width="75%" height="75%" /></p>
<p>Cluster 1 picks out the stations in eastern Oregon’s “High Desert”
region, while Cluster 2 highlights lower elevation stations in eastern
Oregon. Cluster 3 includes stations in the valleys of western Oregon,
while Cluster 4 indentifies coastal and lower Columbia Valley
stations.</p>
<p>(A listing of the cluster number that each station belongs to can be
gotten by the following code.)</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">cbind</span>(station, clusternum, <span class="fu">as.character</span>(Name)))</span></code></pre></div>
</div>
<div id="examining-the-cluster-analysis" class="section level2"
number="2.3">
<h2><span class="header-section-number">2.3</span> Examining the cluster
analysis</h2>
<p>The efficacy of the analyis in regionalizing or grouping the stations
can be gauged usigns plots and a analysis that looks at the “separation”
of the groups of stations.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="co"># examine clusters -- simple plots </span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="fu">tapply</span>(tjan, clusternum, mean)</span></code></pre></div>
<pre><code>##          1          2          3          4 
## -1.7166667 -0.7166667  4.0000000  5.3052632</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="fu">histogram</span>(<span class="sc">~</span> tjan <span class="sc">|</span> <span class="fu">factor</span>(clusternum), <span class="at">nint=</span><span class="dv">20</span>, <span class="at">layout=</span><span class="fu">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))</span></code></pre></div>
<p><img src="multivariate_files/figure-html/multivariate-15-1.png" width="75%" height="75%" /></p>
<p>Note how the histograms for some of the clusters seem multimodal
multimodal–this is a clear sign of inhomogeneity of the clusters. Let’s
see how the clusters differ. This can be done via discriminant analysis
<a
href="https://pjbartlein.github.io/REarthSysSci/topics/da.html">Discriminant
Analysis</a>, which (like PCA) defines some new variables with desirable
properties, in this case, new variables that maximally “spread out” the
groups along the new variable axes:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="co"># discriminant analysis </span></span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a>cluster_lda1 <span class="ot">&lt;-</span> <span class="fu">lda</span>(clusternum <span class="sc">~</span> tjan <span class="sc">+</span> tjul <span class="sc">+</span> tann <span class="sc">+</span> pjan <span class="sc">+</span> pjul <span class="sc">+</span> pann)</span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a>cluster_lda1</span></code></pre></div>
<pre><code>## Call:
## lda(clusternum ~ tjan + tjul + tann + pjan + pjul + pann)
## 
## Prior probabilities of groups:
##         1         2         3         4 
## 0.3913043 0.1304348 0.2717391 0.2065217 
## 
## Group means:
##         tjan     tjul      tann      pjan      pjul      pann
## 1 -1.7166667 18.75556  8.038889  52.75000  9.944444  381.4722
## 2 -0.7166667 23.89167 11.225000  37.91667  6.166667  268.6667
## 3  4.0000000 19.36000 11.312000 186.64000  7.880000 1080.1600
## 4  5.3052632 15.95263 10.657895 316.52632 19.368421 1891.8947
## 
## Coefficients of linear discriminants:
##               LD1           LD2          LD3
## tjan -0.078899639 -0.2453913360 -1.615938671
## tjul  0.161194061 -0.7350202197 -1.079992203
## tann -0.588385785  0.2854664305  2.730086421
## pjan -0.012082646 -0.0113134200 -0.004373537
## pjul -0.015570218  0.0896595430  0.081419236
## pann -0.001246184  0.0007128788  0.001468877
## 
## Proportion of trace:
##    LD1    LD2    LD3 
## 0.8257 0.1425 0.0318</code></pre>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="fu">plot</span>(cluster_lda1, <span class="at">col =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="multivariate_files/figure-html/multivariate-16-1.png" width="75%" height="75%" /></p>
<p>Get the discriminant scores (the values of the new discriminant
function):</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="co"># discriminant scores</span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>cluster_dscore <span class="ot">&lt;-</span> <span class="fu">predict</span>(cluster_lda1, <span class="at">dimen=</span>nclust)<span class="sc">$</span>x</span></code></pre></div>
<p>Produce a grouped box plot for each of first three discriminant
function:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a><span class="co"># residual grouped box plot</span></span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>opar <span class="ot">&lt;-</span> <span class="fu">par</span>(<span class="at">mfrow=</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a><span class="fu">boxplot</span>(cluster_dscore[, <span class="dv">1</span>] <span class="sc">~</span> clusternum, <span class="at">ylab =</span> <span class="st">&quot;LD 1&quot;</span>, <span class="at">ylim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">8</span>,<span class="dv">8</span>))</span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a><span class="fu">boxplot</span>(cluster_dscore[, <span class="dv">2</span>] <span class="sc">~</span> clusternum, <span class="at">ylab =</span> <span class="st">&quot;LD 2&quot;</span>, <span class="at">ylim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">8</span>,<span class="dv">8</span>))</span>
<span id="cb24-5"><a href="#cb24-5" tabindex="-1"></a><span class="fu">boxplot</span>(cluster_dscore[, <span class="dv">3</span>] <span class="sc">~</span> clusternum, <span class="at">ylab =</span> <span class="st">&quot;LD 2&quot;</span>, <span class="at">ylim=</span><span class="fu">c</span>(<span class="sc">-</span><span class="dv">8</span>,<span class="dv">8</span>))</span></code></pre></div>
<p><img src="multivariate_files/figure-html/multivariate-18-1.png" width="75%" height="75%" /></p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="fu">par</span>(opar)</span></code></pre></div>
<p>The first discriminant function clearly spreads out the four
clusters, while the second could be used to discriminate between
clusters 1 and 4 and clusters 2 and 3.</p>
<p>Finally, get the discriminant function loadings, which measure the
relationship or correlation between each original variables and the new
discriminant function:</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="co"># discriminant &quot;loadings&quot;</span></span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a><span class="fu">cor</span>(orstationc[<span class="dv">5</span><span class="sc">:</span><span class="dv">10</span>],cluster_dscore)</span></code></pre></div>
<pre><code>##             LD1        LD2          LD3
## tjan -0.9196005 -0.1501351 -0.235122328
## tjul  0.4810736 -0.8198778  0.260104905
## tann -0.5795323 -0.7305824  0.063482600
## pjan -0.9632235  0.1729949  0.009078909
## pjul -0.4913887  0.5687402  0.492687963
## pann -0.9540390  0.2165052  0.059543444</code></pre>
<p>The loadings can be used to interpret or describe the discriminant
functions. Here January temperature and January and annual precipitation
are most highly correlated with the first discriminant function. Each of
these have strong latitudinal gradient across the state: high in the
west, low in the east. Note that this gradient corresponds well with a
similar longitudinal gradient in cluster number.</p>
<p>The discriminant analysis here is used to answer the question “how do
the clusters of stations differ climatically?” In this case, it looks
like <code>pann</code> and <code>tann</code> are the variables most
closely correlated with each discriminant function, and because each of
these variables are more-or-less averages of the seasonal extreme
variables, that might explain why the clusters seem inhomogeneous.</p>
<p>The analysis can be repeated, extracting different numbers of
clusters. The easiest way to do this is to copy the three chunks of code
above into a text editor, changing the assignment to nclust and
including or excluding before executing each chunk, one at a time.</p>
</div>
</div>
<div id="distances-among-observations-and-among-variables"
class="section level1" number="3">
<h1><span class="header-section-number">3</span> Distances: among
observations and among variables</h1>
<p>As noted above, in addition to being the input to a cluster analysis,
distance matrices are interesting in their own right. In a typical
rectangular data set, there are two distance matrices that could be
calculated, one comparing variables (i.e. the columns in a data frame),
and the other comparing observations (or rows). In the examples below
the “TraCE 21ka” decadal data are used. There are (96 x 48) 4608 grid
points and 2204 observations. Read the raster brick in the netCDF file
(<code>Trace21_TREFHT_anm2.nc</code>):</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="co"># read the TraCE21 temperature data</span></span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a><span class="co"># read TraCE21-ka transient climate-model simulation decadal data</span></span>
<span id="cb28-3"><a href="#cb28-3" tabindex="-1"></a>tr21dec_path <span class="ot">&lt;-</span> <span class="st">&quot;/Users/bartlein/Projects/RESS/data/nc_files/&quot;</span></span>
<span id="cb28-4"><a href="#cb28-4" tabindex="-1"></a>tr21dec_name <span class="ot">&lt;-</span> <span class="st">&quot;Trace21_TREFHT_anm2.nc&quot;</span></span>
<span id="cb28-5"><a href="#cb28-5" tabindex="-1"></a>tr21dec_file <span class="ot">&lt;-</span> <span class="fu">paste</span>(tr21dec_path, tr21dec_name, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb28-6"><a href="#cb28-6" tabindex="-1"></a>dname <span class="ot">&lt;-</span> <span class="st">&quot;tas_anm_ann&quot;</span></span></code></pre></div>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="co"># read the data using ncdf4</span></span>
<span id="cb29-2"><a href="#cb29-2" tabindex="-1"></a>nc_in <span class="ot">&lt;-</span> <span class="fu">nc_open</span>(tr21dec_file)</span>
<span id="cb29-3"><a href="#cb29-3" tabindex="-1"></a><span class="fu">print</span>(nc_in)</span></code></pre></div>
<pre><code>## File /Users/bartlein/Projects/RESS/data/nc_files/Trace21_TREFHT_anm2.nc (NC_FORMAT_CLASSIC):
## 
##      5 variables (excluding dimension variables):
##         float tas_anm_ann[lon,lat,time]   
##             name: tas_anm_ann
##             long_name: tas anomalies
##             units: C
##             _FillValue: 1.00000003318135e+32
##             comment: K transformed to degrees C
##         float tas_anm_djf[lon,lat,time]   
##             name: tas_anm_djf
##             long_name: tas anomalies
##             units: C
##             _FillValue: 1.00000003318135e+32
##             comment: K transformed to degrees C
##         float tas_anm_mam[lon,lat,time]   
##             name: tas_anm_mam
##             long_name: tas anomalies
##             units: C
##             _FillValue: 1.00000003318135e+32
##             comment: K transformed to degrees C
##         float tas_anm_jja[lon,lat,time]   
##             name: tas_anm_jja
##             long_name: tas anomalies
##             units: C
##             _FillValue: 1.00000003318135e+32
##             comment: K transformed to degrees C
##         float tas_anm_son[lon,lat,time]   
##             name: tas_anm_son
##             long_name: tas anomalies
##             units: C
##             _FillValue: 1.00000003318135e+32
##             comment: K transformed to degrees C
## 
##      3 dimensions:
##         lon  Size:96 
##             name: lon
##             long_name: Longitude
##             units: degrees_east
##             axis: X
##             standard_name: longitude
##             coodinate_defines: gridcell_center
##         lat  Size:48 
##             name: lat
##             long_name: Latitude
##             units: degrees_north
##             axis: Y
##             standard_name: latitude
##             coodinate_defines: gridcell_center
##         time  Size:2204 
##             standard_name: time
##             long_name: time
##             units: ka BP
##             axis: T
##             calendar: noleap
##             CoordinateAxisType: time
## 
##     8 global attributes:
##         title: TraCE 21 Transient Simulation
##         model: CCSM3
##         experiment: Trace 21
##         references: Liu et al. (2009), Science 325:310-314
##         history: P.J. Bartlein, 21 Aug 2014
##         source_file: b30.00_4kaDVTd.cam2.ncrcat
##         Conventions: CF-1.0
##         comment: anomaly ltm base period:  1820, 1830, 1840</code></pre>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a>lon <span class="ot">&lt;-</span> <span class="fu">ncvar_get</span>(nc_in,<span class="st">&quot;lon&quot;</span>)</span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a>lat <span class="ot">&lt;-</span> <span class="fu">ncvar_get</span>(nc_in,<span class="st">&quot;lat&quot;</span>)</span>
<span id="cb31-3"><a href="#cb31-3" tabindex="-1"></a>t <span class="ot">&lt;-</span> <span class="fu">ncvar_get</span>(nc_in, <span class="st">&quot;time&quot;</span>)</span>
<span id="cb31-4"><a href="#cb31-4" tabindex="-1"></a>var_array <span class="ot">&lt;-</span> <span class="fu">ncvar_get</span>(nc_in, dname)</span>
<span id="cb31-5"><a href="#cb31-5" tabindex="-1"></a><span class="fu">dim</span>(var_array)</span></code></pre></div>
<pre><code>## [1]   96   48 2204</code></pre>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a>plt_xvals <span class="ot">&lt;-</span> t</span></code></pre></div>
<p>The object <code>plt_xvals</code> is used to easily access time
labels for some of the later plots.</p>
<p>Reshape the 3-d array into a 2-d “tidy” matrix with individual
variables (grid-point locations) in columns, and observations (time) in
rows:</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a><span class="co"># reshape the 3-d array</span></span>
<span id="cb34-2"><a href="#cb34-2" tabindex="-1"></a>nlon <span class="ot">&lt;-</span> <span class="fu">dim</span>(var_array)[<span class="dv">1</span>]; nlat <span class="ot">&lt;-</span> <span class="fu">dim</span>(var_array)[<span class="dv">2</span>]; nt <span class="ot">&lt;-</span> <span class="fu">dim</span>(var_array)[<span class="dv">3</span>]</span>
<span id="cb34-3"><a href="#cb34-3" tabindex="-1"></a>var_vec_long <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(var_array)</span>
<span id="cb34-4"><a href="#cb34-4" tabindex="-1"></a><span class="fu">length</span>(var_vec_long)</span>
<span id="cb34-5"><a href="#cb34-5" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">matrix</span>(var_vec_long, <span class="at">nrow =</span> nlon <span class="sc">*</span> nlat, <span class="at">ncol =</span> nt))</span>
<span id="cb34-6"><a href="#cb34-6" tabindex="-1"></a><span class="fu">dim</span>(X)</span></code></pre></div>
<div id="distances-among-times" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Distances among
times</h2>
<p>Notice the use of the transposition function (<code>t()</code>) to
get the data in the proper “tidy” arrangement for a data frame
(variables in columns, observations in rows). Distance matrix
calculation can take a while, but the <code>{parallelDist}</code>
package can be used to greatly speed things up, by handing off the
creation of different blocks of the matrix to different processors in a
multi-processor machine.</p>
<p>The <code>parDist()</code> function gets the dissimilarity
matrix:</p>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" tabindex="-1"></a><span class="co"># disatances among observations (time)</span></span>
<span id="cb35-2"><a href="#cb35-2" tabindex="-1"></a>dist_obs <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">parDist</span>(X)) </span>
<span id="cb35-3"><a href="#cb35-3" tabindex="-1"></a><span class="fu">dim</span>(dist_obs)</span></code></pre></div>
<p>The dissimilarity matrix (<code>dist_obs</code>) can be conveniently
visualized by turning it into a <code>{terra}</code> SpatRaster, and
then using the <code>{tidyterra}</code> <code>geom_spatraster()</code>
function to visualize it. The rendering of the image is faster if the
image is written to a file:</p>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="co"># turn matrix into a SpatRaster</span></span>
<span id="cb36-2"><a href="#cb36-2" tabindex="-1"></a>dist_var_terra <span class="ot">&lt;-</span> <span class="fu">flip</span>(<span class="fu">rast</span>(dist_var))</span>
<span id="cb36-3"><a href="#cb36-3" tabindex="-1"></a>dist_var_terra</span></code></pre></div>
<p>Use <code>ggplot2()</code> to plot the matrix (writing the *.png
image to a file):</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" tabindex="-1"></a><span class="co"># plot the dissimilarity matrix</span></span>
<span id="cb37-2"><a href="#cb37-2" tabindex="-1"></a><span class="fu">png</span>(<span class="st">&quot;TraCE_decadal_dist_obs_v1.png&quot;</span>, <span class="at">width=</span><span class="dv">720</span>, <span class="at">height=</span><span class="dv">720</span>) <span class="co"># open the file</span></span>
<span id="cb37-3"><a href="#cb37-3" tabindex="-1"></a>  <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb37-4"><a href="#cb37-4" tabindex="-1"></a>    <span class="fu">geom_spatraster</span>(<span class="at">data =</span> <span class="fu">t</span>(dist_obs_terra)) <span class="sc">+</span></span>
<span id="cb37-5"><a href="#cb37-5" tabindex="-1"></a>      <span class="fu">scale_fill_distiller</span>(<span class="at">type =</span> <span class="st">&quot;seq&quot;</span>, <span class="at">palette =</span> <span class="st">&quot;Greens&quot;</span>) <span class="sc">+</span></span>
<span id="cb37-6"><a href="#cb37-6" tabindex="-1"></a>      <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">2200</span>, <span class="at">by=</span><span class="dv">500</span>), <span class="at">minor_breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">2200</span>, <span class="at">by=</span><span class="dv">100</span>)) <span class="sc">+</span></span>
<span id="cb37-7"><a href="#cb37-7" tabindex="-1"></a>      <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">2200</span>, <span class="at">by=</span><span class="dv">500</span>), <span class="at">minor_breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">2200</span>, <span class="at">by=</span><span class="dv">100</span>)) <span class="sc">+</span></span>
<span id="cb37-8"><a href="#cb37-8" tabindex="-1"></a>      <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Dissimilarity matrix among observations (times)&quot;</span>,</span>
<span id="cb37-9"><a href="#cb37-9" tabindex="-1"></a>        <span class="at">x =</span> <span class="st">&quot;Decades since 0 BP&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Decades since 0 BP&quot;</span>, <span class="at">fill=</span><span class="st">&quot;distance&quot;</span>) <span class="sc">+</span></span>
<span id="cb37-10"><a href="#cb37-10" tabindex="-1"></a>      <span class="fu">theme</span>(<span class="at">panel.grid.major=</span><span class="fu">element_line</span>(<span class="at">colour=</span><span class="st">&quot;black&quot;</span>), <span class="at">panel.grid.minor=</span><span class="fu">element_line</span>(<span class="at">colour=</span><span class="st">&quot;black&quot;</span>),</span>
<span id="cb37-11"><a href="#cb37-11" tabindex="-1"></a>            <span class="at">axis.text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">12</span>), <span class="at">axis.title=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>), <span class="at">aspect.ratio =</span> <span class="dv">1</span>,</span>
<span id="cb37-12"><a href="#cb37-12" tabindex="-1"></a>            <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">20</span>))</span>
<span id="cb37-13"><a href="#cb37-13" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p>Here’s the image:</p>
<p><img src="images/TraCE_decadal_dist_obs_v1.png" /></p>
<p>The observations here are ordered in time, and the blocks in the
distance matrix indicate that there are two distinct climate regimes:
one from the beginning of the record to around 14.7 ka (14,700 years
before 1950 CE, e.g. in decade 1470 before the end of the record at 0
BP). Within each major interval, there are subintervals indicated by the
dark squares. For example during the Holocene (the past 11,700 years,
there are two blocks, the early Holocene, 11,700 to 8200 yr BP, and the
middle and late Holocene, after 8200 yr BP). There are also patterns
within the blocks. During the interval from 14,700 to 11,700 yr BP, the
warm by variable Bølling-Allerød from 14,700 yr BP to 12,900 yr BP,
followed by the colder Younger Dryas interval, 12,900 to 11,700 yr BP
can be seen.</p>
</div>
<div id="distance-among-locations" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Distance among
locations</h2>
<p>The dissimilarities among locations/variables can be gotten in a
similar fashion, but note the use of the transpose function
<code>t()</code> to get the variables in rows, and observations in
columns:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="co"># variables (grid points)</span></span>
<span id="cb38-2"><a href="#cb38-2" tabindex="-1"></a>dist_var <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">parDist</span>(<span class="fu">t</span>(X)))</span>
<span id="cb38-3"><a href="#cb38-3" tabindex="-1"></a><span class="fu">dim</span>(dist_var)</span></code></pre></div>
<pre><code>## [1] 6 6</code></pre>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a><span class="co"># turn matrix into a SpatRaster</span></span>
<span id="cb40-2"><a href="#cb40-2" tabindex="-1"></a>dist_var_terra <span class="ot">&lt;-</span> <span class="fu">flip</span>(<span class="fu">rast</span>(dist_var))</span>
<span id="cb40-3"><a href="#cb40-3" tabindex="-1"></a>dist_var_terra</span></code></pre></div>
<pre><code>## class       : SpatRaster 
## dimensions  : 6, 6, 1  (nrow, ncol, nlyr)
## resolution  : 1, 1  (x, y)
## extent      : 0, 6, 0, 6  (xmin, xmax, ymin, ymax)
## coord. ref. :  
## source(s)   : memory
## name        :    lyr.1 
## min value   :     0.00 
## max value   : 10405.97</code></pre>
<p>Plot the matrix:</p>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" tabindex="-1"></a><span class="co"># turn matrix into a SpatRaster</span></span>
<span id="cb42-2"><a href="#cb42-2" tabindex="-1"></a>dist_var_terra <span class="ot">&lt;-</span> <span class="fu">flip</span>(<span class="fu">rast</span>(dist_var))</span>
<span id="cb42-3"><a href="#cb42-3" tabindex="-1"></a>dist_var_terra</span></code></pre></div>
<pre><code>## class       : SpatRaster 
## dimensions  : 6, 6, 1  (nrow, ncol, nlyr)
## resolution  : 1, 1  (x, y)
## extent      : 0, 6, 0, 6  (xmin, xmax, ymin, ymax)
## coord. ref. :  
## source(s)   : memory
## name        :    lyr.1 
## min value   :     0.00 
## max value   : 10405.97</code></pre>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="co"># plot the dissimilarity matrix</span></span>
<span id="cb44-2"><a href="#cb44-2" tabindex="-1"></a><span class="fu">png</span>(<span class="st">&quot;TraCE_decadal_dist_var_v1.png&quot;</span>, <span class="at">width=</span><span class="dv">720</span>, <span class="at">height=</span><span class="dv">720</span>) <span class="co"># open the file</span></span>
<span id="cb44-3"><a href="#cb44-3" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb44-4"><a href="#cb44-4" tabindex="-1"></a>  <span class="fu">geom_spatraster</span>(<span class="at">data =</span> <span class="fu">t</span>(dist_var_terra)) <span class="sc">+</span></span>
<span id="cb44-5"><a href="#cb44-5" tabindex="-1"></a>  <span class="fu">scale_fill_distiller</span>(<span class="at">type =</span> <span class="st">&quot;seq&quot;</span>, <span class="at">palette =</span> <span class="st">&quot;Greens&quot;</span>) <span class="sc">+</span></span>
<span id="cb44-6"><a href="#cb44-6" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">4600</span>, <span class="at">by=</span><span class="dv">500</span>), <span class="at">minor_breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">4600</span>, <span class="at">by=</span><span class="dv">250</span>)) <span class="sc">+</span></span>
<span id="cb44-7"><a href="#cb44-7" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">4600</span>, <span class="at">by=</span><span class="dv">500</span>), <span class="at">minor_breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">4600</span>, <span class="at">by=</span><span class="dv">250</span>)) <span class="sc">+</span></span>
<span id="cb44-8"><a href="#cb44-8" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">&quot;Dissimilarity matrix among variables (grid points/locations)&quot;</span>,</span>
<span id="cb44-9"><a href="#cb44-9" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">&quot;Location&quot;</span>, <span class="at">y =</span> <span class="st">&quot;Location&quot;</span>, <span class="at">fill=</span><span class="st">&quot;distance&quot;</span>) <span class="sc">+</span></span>
<span id="cb44-10"><a href="#cb44-10" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">panel.grid.major=</span><span class="fu">element_line</span>(<span class="at">colour=</span><span class="st">&quot;black&quot;</span>), <span class="at">panel.grid.minor=</span><span class="fu">element_line</span>(<span class="at">colour=</span><span class="st">&quot;black&quot;</span>),</span>
<span id="cb44-11"><a href="#cb44-11" tabindex="-1"></a>            <span class="at">axis.text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">12</span>), <span class="at">axis.title=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">14</span>), <span class="at">aspect.ratio =</span> <span class="dv">1</span>,</span>
<span id="cb44-12"><a href="#cb44-12" tabindex="-1"></a>            <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">20</span>))</span>
<span id="cb44-13"><a href="#cb44-13" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
<p>The image is sort of a map: The grid points are arranged in the data
frame with longitudes varying rapidly (forming the small squares), and
the latitudes slowly, such that the square dark area (low
dissimilarities) in the middle is the tropics, where the spatial
variability of climate is low, while the more spatially variable
higher-latitude regions (at all longitudes) frame the low-variability
region.</p>
<p><img src="images/TraCE_decadal_dist_var_v1.png" /></p>
<p>Here the temporal variability of is emphasized.</p>
</div>
</div>
<div id="heatmaps-2-way-time-and-space-clustering"
class="section level1" number="4">
<h1><span class="header-section-number">4</span> Heatmaps (2-way,
time-and-space clustering)</h1>
<p>“Heatmaps” are a visualization approach that is implemented in the
<code>base</code> package of R. A heatmap basically consists of the
raster-image type plot of the data frame, with the rows and columns of
the matrix reaggranged using cluster analyses of the varables and
observations. (Note that the cluster analyses are carried out
individually, as opposed to simultaneously, an approach known as
“biclustering”.) Typically, for small data sets, the matrix plot is
augmented by dendrograms, and by “side bars” which plot some additional
attribute of the data (location or time, for example) using a scatter
plot, gradient plot, etc.</p>
<div id="a-simple-example-with-hovmöller-matrix-data"
class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> A simple example with
Hovmöller-matrix data</h2>
<p>A simple heatmap can be generated for a “Hovmöller matrix” calculated
using the gridded data. The Hovmöller matrix contains longitudinal
averages and is laid out with one row for each latitude and one column
for each time.</p>
<div id="set-up" class="section level3" number="4.1.1">
<h3><span class="header-section-number">4.1.1</span> Set up</h3>
<p>Create the Hovmöller matrix:</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" tabindex="-1"></a><span class="co"># Hovmöller matrix</span></span>
<span id="cb46-2"><a href="#cb46-2" tabindex="-1"></a>Y <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fl">0.0</span>, <span class="at">nrow=</span>nlat, <span class="at">ncol=</span>nt)</span>
<span id="cb46-3"><a href="#cb46-3" tabindex="-1"></a><span class="fu">dim</span>(X)</span>
<span id="cb46-4"><a href="#cb46-4" tabindex="-1"></a><span class="cf">for</span> (n <span class="cf">in</span> (<span class="dv">1</span><span class="sc">:</span>nt)) {</span>
<span id="cb46-5"><a href="#cb46-5" tabindex="-1"></a>  <span class="cf">for</span> (k <span class="cf">in</span> (<span class="dv">1</span><span class="sc">:</span>nlat)) {</span>
<span id="cb46-6"><a href="#cb46-6" tabindex="-1"></a>    <span class="cf">for</span> (j <span class="cf">in</span> (<span class="dv">1</span><span class="sc">:</span>nlon)) {</span>
<span id="cb46-7"><a href="#cb46-7" tabindex="-1"></a>      Y[k, n] <span class="ot">&lt;-</span> Y[k, n] <span class="sc">+</span> var_array[j, k, n]</span>
<span id="cb46-8"><a href="#cb46-8" tabindex="-1"></a>    }</span>
<span id="cb46-9"><a href="#cb46-9" tabindex="-1"></a>    Y[k, n] <span class="ot">&lt;-</span> Y[k, n]<span class="sc">/</span>nlon</span>
<span id="cb46-10"><a href="#cb46-10" tabindex="-1"></a>  }</span>
<span id="cb46-11"><a href="#cb46-11" tabindex="-1"></a>}</span>
<span id="cb46-12"><a href="#cb46-12" tabindex="-1"></a><span class="fu">dim</span>(Y)</span>
<span id="cb46-13"><a href="#cb46-13" tabindex="-1"></a>ncols <span class="ot">&lt;-</span> <span class="fu">dim</span>(Y)[<span class="dv">2</span>]</span></code></pre></div>
<p>Set row and column names for the matrix.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a><span class="co"># set row and column names</span></span>
<span id="cb47-2"><a href="#cb47-2" tabindex="-1"></a><span class="fu">rownames</span>(Y) <span class="ot">&lt;-</span> <span class="fu">str_pad</span>(<span class="fu">sprintf</span>(<span class="st">&quot;%.3f&quot;</span>, lat), <span class="dv">5</span>, <span class="at">pad=</span><span class="st">&quot;0&quot;</span>)</span>
<span id="cb47-3"><a href="#cb47-3" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rownames</span>(Y));<span class="fu">tail</span>(<span class="fu">rownames</span>(Y)); <span class="fu">length</span>(<span class="fu">rownames</span>(Y))</span>
<span id="cb47-4"><a href="#cb47-4" tabindex="-1"></a><span class="fu">colnames</span>(Y) <span class="ot">&lt;-</span> <span class="fu">str_pad</span>(<span class="fu">sprintf</span>(<span class="st">&quot;%.3f&quot;</span>, plt_xvals), <span class="dv">5</span>, <span class="at">pad=</span><span class="st">&quot;0&quot;</span>)</span>
<span id="cb47-5"><a href="#cb47-5" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colnames</span>(Y));<span class="fu">tail</span>(<span class="fu">colnames</span>(Y)); <span class="fu">length</span>(<span class="fu">colnames</span>(Y))</span></code></pre></div>
<p>A matrix plot of the Hovmöller matrix will illustrate the data (and
is an actual Hovmöller plot). To get high (northern) latitudes at the
top of the plot, and high (southern) latitudes at the bottom, flip the
data:</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" tabindex="-1"></a><span class="co"># flip X</span></span>
<span id="cb48-2"><a href="#cb48-2" tabindex="-1"></a>Y <span class="ot">&lt;-</span> Y[(nlat<span class="sc">:</span><span class="dv">1</span>),]</span>
<span id="cb48-3"><a href="#cb48-3" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rownames</span>(Y)); <span class="fu">tail</span>(<span class="fu">rownames</span>(Y)); <span class="fu">length</span>(<span class="fu">rownames</span>(Y))</span></code></pre></div>
<p>For this example, the data are subsampled, by taking every 20th
observation (to allow the dendrograms to be visualized):</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" tabindex="-1"></a><span class="co"># subset</span></span>
<span id="cb49-2"><a href="#cb49-2" tabindex="-1"></a>Y2 <span class="ot">&lt;-</span> Y[,<span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">2201</span>, <span class="at">by=</span><span class="dv">20</span>)]</span>
<span id="cb49-3"><a href="#cb49-3" tabindex="-1"></a><span class="fu">dim</span>(Y2)</span>
<span id="cb49-4"><a href="#cb49-4" tabindex="-1"></a>ncols2 <span class="ot">&lt;-</span> <span class="fu">dim</span>(Y2)[<span class="dv">2</span>]</span>
<span id="cb49-5"><a href="#cb49-5" tabindex="-1"></a><span class="fu">rownames</span>(Y2)</span></code></pre></div>
<p>More set up, Generate colors for the matrix pot and the side bars,
which here will illustrate the latitude and time of each data point.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" tabindex="-1"></a><span class="co"># generate colors for side bars</span></span>
<span id="cb50-2"><a href="#cb50-2" tabindex="-1"></a>rcol <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>,<span class="st">&quot;PRGn&quot;</span>))(nlat)</span>
<span id="cb50-3"><a href="#cb50-3" tabindex="-1"></a>ccol <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">9</span>,<span class="st">&quot;Purples&quot;</span>))(ncols)</span>
<span id="cb50-4"><a href="#cb50-4" tabindex="-1"></a>ccol2 <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">9</span>,<span class="st">&quot;Purples&quot;</span>))(ncols2)</span></code></pre></div>
<p>Generate colors for the heatmap</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" tabindex="-1"></a>nclr <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb51-2"><a href="#cb51-2" tabindex="-1"></a>zcol <span class="ot">&lt;-</span> (<span class="fu">rev</span>(<span class="fu">brewer.pal</span>(nclr,<span class="st">&quot;RdBu&quot;</span>)))</span>
<span id="cb51-3"><a href="#cb51-3" tabindex="-1"></a>breaks <span class="ot">=</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">20</span>,<span class="sc">-</span><span class="dv">10</span>,<span class="sc">-</span><span class="dv">5</span>,<span class="sc">-</span><span class="dv">2</span>,<span class="sc">-</span><span class="dv">1</span>,<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">5</span>,<span class="dv">10</span>,<span class="dv">20</span>)</span>
<span id="cb51-4"><a href="#cb51-4" tabindex="-1"></a><span class="fu">length</span>(breaks)</span></code></pre></div>
</div>
<div id="hovmöller-matrix-plot-subsetted-decadal-data"
class="section level3" number="4.1.2">
<h3><span class="header-section-number">4.1.2</span> Hovmöller-matrix
plot (subsetted decadal data)</h3>
<p>Produce the heatmap for the Hovmöller matrix:</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" tabindex="-1"></a>file_label <span class="ot">&lt;-</span>  <span class="st">&quot;TraCE_decadal&quot;</span></span>
<span id="cb52-2"><a href="#cb52-2" tabindex="-1"></a>png_file <span class="ot">&lt;-</span> <span class="fu">paste</span>(file_label, <span class="st">&quot;_hov_01&quot;</span>, <span class="st">&quot;.png&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb52-3"><a href="#cb52-3" tabindex="-1"></a><span class="fu">png</span>(png_file, <span class="at">width=</span><span class="dv">960</span>, <span class="at">height=</span><span class="dv">960</span>)</span>
<span id="cb52-4"><a href="#cb52-4" tabindex="-1"></a><span class="fu">system.time</span>( X_heatmap <span class="ot">&lt;-</span> <span class="fu">heatmap.2</span>(Y2, <span class="at">cexRow =</span> <span class="fl">0.8</span>, <span class="at">cexCol =</span> <span class="fl">0.8</span>, <span class="at">scale=</span><span class="st">&quot;none&quot;</span>,</span>
<span id="cb52-5"><a href="#cb52-5" tabindex="-1"></a>        <span class="at">Rowv=</span><span class="cn">NA</span>, <span class="at">Colv=</span><span class="cn">NA</span>, <span class="at">dendrogram =</span> <span class="st">&quot;none&quot;</span>, <span class="co"># just plot data, don&#39;t do clustering</span></span>
<span id="cb52-6"><a href="#cb52-6" tabindex="-1"></a>        <span class="at">RowSideColors=</span>rcol, <span class="at">ColSideColors=</span>ccol2, <span class="at">col=</span>zcol, <span class="at">breaks=</span>breaks, <span class="at">trace=</span><span class="st">&quot;none&quot;</span>, <span class="at">density.info =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb52-7"><a href="#cb52-7" tabindex="-1"></a>        <span class="at">lmat=</span><span class="fu">rbind</span>(<span class="fu">c</span>(<span class="dv">6</span>, <span class="dv">0</span>, <span class="dv">5</span>), <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">2</span>), <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">3</span>)), <span class="at">lwid=</span><span class="fu">c</span>(<span class="fl">1.0</span>, <span class="fl">0.2</span>, <span class="fl">5.0</span>), <span class="at">lhei =</span> <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="fl">0.2</span>, <span class="fl">4.0</span>),</span>
<span id="cb52-8"><a href="#cb52-8" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;ka&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Latitude&quot;</span>, <span class="at">key =</span> <span class="cn">TRUE</span>, <span class="at">key.title =</span> <span class="cn">NA</span>)</span>
<span id="cb52-9"><a href="#cb52-9" tabindex="-1"></a>)</span>
<span id="cb52-10"><a href="#cb52-10" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p><img src="images/TraCE_decadal_hov_01.png" /></p>
<p>Notice the way the side bars work in illustrating the location and
time of each data value.</p>
</div>
<div id="hovmöller-matrix-heatmap-subsetted-decadal-data"
class="section level3" number="4.1.3">
<h3><span class="header-section-number">4.1.3</span> Hovmöller-matrix
heatmap (subsetted decadal data)</h3>
<p>Now here’s the heatmap:</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" tabindex="-1"></a>png_file <span class="ot">&lt;-</span> <span class="fu">paste</span>(file_label, <span class="st">&quot;_heatmap_01&quot;</span>, <span class="st">&quot;.png&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb53-2"><a href="#cb53-2" tabindex="-1"></a><span class="fu">png</span>(png_file, <span class="at">width=</span><span class="dv">960</span>, <span class="at">height=</span><span class="dv">960</span>)</span>
<span id="cb53-3"><a href="#cb53-3" tabindex="-1"></a><span class="fu">system.time</span>( X_heatmap <span class="ot">&lt;-</span> <span class="fu">heatmap.2</span>(Y2, <span class="at">cexRow =</span> <span class="fl">0.8</span>, <span class="at">cexCol =</span> <span class="fl">0.8</span>, <span class="at">scale=</span><span class="st">&quot;none&quot;</span>,</span>
<span id="cb53-4"><a href="#cb53-4" tabindex="-1"></a>        <span class="at">RowSideColors=</span>rcol, <span class="at">ColSideColors=</span>ccol2, <span class="at">col=</span>zcol, <span class="at">breaks=</span>breaks, <span class="at">trace=</span><span class="st">&quot;none&quot;</span>, <span class="at">density.info =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb53-5"><a href="#cb53-5" tabindex="-1"></a>        <span class="at">lmat=</span><span class="fu">rbind</span>(<span class="fu">c</span>(<span class="dv">6</span>, <span class="dv">0</span>, <span class="dv">5</span>), <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">2</span>), <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">3</span>)), <span class="at">lwid=</span><span class="fu">c</span>(<span class="fl">1.0</span>, <span class="fl">0.2</span>, <span class="fl">5.0</span>), <span class="at">lhei =</span> <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="fl">0.2</span>, <span class="fl">4.0</span>),</span>
<span id="cb53-6"><a href="#cb53-6" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;ka&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Latitude&quot;</span>, <span class="at">key =</span> <span class="cn">TRUE</span>, <span class="at">key.title =</span> <span class="cn">NA</span>)</span>
<span id="cb53-7"><a href="#cb53-7" tabindex="-1"></a>)</span>
<span id="cb53-8"><a href="#cb53-8" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p><img src="images/TraCE_decadal_heatmap_01.png" /></p>
<p>The heatmap has rearranged the data according to the cluster analyses
of the observations and variables, placing more similar data points next
to one another. Notice how the side bars allow one to figure out the
location and time of individual points. Here, and not surprisingly,
there appear to be four distinct time-space clusters (or “regimes”) in
the data): glacial vs. interglacial (Holocene) and high latitudes (in
both hemispheres) vs. low latitudes.</p>
</div>
</div>
<div id="a-second-example-with-decadal-data" class="section level2"
number="4.2">
<h2><span class="header-section-number">4.2</span> A second example with
decadal data</h2>
<p>A second example uses the decadal data directly.</p>
<div id="hovmöller-matrix-plot" class="section level3" number="4.2.1">
<h3><span class="header-section-number">4.2.1</span> Hovmöller-matrix
plot</h3>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" tabindex="-1"></a>png_file <span class="ot">&lt;-</span> <span class="fu">paste</span>(file_label, <span class="st">&quot;_hov_02&quot;</span>, <span class="st">&quot;.png&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb54-2"><a href="#cb54-2" tabindex="-1"></a><span class="fu">png</span>(png_file, <span class="at">width=</span><span class="dv">960</span>, <span class="at">height=</span><span class="dv">960</span>)</span>
<span id="cb54-3"><a href="#cb54-3" tabindex="-1"></a><span class="fu">system.time</span>( X_heatmap <span class="ot">&lt;-</span> <span class="fu">heatmap.2</span>(Y, <span class="at">cexRow =</span> <span class="fl">0.8</span>, <span class="at">cexCol =</span> <span class="fl">0.8</span>, <span class="at">scale=</span><span class="st">&quot;none&quot;</span>, </span>
<span id="cb54-4"><a href="#cb54-4" tabindex="-1"></a>        <span class="at">Rowv=</span><span class="cn">NA</span>, <span class="at">Colv=</span><span class="cn">NA</span>, <span class="at">dendrogram =</span> <span class="st">&quot;none&quot;</span>, <span class="co"># just plot data, don&#39;t do clustering</span></span>
<span id="cb54-5"><a href="#cb54-5" tabindex="-1"></a>        <span class="at">RowSideColors=</span>rcol, <span class="at">ColSideColors=</span>ccol, <span class="at">col=</span>zcol, <span class="at">breaks=</span>breaks, <span class="at">trace=</span><span class="st">&quot;none&quot;</span>, <span class="at">density.info =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb54-6"><a href="#cb54-6" tabindex="-1"></a>        <span class="at">lmat=</span><span class="fu">rbind</span>(<span class="fu">c</span>(<span class="dv">6</span>, <span class="dv">0</span>, <span class="dv">5</span>), <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">2</span>), <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">3</span>)), <span class="at">lwid=</span><span class="fu">c</span>(<span class="fl">1.0</span>, <span class="fl">0.2</span>, <span class="fl">5.0</span>), <span class="at">lhei =</span> <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="fl">0.2</span>, <span class="fl">4.0</span>),</span>
<span id="cb54-7"><a href="#cb54-7" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;ka&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Latitude&quot;</span>, <span class="at">key =</span> <span class="cn">TRUE</span>, <span class="at">key.title =</span> <span class="cn">NA</span>)</span>
<span id="cb54-8"><a href="#cb54-8" tabindex="-1"></a>)</span>
<span id="cb54-9"><a href="#cb54-9" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p><img src="images/TraCE_decadal_hov_02.png" /></p>
</div>
<div id="hovmöller-matrix-heatmap" class="section level3"
number="4.2.2">
<h3><span class="header-section-number">4.2.2</span> Hovmöller-matrix
heatmap</h3>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" tabindex="-1"></a>png_file <span class="ot">&lt;-</span> <span class="fu">paste</span>(file_label, <span class="st">&quot;_heatmap_02&quot;</span>, <span class="st">&quot;.png&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb55-2"><a href="#cb55-2" tabindex="-1"></a><span class="fu">png</span>(png_file, <span class="at">width=</span><span class="dv">960</span>, <span class="at">height=</span><span class="dv">960</span>)</span>
<span id="cb55-3"><a href="#cb55-3" tabindex="-1"></a><span class="fu">system.time</span>( X_heatmap <span class="ot">&lt;-</span> <span class="fu">heatmap.2</span>(Y, <span class="at">cexRow =</span> <span class="fl">0.8</span>, <span class="at">cexCol =</span> <span class="fl">0.8</span>, <span class="at">scale=</span><span class="st">&quot;none&quot;</span>, </span>
<span id="cb55-4"><a href="#cb55-4" tabindex="-1"></a>        <span class="at">RowSideColors=</span>rcol, <span class="at">ColSideColors=</span>ccol, <span class="at">col=</span>zcol, <span class="at">breaks=</span>breaks, <span class="at">trace=</span><span class="st">&quot;none&quot;</span>, <span class="at">density.info =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb55-5"><a href="#cb55-5" tabindex="-1"></a>        <span class="at">lmat=</span><span class="fu">rbind</span>(<span class="fu">c</span>(<span class="dv">6</span>, <span class="dv">0</span>, <span class="dv">5</span>), <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">2</span>), <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">3</span>)), <span class="at">lwid=</span><span class="fu">c</span>(<span class="fl">1.0</span>, <span class="fl">0.2</span>, <span class="fl">5.0</span>), <span class="at">lhei =</span> <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="fl">0.2</span>, <span class="fl">4.0</span>),</span>
<span id="cb55-6"><a href="#cb55-6" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;ka&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Latitude&quot;</span>, <span class="at">key =</span> <span class="cn">TRUE</span>, <span class="at">key.title =</span> <span class="cn">NA</span>)</span>
<span id="cb55-7"><a href="#cb55-7" tabindex="-1"></a>)</span>
<span id="cb55-8"><a href="#cb55-8" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<p><img src="images/TraCE_decadal_heatmap_02.png" /></p>
</div>
</div>
<div id="full-grid-example" class="section level2" number="4.3">
<h2><span class="header-section-number">4.3</span> Full-grid
example</h2>
<p>Finally, a heatmap of the gridded data (as opposed to the Hovmöller
matrix data) can be obtained</p>
<div id="set-up-1" class="section level3" number="4.3.1">
<h3><span class="header-section-number">4.3.1</span> Set up</h3>
<p>There are several set up steps necessary. Reshape the input array
again, and generate row and column names:</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" tabindex="-1"></a><span class="co"># reshape the 3-d array</span></span>
<span id="cb56-2"><a href="#cb56-2" tabindex="-1"></a>var_vec_long <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(var_array)</span>
<span id="cb56-3"><a href="#cb56-3" tabindex="-1"></a><span class="fu">length</span>(var_vec_long)</span>
<span id="cb56-4"><a href="#cb56-4" tabindex="-1"></a>X <span class="ot">&lt;-</span> <span class="fu">matrix</span>(var_vec_long, <span class="at">nrow =</span> nlon <span class="sc">*</span> nlat, <span class="at">ncol =</span> nt) <span class="co"># Don&#39;t transpose as usual this time</span></span>
<span id="cb56-5"><a href="#cb56-5" tabindex="-1"></a><span class="fu">dim</span>(X)</span>
<span id="cb56-6"><a href="#cb56-6" tabindex="-1"></a>ncols <span class="ot">&lt;-</span> <span class="fu">dim</span>(X)[<span class="dv">3</span>]</span></code></pre></div>
<div class="sourceCode" id="cb57"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" tabindex="-1"></a><span class="co"># generate row and column names</span></span>
<span id="cb57-2"><a href="#cb57-2" tabindex="-1"></a>grid <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(lon, lat)</span>
<span id="cb57-3"><a href="#cb57-3" tabindex="-1"></a><span class="fu">names</span>(grid) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;lon&quot;</span>, <span class="st">&quot;lat&quot;</span>)</span>
<span id="cb57-4"><a href="#cb57-4" tabindex="-1"></a><span class="fu">rownames</span>(X) <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;E&quot;</span>, <span class="fu">str_pad</span>(<span class="fu">sprintf</span>(<span class="st">&quot;%.2f&quot;</span>, <span class="fu">round</span>(grid<span class="sc">$</span>lon, <span class="dv">2</span>)), <span class="dv">5</span>, <span class="at">pad=</span><span class="st">&quot;0&quot;</span>),</span>
<span id="cb57-5"><a href="#cb57-5" tabindex="-1"></a>  <span class="st">&quot;N&quot;</span>, <span class="fu">str_pad</span>(<span class="fu">sprintf</span>(<span class="st">&quot;%.2f&quot;</span>, <span class="fu">round</span>(grid<span class="sc">$</span>lat, <span class="dv">2</span>)), <span class="dv">5</span>, <span class="at">pad=</span><span class="st">&quot;0&quot;</span>), <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb57-6"><a href="#cb57-6" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rownames</span>(X), <span class="dv">20</span>); <span class="fu">tail</span>(<span class="fu">rownames</span>(X), <span class="dv">20</span>); <span class="fu">length</span>(<span class="fu">rownames</span>(X))</span>
<span id="cb57-7"><a href="#cb57-7" tabindex="-1"></a></span>
<span id="cb57-8"><a href="#cb57-8" tabindex="-1"></a><span class="fu">colnames</span>(X) <span class="ot">&lt;-</span> <span class="fu">str_pad</span>(<span class="fu">sprintf</span>(<span class="st">&quot;%.3f&quot;</span>, t), <span class="dv">5</span>, <span class="at">pad=</span><span class="st">&quot;0&quot;</span>)</span>
<span id="cb57-9"><a href="#cb57-9" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">colnames</span>(X)); <span class="fu">tail</span>(<span class="fu">colnames</span>(X)); <span class="fu">length</span>(<span class="fu">colnames</span>(X))</span></code></pre></div>
<p>Flip the data…</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" tabindex="-1"></a><span class="co"># flip X</span></span>
<span id="cb58-2"><a href="#cb58-2" tabindex="-1"></a>X <span class="ot">&lt;-</span> X[((nlon<span class="sc">*</span>nlat)<span class="sc">:</span><span class="dv">1</span>),]</span>
<span id="cb58-3"><a href="#cb58-3" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rownames</span>(X)); <span class="fu">tail</span>(<span class="fu">rownames</span>(X)); <span class="fu">length</span>(<span class="fu">rownames</span>(X))</span></code></pre></div>
<p>… and generate colors:</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" tabindex="-1"></a><span class="co"># generate colors for vertical and horizontal side bars</span></span>
<span id="cb59-2"><a href="#cb59-2" tabindex="-1"></a>icol <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>,<span class="st">&quot;PRGn&quot;</span>))(nlat)</span>
<span id="cb59-3"><a href="#cb59-3" tabindex="-1"></a>idx <span class="ot">&lt;-</span> <span class="fu">sort</span>(<span class="fu">rep</span>((<span class="dv">1</span><span class="sc">:</span>nlat), nlon))</span>
<span id="cb59-4"><a href="#cb59-4" tabindex="-1"></a>rcol <span class="ot">&lt;-</span> icol[idx]</span>
<span id="cb59-5"><a href="#cb59-5" tabindex="-1"></a>ccol <span class="ot">&lt;-</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">9</span>,<span class="st">&quot;Purples&quot;</span>))(nt)</span></code></pre></div>
</div>
<div id="matrix-plot" class="section level3" number="4.3.2">
<h3><span class="header-section-number">4.3.2</span> Matrix plot</h3>
<p>Here’s a matrix plot of the data: Here, there are too many points to
make the dendrograms interpretable in any way, and so they are
suppressed.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" tabindex="-1"></a>png_file <span class="ot">&lt;-</span> <span class="fu">paste</span>(file_label, <span class="st">&quot;_hov_03&quot;</span>, <span class="st">&quot;.png&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb60-2"><a href="#cb60-2" tabindex="-1"></a><span class="fu">png</span>(png_file, <span class="at">width=</span><span class="dv">960</span>, <span class="at">height=</span><span class="dv">960</span>)</span>
<span id="cb60-3"><a href="#cb60-3" tabindex="-1"></a><span class="fu">system.time</span>( X_heatmap <span class="ot">&lt;-</span> <span class="fu">heatmap.2</span>(X, <span class="at">cexRow =</span> <span class="fl">0.8</span>, <span class="at">cexCol =</span> <span class="fl">0.8</span>, <span class="at">scale=</span><span class="st">&quot;none&quot;</span>,</span>
<span id="cb60-4"><a href="#cb60-4" tabindex="-1"></a>        <span class="at">Rowv=</span><span class="cn">NA</span>, <span class="at">Colv=</span><span class="cn">NA</span>, <span class="at">dendrogram =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb60-5"><a href="#cb60-5" tabindex="-1"></a>        <span class="at">RowSideColors=</span>rcol, <span class="at">ColSideColors=</span>ccol, <span class="at">col=</span>zcol, <span class="at">breaks=</span>breaks, <span class="at">trace=</span><span class="st">&quot;none&quot;</span>, <span class="at">density.info =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb60-6"><a href="#cb60-6" tabindex="-1"></a>        <span class="at">lmat=</span><span class="fu">rbind</span>(<span class="fu">c</span>(<span class="dv">6</span>, <span class="dv">0</span>, <span class="dv">5</span>), <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">2</span>), <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">3</span>)), <span class="at">lwid=</span><span class="fu">c</span>(<span class="fl">1.0</span>, <span class="fl">0.2</span>, <span class="fl">5.0</span>), <span class="at">lhei =</span> <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="fl">0.2</span>, <span class="fl">4.0</span>),</span>
<span id="cb60-7"><a href="#cb60-7" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;ka&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Locations&quot;</span>, <span class="at">key =</span> <span class="cn">TRUE</span>, <span class="at">key.title =</span> <span class="cn">NA</span>, <span class="at">labRow =</span> <span class="cn">FALSE</span>, <span class="at">labCol =</span> <span class="cn">FALSE</span>)</span>
<span id="cb60-8"><a href="#cb60-8" tabindex="-1"></a>)</span>
<span id="cb60-9"><a href="#cb60-9" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>##    user  system elapsed 
## 1700.08   42.51 1744.53</code></pre>
<p><img src="images/TraCE_decadal_hov_03.png" /></p>
</div>
<div id="heatmap" class="section level3" number="4.3.3">
<h3><span class="header-section-number">4.3.3</span> Heatmap</h3>
<p>Heatmap of the grid-point data</p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" tabindex="-1"></a>png_file <span class="ot">&lt;-</span> <span class="fu">paste</span>(file_label, <span class="st">&quot;_heatmap_03&quot;</span>, <span class="st">&quot;.png&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb62-2"><a href="#cb62-2" tabindex="-1"></a><span class="fu">png</span>(png_file, <span class="at">width=</span><span class="dv">960</span>, <span class="at">height=</span><span class="dv">960</span>)</span>
<span id="cb62-3"><a href="#cb62-3" tabindex="-1"></a><span class="fu">system.time</span>( X_heatmap <span class="ot">&lt;-</span> <span class="fu">heatmap.2</span>(X, <span class="at">cexRow =</span> <span class="fl">0.8</span>, <span class="at">cexCol =</span> <span class="fl">0.8</span>, <span class="at">scale=</span><span class="st">&quot;none&quot;</span>, <span class="at">dendrogram =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb62-4"><a href="#cb62-4" tabindex="-1"></a>        <span class="at">RowSideColors=</span>rcol, <span class="at">ColSideColors=</span>ccol, <span class="at">col=</span>zcol, <span class="at">breaks=</span>breaks, <span class="at">trace=</span><span class="st">&quot;none&quot;</span>, <span class="at">density.info =</span> <span class="st">&quot;none&quot;</span>,</span>
<span id="cb62-5"><a href="#cb62-5" tabindex="-1"></a>        <span class="at">lmat=</span><span class="fu">rbind</span>(<span class="fu">c</span>(<span class="dv">6</span>, <span class="dv">0</span>, <span class="dv">5</span>), <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">2</span>), <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">3</span>)), <span class="at">lwid=</span><span class="fu">c</span>(<span class="fl">1.0</span>, <span class="fl">0.2</span>, <span class="fl">5.0</span>), <span class="at">lhei =</span> <span class="fu">c</span>(<span class="fl">1.5</span>, <span class="fl">0.2</span>, <span class="fl">4.0</span>),</span>
<span id="cb62-6"><a href="#cb62-6" tabindex="-1"></a>        <span class="at">xlab =</span> <span class="st">&quot;ka&quot;</span>, <span class="at">ylab =</span> <span class="st">&quot;Locations&quot;</span>, <span class="at">key =</span> <span class="cn">TRUE</span>, <span class="at">key.title =</span> <span class="cn">NA</span>, <span class="at">labRow =</span> <span class="cn">FALSE</span>, <span class="at">labCol =</span> <span class="cn">FALSE</span>)</span>
<span id="cb62-7"><a href="#cb62-7" tabindex="-1"></a>)</span>
<span id="cb62-8"><a href="#cb62-8" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>##    user  system elapsed 
##  643.89   42.26  688.10 </code></pre>
<p><img src="images/TraCE_decadal_heatmap_03.png" /></p>
</div>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
