<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>netCDF in R</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; } /* Alert */
code span.an { color: #008000; } /* Annotation */
code span.at { } /* Attribute */
code span.bu { } /* BuiltIn */
code span.cf { color: #0000ff; } /* ControlFlow */
code span.ch { color: #008080; } /* Char */
code span.cn { } /* Constant */
code span.co { color: #008000; } /* Comment */
code span.cv { color: #008000; } /* CommentVar */
code span.do { color: #008000; } /* Documentation */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.im { } /* Import */
code span.in { color: #008000; } /* Information */
code span.kw { color: #0000ff; } /* Keyword */
code span.op { } /* Operator */
code span.ot { color: #ff4000; } /* Other */
code span.pp { color: #ff4000; } /* Preprocessor */
code span.sc { color: #008080; } /* SpecialChar */
code span.ss { color: #008080; } /* SpecialString */
code span.st { color: #008080; } /* String */
code span.va { } /* Variable */
code span.vs { color: #008080; } /* VerbatimString */
code span.wa { color: #008000; font-weight: bold; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link rel="stylesheet" href="html-md-01.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">R for Earth-System Science</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="overview.html">Overview</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Topics
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Lecture topics</li>
    <li>
      <a href="intro.html">Introduction: R, RStudio, and other infrastructure</a>
    </li>
    <li>
      <a href="usingR.html">Using R</a>
    </li>
    <li>
      <a href="ESSdata.html">Earth-system science data</a>
    </li>
    <li>
      <a href="netCDF.html">netCDF in R</a>
    </li>
    <li>
      <a href="raster_intro.html">raster and netCDF in R</a>
    </li>
    <li>
      <a href="hdf5_intro.html">HDF in R</a>
    </li>
    <li>
      <a href="geospatial.html">Geospatial analyses</a>
    </li>
    <li>
      <a href="RMaps.html">Maps in R</a>
    </li>
    <li>
      <a href="RMaps2.html">Maps in R (2)</a>
    </li>
    <li>
      <a href="rasterVis01.html">Visualization:  raster and rasterVis</a>
    </li>
    <li>
      <a href="vis_ggplot01.html">Visualization for explanation</a>
    </li>
    <li>
      <a href="highdim.html">Visualization of high-dimensional data</a>
    </li>
    <li>
      <a href="PCA.html">Principal components analysis</a>
    </li>
    <li>
      <a href="multivariate.html">Other multivariate analyses</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tasks
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="geog495_ex01.html">Install R and RStudio</a>
    </li>
    <li>
      <a href="packages-and-data.html">Install packages and data</a>
    </li>
    <li>
      <a href="GitHub.html">Git and GitHub</a>
    </li>
    <li>
      <a href="Ex_UsingR.html">Using R exercise</a>
    </li>
    <li>
      <a href="markdown.html">Markdown and RMarkdown</a>
    </li>
    <li>
      <a href="install_netCDF.html">Install netCDF</a>
    </li>
    <li>
      <a href="install_Panoply.html">Install Panoply</a>
    </li>
    <li>
      <a href="transfer.html">File transfer</a>
    </li>
    <li>
      <a href="project.html">GitHub project repository and web page</a>
    </li>
    <li>
      <a href="ltms-and-anomalies.html">Long-term means and anomalies</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Resources
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="readings.html">Readings</a>
    </li>
    <li>
      <a href="websites.html">Old course web pages</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="about.html">About</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">netCDF in R</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#introduction" id="toc-introduction"><span
class="toc-section-number">1</span> Introduction</a>
<ul>
<li><a href="#reading-restructuring-and-writing-netcdf-files-in-r"
id="toc-reading-restructuring-and-writing-netcdf-files-in-r"><span
class="toc-section-number">1.1</span> Reading, restructuring and writing
netCDF files in R</a></li>
</ul></li>
<li><a href="#reading-a-netcdf-data-set-using-the-ncdf4-package"
id="toc-reading-a-netcdf-data-set-using-the-ncdf4-package"><span
class="toc-section-number">2</span> Reading a netCDF data set using the
<em>ncdf4</em> package</a>
<ul>
<li><a href="#open-the-netcdf-file" id="toc-open-the-netcdf-file"><span
class="toc-section-number">2.1</span> Open the netCDF file</a></li>
<li><a href="#get-coordinate-including-time-variables"
id="toc-get-coordinate-including-time-variables"><span
class="toc-section-number">2.2</span> Get coordinate (including time)
variables</a></li>
<li><a href="#get-a-variable" id="toc-get-a-variable"><span
class="toc-section-number">2.3</span> Get a variable</a></li>
</ul></li>
<li><a href="#reshaping-from-raster-to-rectangular"
id="toc-reshaping-from-raster-to-rectangular"><span
class="toc-section-number">3</span> Reshaping from raster to
rectangular</a>
<ul>
<li><a href="#convert-the-time-variable"
id="toc-convert-the-time-variable"><span
class="toc-section-number">3.1</span> Convert the time variable</a></li>
<li><a href="#replace-netcdf-fillvalues-with-r-nas"
id="toc-replace-netcdf-fillvalues-with-r-nas"><span
class="toc-section-number">3.2</span> Replace netCDF fillvalues with R
NAs</a></li>
<li><a
href="#get-a-single-time-slice-of-the-data-create-an-r-data-frame-and-write-a-.csv-file"
id="toc-get-a-single-time-slice-of-the-data-create-an-r-data-frame-and-write-a-.csv-file"><span
class="toc-section-number">3.3</span> Get a single time slice of the
data, create an R data frame, and write a .csv file</a>
<ul>
<li><a href="#get-a-single-slice-of-data"
id="toc-get-a-single-slice-of-data"><span
class="toc-section-number">3.3.1</span> Get a single slice of
data</a></li>
<li><a href="#create-a-data-frame" id="toc-create-a-data-frame"><span
class="toc-section-number">3.3.2</span> Create a data frame</a></li>
<li><a href="#write-out-the-data-frame"
id="toc-write-out-the-data-frame"><span
class="toc-section-number">3.3.3</span> Write out the data
frame</a></li>
</ul></li>
<li><a
href="#convert-the-whole-array-to-a-data-frame-and-calculate-mtwa-mtco-and-the-annual-mean"
id="toc-convert-the-whole-array-to-a-data-frame-and-calculate-mtwa-mtco-and-the-annual-mean"><span
class="toc-section-number">3.4</span> Convert the whole array to a data
frame, and calculate MTWA, MTCO and the annual mean</a>
<ul>
<li><a href="#reshape-the-whole-array"
id="toc-reshape-the-whole-array"><span
class="toc-section-number">3.4.1</span> Reshape the whole array</a></li>
<li><a href="#get-the-annual-mean" id="toc-get-the-annual-mean"><span
class="toc-section-number">3.4.2</span> Get the annual mean</a></li>
<li><a href="#write-out-the-second-data-frame"
id="toc-write-out-the-second-data-frame"><span
class="toc-section-number">3.4.3</span> Write out the second data
frame</a></li>
</ul></li>
</ul></li>
<li><a href="#data-frame-to-array-conversionrectangular-to-raster"
id="toc-data-frame-to-array-conversionrectangular-to-raster"><span
class="toc-section-number">4</span> Data frame-to-array
conversion(rectangular to raster)</a>
<ul>
<li><a href="#convert-a-full-r-data-frame-to-an-array"
id="toc-convert-a-full-r-data-frame-to-an-array"><span
class="toc-section-number">4.1</span> Convert a “full” R data frame to
an array</a>
<ul>
<li><a href="#initial-set-up-create-dimension-variables"
id="toc-initial-set-up-create-dimension-variables"><span
class="toc-section-number">4.1.1</span> Initial set up – create
dimension variables</a></li>
<li><a href="#reshaping-a-full-data-frame-to-an-array"
id="toc-reshaping-a-full-data-frame-to-an-array"><span
class="toc-section-number">4.1.2</span> Reshaping a “full” data frame to
an array</a></li>
<li><a href="#check-the-conversion" id="toc-check-the-conversion"><span
class="toc-section-number">4.1.3</span> Check the conversion</a></li>
</ul></li>
<li><a href="#convert-a-short-r-data-frame-to-an-array"
id="toc-convert-a-short-r-data-frame-to-an-array"><span
class="toc-section-number">4.2</span> Convert a “short” R data frame to
an array</a>
<ul>
<li><a href="#initial-set-up" id="toc-initial-set-up"><span
class="toc-section-number">4.2.1</span> Initial set up</a></li>
<li><a href="#explicit-copying-from-a-data-frame-to-array"
id="toc-explicit-copying-from-a-data-frame-to-array"><span
class="toc-section-number">4.2.2</span> Explicit copying from a data
frame to array</a></li>
<li><a href="#partial-loop-avoidance"
id="toc-partial-loop-avoidance"><span
class="toc-section-number">4.2.3</span> Partial loop avoidance</a></li>
<li><a href="#complete-loop-avoidance-approach"
id="toc-complete-loop-avoidance-approach"><span
class="toc-section-number">4.2.4</span> Complete loop-avoidance
approach</a></li>
</ul></li>
</ul></li>
<li><a href="#create-and-write-a-netcdf-file"
id="toc-create-and-write-a-netcdf-file"><span
class="toc-section-number">5</span> Create and write a netCDF
file</a></li>
<li><a href="#reading-and-writing-a-projected-netcdf-file"
id="toc-reading-and-writing-a-projected-netcdf-file"><span
class="toc-section-number">6</span> Reading and writing a projected
netCDF file</a>
<ul>
<li><a href="#open-the-netcdf-file-1"
id="toc-open-the-netcdf-file-1"><span
class="toc-section-number">6.1</span> Open the netCDF file</a></li>
<li><a href="#get-coordinate-including-time-variables-1"
id="toc-get-coordinate-including-time-variables-1"><span
class="toc-section-number">6.2</span> Get coordinate (including time)
variables</a></li>
<li><a href="#get-a-variable-1" id="toc-get-a-variable-1"><span
class="toc-section-number">6.3</span> Get a variable</a></li>
</ul></li>
<li><a href="#map-the-data" id="toc-map-the-data"><span
class="toc-section-number">7</span> Map the data</a>
<ul>
<li><a href="#get-a-single-slice-of-data-1"
id="toc-get-a-single-slice-of-data-1"><span
class="toc-section-number">7.1</span> Get a single slice of
data</a></li>
<li><a href="#maps" id="toc-maps"><span
class="toc-section-number">7.2</span> Maps</a></li>
</ul></li>
<li><a href="#create-and-write-a-projected-netcdf-file"
id="toc-create-and-write-a-projected-netcdf-file"><span
class="toc-section-number">8</span> Create and write a projected netCDF
file</a></li>
</ul>
</div>

<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('right', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<div id="introduction" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>NetCDF is a widely used format for exchanging or distributing climate
data, and has also been adopted in other fields, particularly in
bioinformatics, and in other disciplines where large multidimensional
arrays of data are generated. NetCDF files are <em>self-describing</em>,
in the sense that they contain metadata that describes what is contained
in a file, such as the latitude and longitude layout of the grid, the
names and units of variables in the data set, and “attributes” that
describe things like missing value codes, or offsets and scale factors
that may have been used to compress the data. NetCDF files are also
<em>machine-independent</em> because can be transferred among servers
and computers that are running different operating systems, without
having to convert the files in some way. Originally developed for
storing and distributing climate data, such as those generated by
climate simulation or reanalysis models, the format and protocols can be
used for other gridded data sets. NetCDF libraries are developed and
maintained by Unidata <a
href="http://www.unidata.ucar.edu/software/netcdf/">http://www.unidata.ucar.edu/software/netcdf/</a>
and easy-to-use applications for producing simple visualizations of
NetCDF files exist, such as Panoply, <a
href="http://www.giss.nasa.gov/tools/panoply/">http://www.giss.nasa.gov/tools/panoply/</a>.</p>
<p>There are two versions of netCDF; netCDF3, which is widely used, but
has some size and performance limitations, and netCDF4, which supports
larger data sets and includes additional capabilities like file
compression.</p>
<p>R has the capability of reading and writing (and hence analyzing)
netCDF files, using the <code>ncdf</code> and <code>ncdf4</code>
packages provided by David Pierce, and through other packages like
<code>raster</code>, <code>metR1</code>, and <code>RNetCDF</code>. The
<code>ncdf4.helpers</code> and <code>easyNCDF</code> packages provide
some additional tools.</p>
<p>The <code>ncdf4</code> package is available on both Windows and Mac
OS X (and Linux), and supports both the older NetCDF3 format as well as
netCDF4. (See the <code>ncdf/ncdf4</code> web page at <a
href="http://cirrus.ucsd.edu/~pierce/ncdf/index.html">http://cirrus.ucsd.edu/~pierce/ncdf/index.html</a>
for further discussion.)</p>
<div id="reading-restructuring-and-writing-netcdf-files-in-r"
class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> Reading,
restructuring and writing netCDF files in R</h2>
<p>There is a common “design pattern” in analyzing data stored as
netCDF, HDF or in the native format of the <code>raster</code> package,
that includes</p>
<ol style="list-style-type: decimal">
<li>data input (using, for example, <code>ncdf4</code>,
<code>rhdf5</code> <code>raster</code>);</li>
<li>recasting/reshaping the raster brick input data into a rectangular
data frame;</li>
<li>analysis and visualization;</li>
<li>recasting/reshaping a “results” data frame back to a raster or
raster brick; and</li>
<li>data output, using the same packages as in step 1.</li>
</ol>
<p>The examples provided here include</p>
<ul>
<li>reading a netCDF file using the ncdf4 package (netCDF4)</li>
<li>reshaping a netCDF “brick” of data into a data frame</li>
<li>reshaping a data frame into an array or “brick”</li>
<li>writing a netCDF file using the ncdf4 package</li>
</ul>
<p>The examples make use of a netCDF file of climate data from the
Climate Research Unit <a
href="http://www.cru.uea.ac.uk/data">http://www.cru.uea.ac.uk/data</a>,
consisting of long-term mean values (1961-1990) of near-surface air
temperature on a 0.5-degree grid (for land points). The dimensions of
the array are 720 (longitudes) x 360 (latitudes) x 12 (months), thus
forming a raster “stack” or “brick” consisting of 12 layers.</p>
<p>The data are available on <code>ClimateLab.uoregon.edu</code> (see
File transfer on the Tasks tab), in the <code>/nc_files</code> folder,
with the file name <code>cru10min30_tmp.nc</code>. Download the netCDF
file to a convenient folder.</p>
<p><a href="netCDF.html">[Back to top]</a></p>
</div>
</div>
<div id="reading-a-netcdf-data-set-using-the-ncdf4-package"
class="section level1" number="2">
<h1><span class="header-section-number">2</span> Reading a netCDF data
set using the <em>ncdf4</em> package</h1>
<p>To begin, load the <code>ncdf4</code> package.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># load the ncdf4 package</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(ncdf4)</span></code></pre></div>
<p>The file is assumed to be a CF-compliant netCDF file, in which the
three main spatiotemporal dimensions allear the the relative order of
time (T-coordinate), height or depth (Z-coordinate), latitude (or
Y-coordinate), and longitude (or X-coordinate). In this example, the
file is a 3-D file with T, Y and X coordinates (month of the year,
latitude, and longitude). First, set the values for some temporary
variables. <code>ncpath</code> is the path to where the file was
downloaded, <code>ncname</code> is the name of the netCDF file, while
<code>dname</code> is the name of the variable that will be read in.</p>
<div id="open-the-netcdf-file" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Open the netCDF
file</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># set path and filename</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>ncpath <span class="ot">&lt;-</span> <span class="st">&quot;/Users/bartlein/Projects/ESSD/data/nc_files/&quot;</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>ncname <span class="ot">&lt;-</span> <span class="st">&quot;cru10min30_tmp&quot;</span>  </span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>ncfname <span class="ot">&lt;-</span> <span class="fu">paste</span>(ncpath, ncname, <span class="st">&quot;.nc&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>dname <span class="ot">&lt;-</span> <span class="st">&quot;tmp&quot;</span>  <span class="co"># note: tmp means temperature (not temporary)</span></span></code></pre></div>
<p>Open the NetCDF data set, and print some basic information. The
<code>print()</code> function applied to the <code>ncin</code> object
produces information similar to that produced by the command-line
utility <code>ncdump</code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># open a netCDF file</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>ncin <span class="ot">&lt;-</span> <span class="fu">nc_open</span>(ncfname)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="fu">print</span>(ncin)</span></code></pre></div>
<pre><code>## File /Users/bartlein/Projects/ESSD/data/nc_files/cru10min30_tmp.nc (NC_FORMAT_CLASSIC):
## 
##      2 variables (excluding dimension variables):
##         float climatology_bounds[nv,time]   
##         float tmp[lon,lat,time]   
##             long_name: air_temperature
##             units: degC
##             _FillValue: -99
##             source: E:\Projects\cru\data\cru_cl_2.0\nc_files\cru10min_tmp.nc
## 
##      4 dimensions:
##         lon  Size:720 
##             standard_name: longitude
##             long_name: longitude
##             units: degrees_east
##             axis: X
##         lat  Size:360 
##             standard_name: latitude
##             long_name: latitude
##             units: degrees_north
##             axis: Y
##         time  Size:12 
##             standard_name: time
##             long_name: time
##             units: days since 1900-01-01 00:00:00.0 -0:00
##             axis: T
##             calendar: standard
##             climatology: climatology_bounds
##         nv  Size:2 (no dimvar)
## 
##     7 global attributes:
##         data: CRU CL 2.0 1961-1990 Monthly Averages
##         title: CRU CL 2.0 -- 10min grid sampled every 0.5 degree
##         institution: http://www.cru.uea.ac.uk/
##         source: http://www.cru.uea.ac.uk/~markn/cru05/cru05_intro.html
##         references: New et al. (2002) Climate Res 21:1-25
##         history: P.J. Bartlein, 19 Jun 2005
##         Conventions: CF-1.0</code></pre>
<p>Note that in an <code>ncdump</code> of the file, the coordinates of
the variable <code>tmp</code> are listed in the reverse order as they
are here (e.g. <code>tmp(time, lat, lon)</code>).</p>
</div>
<div id="get-coordinate-including-time-variables" class="section level2"
number="2.2">
<h2><span class="header-section-number">2.2</span> Get coordinate
(including time) variables</h2>
<p>Next, get the coordinate variables <code>longitude</code> and
<code>latitude</code> are read using the <code>ncvar_get()</code>
function, and the first few values of each are listed using the
<code>head()</code> and <code>tail()</code> functions. The number of
longitude and latitude values can be verified using the
<code>dim()</code> function:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># get longitude and latitude</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>lon <span class="ot">&lt;-</span> <span class="fu">ncvar_get</span>(ncin,<span class="st">&quot;lon&quot;</span>)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>nlon <span class="ot">&lt;-</span> <span class="fu">dim</span>(lon)</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="fu">head</span>(lon)</span></code></pre></div>
<pre><code>## [1] -179.75 -179.25 -178.75 -178.25 -177.75 -177.25</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>lat <span class="ot">&lt;-</span> <span class="fu">ncvar_get</span>(ncin,<span class="st">&quot;lat&quot;</span>)</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>nlat <span class="ot">&lt;-</span> <span class="fu">dim</span>(lat)</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="fu">head</span>(lat)</span></code></pre></div>
<pre><code>## [1] -89.75 -89.25 -88.75 -88.25 -87.75 -87.25</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">c</span>(nlon,nlat))</span></code></pre></div>
<pre><code>## [1] 720 360</code></pre>
<p>Get the time variable and its attributes using the
<code>ncvar_get()</code> and <code>ncatt_get()</code> functions, and
list them, and also get the number of time steps using the
<code>dim()</code> function.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co"># get time</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>time <span class="ot">&lt;-</span> <span class="fu">ncvar_get</span>(ncin,<span class="st">&quot;time&quot;</span>)</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>time</span></code></pre></div>
<pre><code>##  [1] 27773.5 27803.5 27833.5 27864.0 27894.5 27925.0 27955.5 27986.5 28017.0 28047.5 28078.0 28108.5</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>tunits <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,<span class="st">&quot;time&quot;</span>,<span class="st">&quot;units&quot;</span>)</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>nt <span class="ot">&lt;-</span> <span class="fu">dim</span>(time)</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>nt</span></code></pre></div>
<pre><code>## [1] 12</code></pre>
<p>Print the time units string. Note the structure of the time units
attribute. The object <code>tunits</code> has two components
<code>hasatt</code> (a logical variable), and <code>tunits$value</code>,
the actual “time since” string.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>tunits</span></code></pre></div>
<pre><code>## $hasatt
## [1] TRUE
## 
## $value
## [1] &quot;days since 1900-01-01 00:00:00.0 -0:00&quot;</code></pre>
</div>
<div id="get-a-variable" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> Get a variable</h2>
<p>Get the the variable (<code>tmp</code>) and its attributes, and
verify the size of the array.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="co"># get temperature</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>tmp_array <span class="ot">&lt;-</span> <span class="fu">ncvar_get</span>(ncin,dname)</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a>dlname <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,dname,<span class="st">&quot;long_name&quot;</span>)</span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a>dunits <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,dname,<span class="st">&quot;units&quot;</span>)</span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a>fillvalue <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,dname,<span class="st">&quot;_FillValue&quot;</span>)</span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a><span class="fu">dim</span>(tmp_array)</span></code></pre></div>
<pre><code>## [1] 720 360  12</code></pre>
<p>Get the global attributes. The attributes can be listed, by simply
typing an attribute name at the command line.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="co"># get global attributes</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>title <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,<span class="dv">0</span>,<span class="st">&quot;title&quot;</span>)</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>institution <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,<span class="dv">0</span>,<span class="st">&quot;institution&quot;</span>)</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>datasource <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,<span class="dv">0</span>,<span class="st">&quot;source&quot;</span>)</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>references <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,<span class="dv">0</span>,<span class="st">&quot;references&quot;</span>)</span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a>history <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,<span class="dv">0</span>,<span class="st">&quot;history&quot;</span>)</span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a>Conventions <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,<span class="dv">0</span>,<span class="st">&quot;Conventions&quot;</span>)</span></code></pre></div>
<p>Close the netCDF file using the <code>nc_close()</code> function.</p>
<p>Check what’s in the current workspace:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="fu">ls</span>()</span></code></pre></div>
<pre><code>##  [1] &quot;BF&quot;               &quot;carp&quot;             &quot;carp_heat&quot;        &quot;col&quot;              &quot;Conventions&quot;     
##  [6] &quot;csv_file&quot;         &quot;csv_name&quot;         &quot;csv_path&quot;         &quot;csvfile&quot;          &quot;csvname&quot;         
## [11] &quot;csvpath&quot;          &quot;cutpts&quot;           &quot;datasource&quot;       &quot;dlname&quot;           &quot;dname&quot;           
## [16] &quot;dunits&quot;           &quot;fillvalue&quot;        &quot;gf&quot;               &quot;grid&quot;             &quot;h1&quot;              
## [21] &quot;h2&quot;               &quot;hdf_file&quot;         &quot;hdf_name&quot;         &quot;hdf_path&quot;         &quot;history&quot;         
## [26] &quot;institution&quot;      &quot;lat&quot;              &quot;lat_array&quot;        &quot;lon&quot;              &quot;lon_array&quot;       
## [31] &quot;mapTheme&quot;         &quot;mb2&quot;              &quot;megabiome_name&quot;   &quot;n&quot;                &quot;napol&quot;           
## [36] &quot;ncfname&quot;          &quot;ncin&quot;             &quot;ncname&quot;           &quot;ncpath&quot;           &quot;nlat&quot;            
## [41] &quot;nlon&quot;             &quot;nm&quot;               &quot;nt&quot;               &quot;ny&quot;               &quot;orstationc&quot;      
## [46] &quot;orstationc_file&quot;  &quot;orstationc_name&quot;  &quot;orstationc_path&quot;  &quot;plt&quot;              &quot;pngfile&quot;         
## [51] &quot;references&quot;       &quot;shp_file&quot;         &quot;shp_name&quot;         &quot;shp_path&quot;         &quot;sort_cols&quot;       
## [56] &quot;stcarp&quot;           &quot;tas_anm_ann&quot;      &quot;tas2&quot;             &quot;tcarp&quot;            &quot;time&quot;            
## [61] &quot;title&quot;            &quot;tmp_array&quot;        &quot;tr21dec_file&quot;     &quot;tr21dec_name&quot;     &quot;tr21dec_path&quot;    
## [66] &quot;tunits&quot;           &quot;UVI&quot;              &quot;vegtype_name&quot;     &quot;world_outline&quot;    &quot;world_outline_gg&quot;
## [71] &quot;world_shp&quot;</code></pre>
<p><a href="netCDF.html">[Back to top]</a></p>
</div>
</div>
<div id="reshaping-from-raster-to-rectangular" class="section level1"
number="3">
<h1><span class="header-section-number">3</span> Reshaping from raster
to rectangular</h1>
<p>NetCDF files or data sets are naturally 2-D raster slabs
(e.g. longitude by latitude “slices”), 3-D bricks (e.g. longitude by
latitude by time), or 4-D arrays (e.g. longitude by latitude by height
by time), while most data analysis routines in R expect 2-D
variable-by-observation “tidy” data frames. (There is an exception to
this expectation in some cases like principle components analysis (PCA)
in which variables are locations and the observations are times.)
Therefore, before analysis, a 3-D or 4-D array in the netCDF files must
be “flattened” into a 2-D array. In addition, time is usually stored as
the CF (Climate Forecast) “time since” format that is not usually
human-readable. Here are some example conversions:</p>
<p>Load some necessary packages (install them if necessary)</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="co"># load some packages</span></span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a><span class="fu">library</span>(chron)</span></code></pre></div>
<pre><code>## 
## Attaching package: &#39;chron&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:raster&#39;:
## 
##     origin, origin&lt;-</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a><span class="fu">library</span>(lattice)</span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span></code></pre></div>
<div id="convert-the-time-variable" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Convert the time
variable</h2>
<p>The time variable, in “time-since” units can be converted into “real”
(or more easily readable) time values by splitting the time
<code>tunits$value</code> string into its component parts, and then
using the <code>chron()</code> function to determine the absolute value
of each time value from the time origin.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a><span class="co"># convert time -- split the time units string into fields</span></span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a>tustr <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(tunits<span class="sc">$</span>value, <span class="st">&quot; &quot;</span>)</span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a>tdstr <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(<span class="fu">unlist</span>(tustr)[<span class="dv">3</span>], <span class="st">&quot;-&quot;</span>)</span>
<span id="cb26-4"><a href="#cb26-4" tabindex="-1"></a>tmonth <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(<span class="fu">unlist</span>(tdstr)[<span class="dv">2</span>])</span>
<span id="cb26-5"><a href="#cb26-5" tabindex="-1"></a>tday <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(<span class="fu">unlist</span>(tdstr)[<span class="dv">3</span>])</span>
<span id="cb26-6"><a href="#cb26-6" tabindex="-1"></a>tyear <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(<span class="fu">unlist</span>(tdstr)[<span class="dv">1</span>])</span>
<span id="cb26-7"><a href="#cb26-7" tabindex="-1"></a><span class="fu">chron</span>(time,<span class="at">origin=</span><span class="fu">c</span>(tmonth, tday, tyear))</span></code></pre></div>
<pre><code>##  [1] (01/16/76 12:00:00) (02/15/76 12:00:00) (03/16/76 12:00:00) (04/16/76 00:00:00) (05/16/76 12:00:00)
##  [6] (06/16/76 00:00:00) (07/16/76 12:00:00) (08/16/76 12:00:00) (09/16/76 00:00:00) (10/16/76 12:00:00)
## [11] (11/16/76 00:00:00) (12/16/76 12:00:00)</code></pre>
<p>The “time-stamp” for this particular data set, which represents
long-term means over the interal 1961-1990 is the mid-point of the
interval for each month of the year, i.e. the middle of January, in the
middle of the range of years (e.g. 1976). This is somewhat arbitrary,
and despite there being a convention for representing “climatological
statistics” there are other ways in which the “time” associated with a
long-term mean is represented.</p>
</div>
<div id="replace-netcdf-fillvalues-with-r-nas" class="section level2"
number="3.2">
<h2><span class="header-section-number">3.2</span> Replace netCDF
fillvalues with R NAs</h2>
<p>In a netCDF file, values of a variable that are either missing or
simply not available (i.e. ocean grid points in a terrestrial data set)
are flagged using specific “fill values” (<code>_FillValue</code>) or
missing values (<code>missing_value</code>), the particular values of
which are set as attributes of a variable. In R, such unavailable data
are indicated using the “<code>NA</code>” value. The following code
fragment illustrates how to replace the netCDF variable’s fill values
with R <code>NA</code>’s .</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" tabindex="-1"></a><span class="co"># replace netCDF fill values with NA&#39;s</span></span>
<span id="cb28-2"><a href="#cb28-2" tabindex="-1"></a>tmp_array[tmp_array<span class="sc">==</span>fillvalue<span class="sc">$</span>value] <span class="ot">&lt;-</span> <span class="cn">NA</span></span></code></pre></div>
<p>The <code>head()</code> function can be used before and after
executing the “square bracket” selection and replacement to verify that
the <code>NA</code> values have indeed replace the netCDF fill
values(<code>head(as.vector(temp_array[,,1]))</code>. The total number
of non-missing (i.e. land, except for Antarctica, which is not present
in the data set) grid cells can be gotten by determining the length of a
vector of values representing one slice from the brick, omitting the
<code>NA</code> values:</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">na.omit</span>(<span class="fu">as.vector</span>(tmp_array[,,<span class="dv">1</span>])))</span></code></pre></div>
<pre><code>## [1] 62961</code></pre>
</div>
<div
id="get-a-single-time-slice-of-the-data-create-an-r-data-frame-and-write-a-.csv-file"
class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Get a single time
slice of the data, create an R data frame, and write a .csv file</h2>
<p>NetCDF variables are read and written as one-dimensional vectors
(e.g. longitudes), two-dimensional arrays or matrices (raster “slices”),
or multi-dimensional arrays (raster “bricks”). In such data structures,
the coordinate values for each grid point are implicit, inferred from
the marginal values of, for example, longitude, latitude and time. In
contrast, in R, the principal data structure for a variable is the data
frame. In the kinds of data sets usually stored as netCDF files, each
row in the data frame will contain the data for an individual grid
point, with each column representing a particular variable, including
explicit values for longitude and latitude (and perhaps time). In the
example CRU data set considered here, the variables would consist of
longitude, latitude and 12 columns of long-term means for each month,
with the full data set thus consisting of 259200 rows (720 by 360) and
14 columns.</p>
<p>This particular structure of this data set can be illustrated by
selecting a single slice from the temperature “brick”, turning it into a
data frame with three variables and 720 by 360 rows, and writing it out
as a .csv file.</p>
<div id="get-a-single-slice-of-data" class="section level3"
number="3.3.1">
<h3><span class="header-section-number">3.3.1</span> Get a single slice
of data</h3>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="co"># get a single slice or layer (January)</span></span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb31-3"><a href="#cb31-3" tabindex="-1"></a>tmp_slice <span class="ot">&lt;-</span> tmp_array[,,m]</span></code></pre></div>
<p>The dimensions of <code>tmp_slice</code>, e.g. 720, 360, can be
verified using the <code>dim()</code> function.</p>
<p>A quick look (map) of the extracted slice of data can be gotten using
the <code>image()</code> function.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="co"># quick map</span></span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a><span class="fu">image</span>(lon,lat,tmp_slice, <span class="at">col=</span><span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>,<span class="st">&quot;RdBu&quot;</span>)))</span></code></pre></div>
<p><img src="netCDF_files/figure-html/image1-1.png" width="75%" height="75%" /></p>
<p>A better map can be obtained using the <code>levelplot()</code>
function from the <code>lattice</code> package. The
<code>expand.grid()</code> function is used to create a set of 720 by
360 pairs of latitude and longitude values (with latitudes varying most
rapidly), one for each element in the <code>tmp_slice</code> array.
Specific values of the cutpoints of temperature categories are defined
to cover the range of temperature values here.</p>
<div class="sourceCode" id="cb33"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" tabindex="-1"></a><span class="co"># levelplot of the slice</span></span>
<span id="cb33-2"><a href="#cb33-2" tabindex="-1"></a>grid <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">lon=</span>lon, <span class="at">lat=</span>lat)</span>
<span id="cb33-3"><a href="#cb33-3" tabindex="-1"></a>cutpts <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">50</span>,<span class="sc">-</span><span class="dv">40</span>,<span class="sc">-</span><span class="dv">30</span>,<span class="sc">-</span><span class="dv">20</span>,<span class="sc">-</span><span class="dv">10</span>,<span class="dv">0</span>,<span class="dv">10</span>,<span class="dv">20</span>,<span class="dv">30</span>,<span class="dv">40</span>,<span class="dv">50</span>)</span>
<span id="cb33-4"><a href="#cb33-4" tabindex="-1"></a><span class="fu">levelplot</span>(tmp_slice <span class="sc">~</span> lon <span class="sc">*</span> lat, <span class="at">data=</span>grid, <span class="at">at=</span>cutpts, <span class="at">cuts=</span><span class="dv">11</span>, <span class="at">pretty=</span>T, </span>
<span id="cb33-5"><a href="#cb33-5" tabindex="-1"></a>  <span class="at">col.regions=</span>(<span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>,<span class="st">&quot;RdBu&quot;</span>))))</span></code></pre></div>
<p><img src="netCDF_files/figure-html/levelplot1-1.png" width="75%" height="75%" /></p>
</div>
<div id="create-a-data-frame" class="section level3" number="3.3.2">
<h3><span class="header-section-number">3.3.2</span> Create a data
frame</h3>
<p>To create a data frame, the <code>expand.grid()</code> and
<code>as.matrix()</code> functions are used to create the 259200 pairs
(i.e. rows) of values of longitude and latitude (the columns), and the
<code>as.vector()</code> function is used to “unstack” the slice of data
into a long vector. The size of the objects that are created can be
verified using the <code>dim()</code> and <code>length()</code>
functions.</p>
<div class="sourceCode" id="cb34"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" tabindex="-1"></a><span class="co"># create dataframe -- reshape data</span></span>
<span id="cb34-2"><a href="#cb34-2" tabindex="-1"></a><span class="co"># matrix (nlon*nlat rows by 2 cols) of lons and lats</span></span>
<span id="cb34-3"><a href="#cb34-3" tabindex="-1"></a>lonlat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">expand.grid</span>(lon,lat))</span>
<span id="cb34-4"><a href="#cb34-4" tabindex="-1"></a><span class="fu">dim</span>(lonlat)</span></code></pre></div>
<pre><code>## [1] 259200      2</code></pre>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" tabindex="-1"></a><span class="co"># vector of `tmp` values</span></span>
<span id="cb36-2"><a href="#cb36-2" tabindex="-1"></a>tmp_vec <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(tmp_slice)</span>
<span id="cb36-3"><a href="#cb36-3" tabindex="-1"></a><span class="fu">length</span>(tmp_vec)</span></code></pre></div>
<pre><code>## [1] 259200</code></pre>
<p>The <code>data.frame()</code> and <code>cbind()</code> functions are
used to assemble the columns of the data frame, which are assigned
appropriate names using the <code>names()</code> function (on the
left-hand side of assignment operator). The <code>head()</code>
function, applied on top of the <code>na.omit()</code> function lists
the first rows of values without <code>NA</code>s:</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="co"># create dataframe and add names</span></span>
<span id="cb38-2"><a href="#cb38-2" tabindex="-1"></a>tmp_df01 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">cbind</span>(lonlat,tmp_vec))</span>
<span id="cb38-3"><a href="#cb38-3" tabindex="-1"></a><span class="fu">names</span>(tmp_df01) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;lon&quot;</span>,<span class="st">&quot;lat&quot;</span>,<span class="fu">paste</span>(dname,<span class="fu">as.character</span>(m), <span class="at">sep=</span><span class="st">&quot;_&quot;</span>))</span>
<span id="cb38-4"><a href="#cb38-4" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">na.omit</span>(tmp_df01), <span class="dv">10</span>)</span></code></pre></div>
<pre><code>##          lon    lat tmp_1
## 49186 -67.25 -55.75   8.2
## 49901 -69.75 -55.25   7.9
## 49902 -69.25 -55.25   8.4
## 49903 -68.75 -55.25   7.8
## 49904 -68.25 -55.25   8.9
## 49905 -67.75 -55.25   9.1
## 49906 -67.25 -55.25   9.0
## 50617 -71.75 -54.75   8.8
## 50619 -70.75 -54.75   8.7
## 50620 -70.25 -54.75   7.9</code></pre>
</div>
<div id="write-out-the-data-frame" class="section level3"
number="3.3.3">
<h3><span class="header-section-number">3.3.3</span> Write out the data
frame</h3>
<p>Finally the data frame is written out to the working directory as a
.csv file, using <code>na.omit()</code> again to drop the observations
with missing data (i.e. ocean points and Antarctica).</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" tabindex="-1"></a><span class="co"># set path and filename</span></span>
<span id="cb40-2"><a href="#cb40-2" tabindex="-1"></a>csvpath <span class="ot">&lt;-</span> <span class="st">&quot;/Users/bartlein/Projects/ESSD/data/csv_files&quot;</span></span>
<span id="cb40-3"><a href="#cb40-3" tabindex="-1"></a>csvname <span class="ot">&lt;-</span> <span class="st">&quot;cru_tmp_1.csv&quot;</span></span>
<span id="cb40-4"><a href="#cb40-4" tabindex="-1"></a>csvfile <span class="ot">&lt;-</span> <span class="fu">paste</span>(csvpath, csvname, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb40-5"><a href="#cb40-5" tabindex="-1"></a><span class="fu">write.table</span>(<span class="fu">na.omit</span>(tmp_df01),csvfile, <span class="at">row.names=</span><span class="cn">FALSE</span>, <span class="at">sep=</span><span class="st">&quot;,&quot;</span>)</span></code></pre></div>
</div>
</div>
<div
id="convert-the-whole-array-to-a-data-frame-and-calculate-mtwa-mtco-and-the-annual-mean"
class="section level2" number="3.4">
<h2><span class="header-section-number">3.4</span> Convert the whole
array to a data frame, and calculate MTWA, MTCO and the annual mean</h2>
<p>The idea here is to convert the <em>nlon</em> by <em>nlat</em> by
<em>nt</em> 3-D array into a (<em>nlon</em> x <em>nlat</em>) by
<em>nt</em> 2-D matrix. (This will work if the netCDF data set was
written as a CF-compliant data set, with arrays dimensioned as in
Fortran, i.e. as nlon x nlat x nt arrays).</p>
<div id="reshape-the-whole-array" class="section level3" number="3.4.1">
<h3><span class="header-section-number">3.4.1</span> Reshape the whole
array</h3>
<p>Convert the array to a vector. First, create a long vector
<code>tmp_vec_long</code> using the <code>as.vector()</code> reshaping
function, and verify its length, which should be 3110400. (This will
work only if the netCDF file (and hence the data array) follow the “CF”
conventions, i.e. that the variable tmp has been defined to have
dimensions <code>nlon</code> by <code>nlat</code> by <code>nt</code>, in
that order.)</p>
<div class="sourceCode" id="cb41"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" tabindex="-1"></a><span class="co"># reshape the array into vector</span></span>
<span id="cb41-2"><a href="#cb41-2" tabindex="-1"></a>tmp_vec_long <span class="ot">&lt;-</span> <span class="fu">as.vector</span>(tmp_array)</span>
<span id="cb41-3"><a href="#cb41-3" tabindex="-1"></a><span class="fu">length</span>(tmp_vec_long)</span></code></pre></div>
<pre><code>## [1] 3110400</code></pre>
<p>Then reshape that vector into a 259200 by 12 matrix using the
<code>matrix()</code> function, and verify its dimensions, which should
be 259200 by 12.</p>
<div class="sourceCode" id="cb43"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" tabindex="-1"></a><span class="co"># reshape the vector into a matrix</span></span>
<span id="cb43-2"><a href="#cb43-2" tabindex="-1"></a>tmp_mat <span class="ot">&lt;-</span> <span class="fu">matrix</span>(tmp_vec_long, <span class="at">nrow=</span>nlon<span class="sc">*</span>nlat, <span class="at">ncol=</span>nt)</span>
<span id="cb43-3"><a href="#cb43-3" tabindex="-1"></a><span class="fu">dim</span>(tmp_mat)</span></code></pre></div>
<pre><code>## [1] 259200     12</code></pre>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">na.omit</span>(tmp_mat))</span></code></pre></div>
<pre><code>##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10] [,11] [,12]
## [1,]  8.2  8.2  6.9  5.2  3.1  1.7  1.1  1.8  3.1   4.9   6.3   7.5
## [2,]  7.9  7.8  6.7  5.1  3.2  1.9  1.4  1.9  3.2   4.8   6.1   7.3
## [3,]  8.4  8.3  7.2  5.5  3.5  2.1  1.7  2.2  3.6   5.3   6.6   7.8
## [4,]  7.8  7.7  6.5  4.8  2.6  1.2  0.8  1.5  3.0   4.7   6.0   7.2
## [5,]  8.9  8.8  7.5  5.7  3.3  1.8  1.5  2.2  3.9   5.7   7.1   8.3
## [6,]  9.1  9.0  7.5  5.7  3.3  1.8  1.3  2.2  3.8   5.7   7.2   8.4</code></pre>
<p>Create the second data frame from the <code>tmp_mat</code>
matrix.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" tabindex="-1"></a><span class="co"># create a dataframe</span></span>
<span id="cb47-2"><a href="#cb47-2" tabindex="-1"></a>lonlat <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">expand.grid</span>(lon,lat))</span>
<span id="cb47-3"><a href="#cb47-3" tabindex="-1"></a>tmp_df02 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">cbind</span>(lonlat,tmp_mat))</span>
<span id="cb47-4"><a href="#cb47-4" tabindex="-1"></a><span class="fu">names</span>(tmp_df02) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;lon&quot;</span>,<span class="st">&quot;lat&quot;</span>,<span class="st">&quot;tmpJan&quot;</span>,<span class="st">&quot;tmpFeb&quot;</span>,<span class="st">&quot;tmpMar&quot;</span>,<span class="st">&quot;tmpApr&quot;</span>,<span class="st">&quot;tmpMay&quot;</span>,<span class="st">&quot;tmpJun&quot;</span>,</span>
<span id="cb47-5"><a href="#cb47-5" tabindex="-1"></a>  <span class="st">&quot;tmpJul&quot;</span>,<span class="st">&quot;tmpAug&quot;</span>,<span class="st">&quot;tmpSep&quot;</span>,<span class="st">&quot;tmpOct&quot;</span>,<span class="st">&quot;tmpNov&quot;</span>,<span class="st">&quot;tmpDec&quot;</span>)</span>
<span id="cb47-6"><a href="#cb47-6" tabindex="-1"></a><span class="co"># options(width=96)</span></span>
<span id="cb47-7"><a href="#cb47-7" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">na.omit</span>(tmp_df02, <span class="dv">20</span>))</span></code></pre></div>
<pre><code>##          lon    lat tmpJan tmpFeb tmpMar tmpApr tmpMay tmpJun tmpJul tmpAug tmpSep tmpOct tmpNov tmpDec
## 49186 -67.25 -55.75    8.2    8.2    6.9    5.2    3.1    1.7    1.1    1.8    3.1    4.9    6.3    7.5
## 49901 -69.75 -55.25    7.9    7.8    6.7    5.1    3.2    1.9    1.4    1.9    3.2    4.8    6.1    7.3
## 49902 -69.25 -55.25    8.4    8.3    7.2    5.5    3.5    2.1    1.7    2.2    3.6    5.3    6.6    7.8
## 49903 -68.75 -55.25    7.8    7.7    6.5    4.8    2.6    1.2    0.8    1.5    3.0    4.7    6.0    7.2
## 49904 -68.25 -55.25    8.9    8.8    7.5    5.7    3.3    1.8    1.5    2.2    3.9    5.7    7.1    8.3
## 49905 -67.75 -55.25    9.1    9.0    7.5    5.7    3.3    1.8    1.3    2.2    3.8    5.7    7.2    8.4</code></pre>
</div>
<div id="get-the-annual-mean" class="section level3" number="3.4.2">
<h3><span class="header-section-number">3.4.2</span> Get the annual
mean</h3>
<p>Get the annual mean, mtwa and mtco (mean temperatures of the warmest
and coldest montth) values and add them to the second data frame.</p>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" tabindex="-1"></a><span class="co"># get the annual mean and MTWA and MTCO</span></span>
<span id="cb49-2"><a href="#cb49-2" tabindex="-1"></a>tmp_df02<span class="sc">$</span>mtwa <span class="ot">&lt;-</span> <span class="fu">apply</span>(tmp_df02[<span class="dv">3</span><span class="sc">:</span><span class="dv">14</span>],<span class="dv">1</span>,max) <span class="co"># mtwa</span></span>
<span id="cb49-3"><a href="#cb49-3" tabindex="-1"></a>tmp_df02<span class="sc">$</span>mtco <span class="ot">&lt;-</span> <span class="fu">apply</span>(tmp_df02[<span class="dv">3</span><span class="sc">:</span><span class="dv">14</span>],<span class="dv">1</span>,min) <span class="co"># mtco</span></span>
<span id="cb49-4"><a href="#cb49-4" tabindex="-1"></a>tmp_df02<span class="sc">$</span>mat <span class="ot">&lt;-</span> <span class="fu">apply</span>(tmp_df02[<span class="dv">3</span><span class="sc">:</span><span class="dv">14</span>],<span class="dv">1</span>,mean) <span class="co"># annual (i.e. row) means</span></span>
<span id="cb49-5"><a href="#cb49-5" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">na.omit</span>(tmp_df02))</span></code></pre></div>
<pre><code>##          lon    lat tmpJan tmpFeb tmpMar tmpApr tmpMay tmpJun tmpJul tmpAug tmpSep tmpOct tmpNov tmpDec
## 49186 -67.25 -55.75    8.2    8.2    6.9    5.2    3.1    1.7    1.1    1.8    3.1    4.9    6.3    7.5
## 49901 -69.75 -55.25    7.9    7.8    6.7    5.1    3.2    1.9    1.4    1.9    3.2    4.8    6.1    7.3
## 49902 -69.25 -55.25    8.4    8.3    7.2    5.5    3.5    2.1    1.7    2.2    3.6    5.3    6.6    7.8
## 49903 -68.75 -55.25    7.8    7.7    6.5    4.8    2.6    1.2    0.8    1.5    3.0    4.7    6.0    7.2
## 49904 -68.25 -55.25    8.9    8.8    7.5    5.7    3.3    1.8    1.5    2.2    3.9    5.7    7.1    8.3
## 49905 -67.75 -55.25    9.1    9.0    7.5    5.7    3.3    1.8    1.3    2.2    3.8    5.7    7.2    8.4
##       mtwa mtco      mat
## 49186  8.2  1.1 4.833333
## 49901  7.9  1.4 4.775000
## 49902  8.4  1.7 5.183333
## 49903  7.8  0.8 4.483333
## 49904  8.9  1.5 5.391667
## 49905  9.1  1.3 5.416667</code></pre>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" tabindex="-1"></a><span class="fu">dim</span>(<span class="fu">na.omit</span>(tmp_df02))</span></code></pre></div>
<pre><code>## [1] 62961    17</code></pre>
</div>
<div id="write-out-the-second-data-frame" class="section level3"
number="3.4.3">
<h3><span class="header-section-number">3.4.3</span> Write out the
second data frame</h3>
<p>Write the second data frame out as a .csv file, dropping NAs.</p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" tabindex="-1"></a><span class="co"># write out the dataframe as a .csv file</span></span>
<span id="cb53-2"><a href="#cb53-2" tabindex="-1"></a>csvpath <span class="ot">&lt;-</span> <span class="st">&quot;/Users/bartlein/Projects/ESSD/data/csv_files/&quot;</span></span>
<span id="cb53-3"><a href="#cb53-3" tabindex="-1"></a>csvname <span class="ot">&lt;-</span> <span class="st">&quot;cru_tmp_2.csv&quot;</span></span>
<span id="cb53-4"><a href="#cb53-4" tabindex="-1"></a>csvfile <span class="ot">&lt;-</span> <span class="fu">paste</span>(csvpath, csvname, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb53-5"><a href="#cb53-5" tabindex="-1"></a><span class="fu">write.table</span>(<span class="fu">na.omit</span>(tmp_df02),csvfile, <span class="at">row.names=</span><span class="cn">FALSE</span>, <span class="at">sep=</span><span class="st">&quot;,&quot;</span>)</span></code></pre></div>
<p>Create a third data frame, with only non-missing values. This will be
used later to demonstrate how to convert a “short” data frame into full
matrix or array for writing out as a netCDF file.</p>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" tabindex="-1"></a><span class="co"># create a dataframe without missing values</span></span>
<span id="cb54-2"><a href="#cb54-2" tabindex="-1"></a>tmp_df03 <span class="ot">&lt;-</span> <span class="fu">na.omit</span>(tmp_df02)</span>
<span id="cb54-3"><a href="#cb54-3" tabindex="-1"></a><span class="fu">head</span>(tmp_df03)</span></code></pre></div>
<pre><code>##          lon    lat tmpJan tmpFeb tmpMar tmpApr tmpMay tmpJun tmpJul tmpAug tmpSep tmpOct tmpNov tmpDec
## 49186 -67.25 -55.75    8.2    8.2    6.9    5.2    3.1    1.7    1.1    1.8    3.1    4.9    6.3    7.5
## 49901 -69.75 -55.25    7.9    7.8    6.7    5.1    3.2    1.9    1.4    1.9    3.2    4.8    6.1    7.3
## 49902 -69.25 -55.25    8.4    8.3    7.2    5.5    3.5    2.1    1.7    2.2    3.6    5.3    6.6    7.8
## 49903 -68.75 -55.25    7.8    7.7    6.5    4.8    2.6    1.2    0.8    1.5    3.0    4.7    6.0    7.2
## 49904 -68.25 -55.25    8.9    8.8    7.5    5.7    3.3    1.8    1.5    2.2    3.9    5.7    7.1    8.3
## 49905 -67.75 -55.25    9.1    9.0    7.5    5.7    3.3    1.8    1.3    2.2    3.8    5.7    7.2    8.4
##       mtwa mtco      mat
## 49186  8.2  1.1 4.833333
## 49901  7.9  1.4 4.775000
## 49902  8.4  1.7 5.183333
## 49903  7.8  0.8 4.483333
## 49904  8.9  1.5 5.391667
## 49905  9.1  1.3 5.416667</code></pre>
<p>Check what’s in the current workspace now:</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" tabindex="-1"></a><span class="fu">ls</span>()</span></code></pre></div>
<pre><code>##  [1] &quot;BF&quot;               &quot;carp&quot;             &quot;carp_heat&quot;        &quot;col&quot;              &quot;Conventions&quot;     
##  [6] &quot;csv_file&quot;         &quot;csv_name&quot;         &quot;csv_path&quot;         &quot;csvfile&quot;          &quot;csvname&quot;         
## [11] &quot;csvpath&quot;          &quot;cutpts&quot;           &quot;datasource&quot;       &quot;dlname&quot;           &quot;dname&quot;           
## [16] &quot;dunits&quot;           &quot;fillvalue&quot;        &quot;gf&quot;               &quot;grid&quot;             &quot;h1&quot;              
## [21] &quot;h2&quot;               &quot;hdf_file&quot;         &quot;hdf_name&quot;         &quot;hdf_path&quot;         &quot;history&quot;         
## [26] &quot;institution&quot;      &quot;lat&quot;              &quot;lat_array&quot;        &quot;lon&quot;              &quot;lon_array&quot;       
## [31] &quot;lonlat&quot;           &quot;m&quot;                &quot;mapTheme&quot;         &quot;mb2&quot;              &quot;megabiome_name&quot;  
## [36] &quot;n&quot;                &quot;napol&quot;            &quot;ncfname&quot;          &quot;ncin&quot;             &quot;ncname&quot;          
## [41] &quot;ncpath&quot;           &quot;nlat&quot;             &quot;nlon&quot;             &quot;nm&quot;               &quot;nt&quot;              
## [46] &quot;ny&quot;               &quot;orstationc&quot;       &quot;orstationc_file&quot;  &quot;orstationc_name&quot;  &quot;orstationc_path&quot; 
## [51] &quot;plt&quot;              &quot;pngfile&quot;          &quot;references&quot;       &quot;shp_file&quot;         &quot;shp_name&quot;        
## [56] &quot;shp_path&quot;         &quot;sort_cols&quot;        &quot;stcarp&quot;           &quot;tas_anm_ann&quot;      &quot;tas2&quot;            
## [61] &quot;tcarp&quot;            &quot;tday&quot;             &quot;tdstr&quot;            &quot;time&quot;             &quot;title&quot;           
## [66] &quot;tmonth&quot;           &quot;tmp_array&quot;        &quot;tmp_df01&quot;         &quot;tmp_df02&quot;         &quot;tmp_df03&quot;        
## [71] &quot;tmp_mat&quot;          &quot;tmp_slice&quot;        &quot;tmp_vec&quot;          &quot;tmp_vec_long&quot;     &quot;tr21dec_file&quot;    
## [76] &quot;tr21dec_name&quot;     &quot;tr21dec_path&quot;     &quot;tunits&quot;           &quot;tustr&quot;            &quot;tyear&quot;           
## [81] &quot;UVI&quot;              &quot;vegtype_name&quot;     &quot;world_outline&quot;    &quot;world_outline_gg&quot; &quot;world_shp&quot;</code></pre>
<p><a href="netCDF.html">[Back to top]</a></p>
<p><a href="netCDF.html">[Back to top]</a></p>
</div>
</div>
</div>
<div id="data-frame-to-array-conversionrectangular-to-raster"
class="section level1" number="4">
<h1><span class="header-section-number">4</span> Data frame-to-array
conversion(rectangular to raster)</h1>
<p>In this set of example code, an R data frame is reshaped into an
array that can be written out as a netCDF file. This could be a trivial
transformation, if all rows and columns of the target array are
contained in the data frame. In many real-world cases, however, the data
frame contains, for example, only land data points, and so transforming
or reshaping the data frame to an array is not straighforward, because
the data frame contains only a subset of points in the full array.</p>
<p>There are several approaches for doing the reshaping, ranging from
explict and slow, to rather cryptic, but fast. The individual approaches
below can be timed using the <code>proc.time()</code> function:</p>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" tabindex="-1"></a><span class="co"># time an R process</span></span>
<span id="cb58-2"><a href="#cb58-2" tabindex="-1"></a>ptm <span class="ot">&lt;-</span> <span class="fu">proc.time</span>() <span class="co"># start the timer</span></span>
<span id="cb58-3"><a href="#cb58-3" tabindex="-1"></a><span class="co"># ... some code ...</span></span>
<span id="cb58-4"><a href="#cb58-4" tabindex="-1"></a><span class="fu">proc.time</span>() <span class="sc">-</span> ptm <span class="co"># how long?</span></span></code></pre></div>
<p>Specific times will vary, of course, from machine to machine.</p>
<div id="convert-a-full-r-data-frame-to-an-array" class="section level2"
number="4.1">
<h2><span class="header-section-number">4.1</span> Convert a “full” R
data frame to an array</h2>
<p>In this first example, a “full” data frame (e.g. one with
<em>nlon</em> x <em>nlat</em> rows, and <em>nt</em> columns) is reshaped
into a 3-D <em>nlon</em>`* by <em>nlat</em> by <em>nt</em> array. (The
example also illustrates the conversion of a <em>nlon</em> x
<em>nlat</em> rows by 1-column variable in a data frame into a 2-D
<em>nlon</em> by <em>nlat</em> array.)</p>
<div id="initial-set-up-create-dimension-variables"
class="section level3" number="4.1.1">
<h3><span class="header-section-number">4.1.1</span> Initial set up –
create dimension variables</h3>
<p>The first step is to create the dimension variables for the “full”
array; in this example, the longitudes (<code>lon</code>), latitudes
(<code>lat</code>) and time (<code>t</code>) variables. These variables
should be defined for the “full” array, and not just for the
observations in the data frame. One approach is to simply copy those
values from an “original” netCDF data set.</p>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" tabindex="-1"></a><span class="co"># copy lon, lat and time from the initial netCDF data set</span></span>
<span id="cb59-2"><a href="#cb59-2" tabindex="-1"></a>lon2 <span class="ot">&lt;-</span> lon</span>
<span id="cb59-3"><a href="#cb59-3" tabindex="-1"></a>lat2 <span class="ot">&lt;-</span> lat</span>
<span id="cb59-4"><a href="#cb59-4" tabindex="-1"></a>time2 <span class="ot">&lt;-</span> time</span>
<span id="cb59-5"><a href="#cb59-5" tabindex="-1"></a>tunits2 <span class="ot">&lt;-</span> tunits</span>
<span id="cb59-6"><a href="#cb59-6" tabindex="-1"></a>nlon2 <span class="ot">&lt;-</span> nlon; nlat2 <span class="ot">&lt;-</span> nlat; nt2 <span class="ot">&lt;-</span> nt</span></code></pre></div>
<p>Another approach is to generate or specify the dimension variables
explicitly. However, this may be problematical if the source file
longitudes and latitudes were not generated in exactly the same way, or
were saved at lower (single) precision.</p>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" tabindex="-1"></a><span class="co"># generate lons, lats and set time</span></span>
<span id="cb60-2"><a href="#cb60-2" tabindex="-1"></a>lon2 <span class="ot">&lt;-</span> <span class="fu">as.array</span>(<span class="fu">seq</span>(<span class="sc">-</span><span class="fl">179.75</span>,<span class="fl">179.75</span>,<span class="fl">0.5</span>))</span>
<span id="cb60-3"><a href="#cb60-3" tabindex="-1"></a>nlon2 <span class="ot">&lt;-</span> <span class="dv">720</span></span>
<span id="cb60-4"><a href="#cb60-4" tabindex="-1"></a>lat2 <span class="ot">&lt;-</span> <span class="fu">as.array</span>(<span class="fu">seq</span>(<span class="sc">-</span><span class="fl">89.75</span>,<span class="fl">89.75</span>,<span class="fl">0.5</span>))</span>
<span id="cb60-5"><a href="#cb60-5" tabindex="-1"></a>nlat2 <span class="ot">&lt;-</span> <span class="dv">360</span></span>
<span id="cb60-6"><a href="#cb60-6" tabindex="-1"></a>time2 <span class="ot">&lt;-</span><span class="fu">as.array</span>(<span class="fu">c</span>(<span class="fl">27773.5</span>, <span class="fl">27803.5</span>, <span class="fl">27833.5</span>, <span class="fl">27864.0</span>, <span class="fl">27894.5</span>, <span class="fl">27925.0</span>,</span>
<span id="cb60-7"><a href="#cb60-7" tabindex="-1"></a>       <span class="fl">27955.5</span>, <span class="fl">27986.5</span>, <span class="fl">28017.0</span>, <span class="fl">28047.5</span>, <span class="fl">28078.0</span>, <span class="fl">28108.5</span>))</span>
<span id="cb60-8"><a href="#cb60-8" tabindex="-1"></a>nt2 <span class="ot">&lt;-</span> <span class="dv">12</span></span>
<span id="cb60-9"><a href="#cb60-9" tabindex="-1"></a>tunits2 <span class="ot">&lt;-</span> <span class="st">&quot;days since 1900-01-01 00:00:00.0 -0:00&quot;</span></span></code></pre></div>
</div>
<div id="reshaping-a-full-data-frame-to-an-array" class="section level3"
number="4.1.2">
<h3><span class="header-section-number">4.1.2</span> Reshaping a “full”
data frame to an array</h3>
<p>In this example, the <code>tmp_df02</code> data frame, which contains
259200 rows and 17 columns (with missing values over the oceans), is
transformed into a <em>nlon</em> by <em>nlat</em> by <em>nt</em> array.
In the new array, <code>lon</code> varies most rapidly, <code>lat</code>
next, and <code>t</code> least rapidly, in a fashion consistent with the
“CF-1.6” conventions for netCDF files. The size (and shapes) of the
various arrays are confirmed by repeated applications of the
<code>dim()</code> function (recalling that <code>dim()</code> will list
the number of columns first, number of rows second (and if approriate,
the number of times third)). The conversion is done in two steps: 1)
converting that part of the the data frame containing the 12 monthly
values into into a 2-d matrix, and then 2) reshaping the 2-d matrix into
a 3-d array.</p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" tabindex="-1"></a>ptm <span class="ot">&lt;-</span> <span class="fu">proc.time</span>() <span class="co"># start the timer</span></span>
<span id="cb61-2"><a href="#cb61-2" tabindex="-1"></a><span class="co"># convert tmp_df02 back into an array</span></span>
<span id="cb61-3"><a href="#cb61-3" tabindex="-1"></a>tmp_mat2 <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(tmp_df02[<span class="dv">3</span><span class="sc">:</span>(<span class="dv">3</span><span class="sc">+</span>nt<span class="dv">-1</span>)])</span>
<span id="cb61-4"><a href="#cb61-4" tabindex="-1"></a><span class="fu">dim</span>(tmp_mat2)</span></code></pre></div>
<pre><code>## [1] 259200     12</code></pre>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" tabindex="-1"></a><span class="co"># then reshape the array</span></span>
<span id="cb63-2"><a href="#cb63-2" tabindex="-1"></a>tmp_array2 <span class="ot">&lt;-</span> <span class="fu">array</span>(tmp_mat2, <span class="at">dim=</span><span class="fu">c</span>(nlon2,nlat2,nt))</span>
<span id="cb63-3"><a href="#cb63-3" tabindex="-1"></a><span class="fu">dim</span>(tmp_array2)</span></code></pre></div>
<pre><code>## [1] 720 360  12</code></pre>
<p>The columns containing <code>mtwa</code>, <code>mtco</code> and
<code>mat</code> are each transformed into 2-D arrays.</p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" tabindex="-1"></a><span class="co"># convert mtwa, mtco and mat to arrays</span></span>
<span id="cb65-2"><a href="#cb65-2" tabindex="-1"></a>mtwa_array2 <span class="ot">&lt;-</span> <span class="fu">array</span>(tmp_df02<span class="sc">$</span>mtwa, <span class="at">dim=</span><span class="fu">c</span>(nlon2,nlat2))</span>
<span id="cb65-3"><a href="#cb65-3" tabindex="-1"></a><span class="fu">dim</span>(mtwa_array2)</span></code></pre></div>
<pre><code>## [1] 720 360</code></pre>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" tabindex="-1"></a>mtco_array2 <span class="ot">&lt;-</span> <span class="fu">array</span>(tmp_df02<span class="sc">$</span>mtco, <span class="at">dim=</span><span class="fu">c</span>(nlon2,nlat2))</span>
<span id="cb67-2"><a href="#cb67-2" tabindex="-1"></a><span class="fu">dim</span>(mtco_array2)</span></code></pre></div>
<pre><code>## [1] 720 360</code></pre>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" tabindex="-1"></a>mat_array2 <span class="ot">&lt;-</span> <span class="fu">array</span>(tmp_df02<span class="sc">$</span>mat, <span class="at">dim=</span><span class="fu">c</span>(nlon2,nlat2))</span>
<span id="cb69-2"><a href="#cb69-2" tabindex="-1"></a><span class="fu">dim</span>(mat_array2)</span></code></pre></div>
<pre><code>## [1] 720 360</code></pre>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" tabindex="-1"></a><span class="fu">proc.time</span>() <span class="sc">-</span> ptm <span class="co"># how long?</span></span></code></pre></div>
<pre><code>##    user  system elapsed 
##   0.060   0.026   0.088</code></pre>
</div>
<div id="check-the-conversion" class="section level3" number="4.1.3">
<h3><span class="header-section-number">4.1.3</span> Check the
conversion</h3>
<p>It’s generally a good idea to plot (map) the resulting arrays to
check for anomalies or misapprehensions about the layout of the data.
First plot the January values, then MTWA, MTCO and MAT.</p>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" tabindex="-1"></a><span class="co"># some plots to check creation of arrays</span></span>
<span id="cb73-2"><a href="#cb73-2" tabindex="-1"></a><span class="fu">library</span>(lattice)</span>
<span id="cb73-3"><a href="#cb73-3" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb73-4"><a href="#cb73-4" tabindex="-1"></a></span>
<span id="cb73-5"><a href="#cb73-5" tabindex="-1"></a><span class="fu">levelplot</span>(tmp_array2[,,<span class="dv">1</span>] <span class="sc">~</span> lon <span class="sc">*</span> lat, <span class="at">data=</span>grid, <span class="at">at=</span>cutpts, <span class="at">cuts=</span><span class="dv">11</span>, <span class="at">pretty=</span>T, </span>
<span id="cb73-6"><a href="#cb73-6" tabindex="-1"></a>  <span class="at">col.regions=</span>(<span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>,<span class="st">&quot;RdBu&quot;</span>))), <span class="at">main=</span><span class="st">&quot;Mean July Temperature (C)&quot;</span>)</span>
<span id="cb73-7"><a href="#cb73-7" tabindex="-1"></a><span class="fu">levelplot</span>(mtwa_array2 <span class="sc">~</span> lon <span class="sc">*</span> lat, <span class="at">data=</span>grid, <span class="at">at=</span>cutpts, <span class="at">cuts=</span><span class="dv">11</span>, <span class="at">pretty=</span>T, </span>
<span id="cb73-8"><a href="#cb73-8" tabindex="-1"></a>  <span class="at">col.regions=</span>(<span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>,<span class="st">&quot;RdBu&quot;</span>))), <span class="at">main=</span><span class="st">&quot;MTWA (C)&quot;</span>)</span>
<span id="cb73-9"><a href="#cb73-9" tabindex="-1"></a><span class="fu">levelplot</span>(mtco_array2 <span class="sc">~</span> lon <span class="sc">*</span> lat, <span class="at">data=</span>grid, <span class="at">at=</span>cutpts, <span class="at">cuts=</span><span class="dv">11</span>, <span class="at">pretty=</span>T, </span>
<span id="cb73-10"><a href="#cb73-10" tabindex="-1"></a>  <span class="at">col.regions=</span>(<span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>,<span class="st">&quot;RdBu&quot;</span>))), <span class="at">main=</span><span class="st">&quot;MTCO (C)&quot;</span>)</span>
<span id="cb73-11"><a href="#cb73-11" tabindex="-1"></a><span class="fu">levelplot</span>(mat_array2 <span class="sc">~</span> lon <span class="sc">*</span> lat, <span class="at">data=</span>grid, <span class="at">at=</span>cutpts, <span class="at">cuts=</span><span class="dv">11</span>, <span class="at">pretty=</span>T, </span>
<span id="cb73-12"><a href="#cb73-12" tabindex="-1"></a>  <span class="at">col.regions=</span>(<span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>,<span class="st">&quot;RdBu&quot;</span>))), <span class="at">main=</span><span class="st">&quot;MAT (C)&quot;</span>)</span></code></pre></div>
<p><img src="netCDF_files/figure-html/array06-1.png" width="75%" height="75%" /><img src="netCDF_files/figure-html/array06-2.png" width="75%" height="75%" /><img src="netCDF_files/figure-html/array06-3.png" width="75%" height="75%" /><img src="netCDF_files/figure-html/array06-4.png" width="75%" height="75%" /></p>
<p>Looks ok.</p>
</div>
</div>
<div id="convert-a-short-r-data-frame-to-an-array"
class="section level2" number="4.2">
<h2><span class="header-section-number">4.2</span> Convert a “short” R
data frame to an array</h2>
<p>In this second example, a “short” data frame, containing, for
example, data only for land grid points, is converted to a “full” array.
Unlike the conversion of a “full” data frame, this can’t be accomplished
by simple conversion and reshaping, but instead the rows in the “short”
data frame have to be assigned to the specific locations. This can be
done explicity, by looping over the individual rows of the data frame,
and copying the values from each row into the appropriate locations of
the array. This can be very slow, but it has the advantage of being
explict. “Loop-avoidance” approaches are much faster, but can be rather
cryptic, and depend on the data frame and “target” arrays being properly
structured. In this example, the “short” data frame
<code>tmp_df03</code> is moved into a 3-D array <code>tmp_array3</code>
using three different approaches.</p>
<div id="initial-set-up" class="section level3" number="4.2.1">
<h3><span class="header-section-number">4.2.1</span> Initial set up</h3>
<p>As before, dimension variables are generated or copied:</p>
<div class="sourceCode" id="cb74"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" tabindex="-1"></a><span class="co"># generate lons, lats and set time</span></span>
<span id="cb74-2"><a href="#cb74-2" tabindex="-1"></a>lon3 <span class="ot">&lt;-</span> <span class="fu">as.array</span>(<span class="fu">seq</span>(<span class="sc">-</span><span class="fl">179.750</span>,<span class="fl">179.750</span>,<span class="fl">0.50</span>))</span>
<span id="cb74-3"><a href="#cb74-3" tabindex="-1"></a>nlon3 <span class="ot">&lt;-</span> <span class="dv">720</span></span>
<span id="cb74-4"><a href="#cb74-4" tabindex="-1"></a>lat3 <span class="ot">&lt;-</span> <span class="fu">as.array</span>(<span class="fu">seq</span>(<span class="sc">-</span><span class="fl">89.750</span>,<span class="fl">89.750</span>,<span class="fl">0.50</span>))</span>
<span id="cb74-5"><a href="#cb74-5" tabindex="-1"></a>nlat3 <span class="ot">&lt;-</span> <span class="dv">360</span></span>
<span id="cb74-6"><a href="#cb74-6" tabindex="-1"></a>time3 <span class="ot">&lt;-</span> <span class="fu">as.array</span>(<span class="fu">c</span>(<span class="fl">27773.5</span>, <span class="fl">27803.5</span>, <span class="fl">27833.5</span>, <span class="fl">27864.0</span>, <span class="fl">27894.5</span>, <span class="fl">27925.0</span>,</span>
<span id="cb74-7"><a href="#cb74-7" tabindex="-1"></a>       <span class="fl">27955.5</span>, <span class="fl">27986.5</span>, <span class="fl">28017.0</span>, <span class="fl">28047.5</span>, <span class="fl">28078.0</span>, <span class="fl">28108.5</span>))</span>
<span id="cb74-8"><a href="#cb74-8" tabindex="-1"></a>nt3 <span class="ot">&lt;-</span> <span class="dv">12</span></span>
<span id="cb74-9"><a href="#cb74-9" tabindex="-1"></a>tunits3 <span class="ot">&lt;-</span> <span class="st">&quot;days since 1900-01-01 00:00:00.0 -0:00&quot;</span></span></code></pre></div>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" tabindex="-1"></a><span class="co"># copy lon, lat and time from initial netCDF data set</span></span>
<span id="cb75-2"><a href="#cb75-2" tabindex="-1"></a>lon4 <span class="ot">&lt;-</span> lon</span>
<span id="cb75-3"><a href="#cb75-3" tabindex="-1"></a>lat4 <span class="ot">&lt;-</span> lat</span>
<span id="cb75-4"><a href="#cb75-4" tabindex="-1"></a>time4 <span class="ot">&lt;-</span> time</span>
<span id="cb75-5"><a href="#cb75-5" tabindex="-1"></a>tunits4 <span class="ot">&lt;-</span> tunits</span>
<span id="cb75-6"><a href="#cb75-6" tabindex="-1"></a>nlon4 <span class="ot">&lt;-</span> nlon; nlat4 <span class="ot">&lt;-</span> nlat; nt4 <span class="ot">&lt;-</span> nt</span></code></pre></div>
<p>Next, an <em>nlon</em> by <em>nlat</em> by <em>nt</em> array is
created, and filled with the original fill value (or an alternative).
Also, three <em>nlon</em> by <em>nlat</em> arrays for <code>MTWA</code>,
<code>MTCO</code>, and <code>MAT</code> are created and filled. The
generated lontitudes and latitudes are used here (as opposed to copies
from the original netCDF file–this is more general)</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" tabindex="-1"></a><span class="co"># create arrays</span></span>
<span id="cb76-2"><a href="#cb76-2" tabindex="-1"></a><span class="co"># nlon * nlat * nt array</span></span>
<span id="cb76-3"><a href="#cb76-3" tabindex="-1"></a>fillvalue <span class="ot">&lt;-</span> <span class="fl">1e32</span></span>
<span id="cb76-4"><a href="#cb76-4" tabindex="-1"></a>tmp_array3 <span class="ot">&lt;-</span> <span class="fu">array</span>(fillvalue, <span class="at">dim=</span><span class="fu">c</span>(nlon3,nlat3,nt3))</span>
<span id="cb76-5"><a href="#cb76-5" tabindex="-1"></a><span class="co"># nlon * nlat arrays for mtwa, mtco and mat</span></span>
<span id="cb76-6"><a href="#cb76-6" tabindex="-1"></a>mtwa_array3 <span class="ot">&lt;-</span> <span class="fu">array</span>(fillvalue, <span class="at">dim=</span><span class="fu">c</span>(nlon3,nlat3))</span>
<span id="cb76-7"><a href="#cb76-7" tabindex="-1"></a>mtco_array3 <span class="ot">&lt;-</span> <span class="fu">array</span>(fillvalue, <span class="at">dim=</span><span class="fu">c</span>(nlon3,nlat3))</span>
<span id="cb76-8"><a href="#cb76-8" tabindex="-1"></a>mat_array3 <span class="ot">&lt;-</span> <span class="fu">array</span>(fillvalue, <span class="at">dim=</span><span class="fu">c</span>(nlon3,nlat3))</span></code></pre></div>
</div>
<div id="explicit-copying-from-a-data-frame-to-array"
class="section level3" number="4.2.2">
<h3><span class="header-section-number">4.2.2</span> Explicit copying
from a data frame to array</h3>
<p>In the first, most explict, approach, we loop over the rows in the
data frame, find the <em>j</em>-th and <em>k</em>-th column and row that
each observation falls in (using the <code>which.min()</code> function),
and then copy the values for each row into the arrays. This takes a
relatively long time for data sets with hundreds of rows and
columns.</p>
<div class="sourceCode" id="cb77"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" tabindex="-1"></a><span class="co"># loop over the rows in the data frame </span></span>
<span id="cb77-2"><a href="#cb77-2" tabindex="-1"></a><span class="co"># most explicit, but takes a VERY LONG TIME</span></span>
<span id="cb77-3"><a href="#cb77-3" tabindex="-1"></a>ptm <span class="ot">&lt;-</span> <span class="fu">proc.time</span>() <span class="co"># time the loop</span></span>
<span id="cb77-4"><a href="#cb77-4" tabindex="-1"></a>nobs <span class="ot">&lt;-</span> <span class="fu">dim</span>(tmp_df03)[<span class="dv">1</span>]</span>
<span id="cb77-5"><a href="#cb77-5" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>nobs) {</span>
<span id="cb77-6"><a href="#cb77-6" tabindex="-1"></a>  </span>
<span id="cb77-7"><a href="#cb77-7" tabindex="-1"></a>  <span class="co"># figure out location in the target array of the values in each row of the data frame</span></span>
<span id="cb77-8"><a href="#cb77-8" tabindex="-1"></a>  j <span class="ot">&lt;-</span> <span class="fu">which.min</span>(<span class="fu">abs</span>(lon3<span class="sc">-</span>tmp_df03<span class="sc">$</span>lon[i]))</span>
<span id="cb77-9"><a href="#cb77-9" tabindex="-1"></a>  k <span class="ot">&lt;-</span> <span class="fu">which.min</span>(<span class="fu">abs</span>(lat3<span class="sc">-</span>tmp_df03<span class="sc">$</span>lat[i]))</span>
<span id="cb77-10"><a href="#cb77-10" tabindex="-1"></a>  </span>
<span id="cb77-11"><a href="#cb77-11" tabindex="-1"></a>  <span class="co"># copy data from the data frame to array</span></span>
<span id="cb77-12"><a href="#cb77-12" tabindex="-1"></a>  tmp_array3[j,k,<span class="dv">1</span><span class="sc">:</span>nt] <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(tmp_df03[i,<span class="dv">3</span><span class="sc">:</span>(nt<span class="sc">+</span><span class="dv">2</span>)])</span>
<span id="cb77-13"><a href="#cb77-13" tabindex="-1"></a>  mtwa_array3[j,k] <span class="ot">&lt;-</span> tmp_df03<span class="sc">$</span>mtwa[i]</span>
<span id="cb77-14"><a href="#cb77-14" tabindex="-1"></a>  mtco_array3[j,k] <span class="ot">&lt;-</span> tmp_df03<span class="sc">$</span>mtco[i]</span>
<span id="cb77-15"><a href="#cb77-15" tabindex="-1"></a>  mat_array3[j,k] <span class="ot">&lt;-</span> tmp_df03<span class="sc">$</span>mat[i]</span>
<span id="cb77-16"><a href="#cb77-16" tabindex="-1"></a>}</span>
<span id="cb77-17"><a href="#cb77-17" tabindex="-1"></a><span class="fu">proc.time</span>() <span class="sc">-</span> ptm <span class="co"># how long?</span></span></code></pre></div>
<pre><code>##    user  system elapsed 
##   81.98   59.23  141.24</code></pre>
</div>
<div id="partial-loop-avoidance" class="section level3" number="4.2.3">
<h3><span class="header-section-number">4.2.3</span> Partial loop
avoidance</h3>
<p>In the second approach, the <code>sapply()</code> function is used to
repeatedly apply a function to create two vectors of indices
(<code>j2</code> and <code>k2</code>) that describe which column and row
of the array each row of the data frame is assigned to. For example, the
function <code>function(x) which.min(abs(lon3-x))</code> finds the
closest longitude of the full array (<code>lon3</code>) to the longitude
of each row of the data frame (tmp_df03$lon, the <code>x</code> argument
of the function).</p>
<div class="sourceCode" id="cb79"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" tabindex="-1"></a><span class="co"># loop-avoidance approaches </span></span>
<span id="cb79-2"><a href="#cb79-2" tabindex="-1"></a><span class="co"># get vectors of the grid-cell indices for each row in the data frame</span></span>
<span id="cb79-3"><a href="#cb79-3" tabindex="-1"></a>ptm <span class="ot">&lt;-</span> <span class="fu">proc.time</span>() </span>
<span id="cb79-4"><a href="#cb79-4" tabindex="-1"></a>j2 <span class="ot">&lt;-</span> <span class="fu">sapply</span>(tmp_df03<span class="sc">$</span>lon, <span class="cf">function</span>(x) <span class="fu">which.min</span>(<span class="fu">abs</span>(lon3<span class="sc">-</span>x)))</span>
<span id="cb79-5"><a href="#cb79-5" tabindex="-1"></a>k2 <span class="ot">&lt;-</span> <span class="fu">sapply</span>(tmp_df03<span class="sc">$</span>lat, <span class="cf">function</span>(x) <span class="fu">which.min</span>(<span class="fu">abs</span>(lat3<span class="sc">-</span>x)))</span></code></pre></div>
<p>Then, the values are copied (one time at a time) by first reshaping
the appropriate column in the data frame (using the
<code>as.matrix()</code> function) into a temporary array
(<code>temp_array</code>), which is then copied into
<code>tmp_array3</code> (with <code>temp</code> meaning “temporary” and
<code>tmp</code> denoting temperature here). Note how the square-bracket
selection on the left side of the assignment
(<code>[cbind(j2,k2)]</code>) puts each row of the data frame into the
proper location in the array.</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" tabindex="-1"></a>fillvalue <span class="ot">&lt;-</span> <span class="fl">1e32</span></span>
<span id="cb80-2"><a href="#cb80-2" tabindex="-1"></a><span class="co"># partial loop avoidance for tmp_array3</span></span>
<span id="cb80-3"><a href="#cb80-3" tabindex="-1"></a>temp_array <span class="ot">&lt;-</span> <span class="fu">array</span>(fillvalue, <span class="at">dim=</span><span class="fu">c</span>(nlon3,nlat3))</span>
<span id="cb80-4"><a href="#cb80-4" tabindex="-1"></a>nobs <span class="ot">&lt;-</span> <span class="fu">dim</span>(tmp_df03)[<span class="dv">1</span>]</span>
<span id="cb80-5"><a href="#cb80-5" tabindex="-1"></a><span class="cf">for</span> (l <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>nt) {</span>
<span id="cb80-6"><a href="#cb80-6" tabindex="-1"></a>  temp_array[<span class="fu">cbind</span>(j2,k2)] <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(tmp_df03[<span class="dv">1</span><span class="sc">:</span>nobs,l<span class="sc">+</span><span class="dv">2</span>]) </span>
<span id="cb80-7"><a href="#cb80-7" tabindex="-1"></a>  tmp_array3[,,l] <span class="ot">&lt;-</span> temp_array</span>
<span id="cb80-8"><a href="#cb80-8" tabindex="-1"></a>}</span></code></pre></div>
<p>The 2-D arrays can be copied directly:</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" tabindex="-1"></a><span class="co"># copy 2-D arrays directly</span></span>
<span id="cb81-2"><a href="#cb81-2" tabindex="-1"></a>mtwa_array3[<span class="fu">cbind</span>(j2,k2)] <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(tmp_df03<span class="sc">$</span>mtwa)</span>
<span id="cb81-3"><a href="#cb81-3" tabindex="-1"></a>mtco_array3[<span class="fu">cbind</span>(j2,k2)] <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(tmp_df03<span class="sc">$</span>mtco) </span>
<span id="cb81-4"><a href="#cb81-4" tabindex="-1"></a>mat_array3[<span class="fu">cbind</span>(j2,k2)] <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(tmp_df03<span class="sc">$</span>mat) </span>
<span id="cb81-5"><a href="#cb81-5" tabindex="-1"></a><span class="fu">proc.time</span>() <span class="sc">-</span> ptm</span></code></pre></div>
<pre><code>##    user  system elapsed 
##   1.176   0.620   1.808</code></pre>
</div>
<div id="complete-loop-avoidance-approach" class="section level3"
number="4.2.4">
<h3><span class="header-section-number">4.2.4</span> Complete
loop-avoidance approach</h3>
<p>Loops can be totally avoided as follows, extending the
<code>[...]</code> selection to all three dimensions of the full array
(<code>tmp_array3</code>). Note that the code fragment
<code>3:(nt3+2)</code> implies that the data are in columns 3 through 14
in the data frame (i.e. <code>lon</code> and <code>lat</code> are in the
first two columns):</p>
<div class="sourceCode" id="cb83"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" tabindex="-1"></a><span class="co"># loop avoidance for all of the variables</span></span>
<span id="cb83-2"><a href="#cb83-2" tabindex="-1"></a>ptm <span class="ot">&lt;-</span> <span class="fu">proc.time</span>() </span>
<span id="cb83-3"><a href="#cb83-3" tabindex="-1"></a>nobs <span class="ot">&lt;-</span> <span class="fu">dim</span>(tmp_df03)[<span class="dv">1</span>]</span>
<span id="cb83-4"><a href="#cb83-4" tabindex="-1"></a>l <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span>nt3,<span class="at">each=</span>nobs)</span>
<span id="cb83-5"><a href="#cb83-5" tabindex="-1"></a>tmp_array3[<span class="fu">cbind</span>(j2,k2,l)] <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(tmp_df03[<span class="dv">1</span><span class="sc">:</span>nobs,<span class="dv">3</span><span class="sc">:</span>(nt3<span class="sc">+</span><span class="dv">2</span>)])</span>
<span id="cb83-6"><a href="#cb83-6" tabindex="-1"></a>mtwa_array3[<span class="fu">cbind</span>(j2,k2)] <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(tmp_df03<span class="sc">$</span>mtwa) </span>
<span id="cb83-7"><a href="#cb83-7" tabindex="-1"></a>mtco_array3[<span class="fu">cbind</span>(j2,k2)] <span class="ot">&lt;-</span> <span class="fu">array</span>(tmp_df03<span class="sc">$</span>mtco) </span>
<span id="cb83-8"><a href="#cb83-8" tabindex="-1"></a>mat_array3[<span class="fu">cbind</span>(j2,k2)] <span class="ot">&lt;-</span> <span class="fu">array</span>(tmp_df03<span class="sc">$</span>mat) </span>
<span id="cb83-9"><a href="#cb83-9" tabindex="-1"></a><span class="fu">proc.time</span>() <span class="sc">-</span> ptm</span></code></pre></div>
<pre><code>##    user  system elapsed 
##   0.045   0.001   0.046</code></pre>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" tabindex="-1"></a><span class="co"># some plots to check creation of arrays</span></span>
<span id="cb85-2"><a href="#cb85-2" tabindex="-1"></a><span class="fu">library</span>(lattice)</span>
<span id="cb85-3"><a href="#cb85-3" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb85-4"><a href="#cb85-4" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb85-5"><a href="#cb85-5" tabindex="-1"></a><span class="fu">levelplot</span>(tmp_array3[,,m] <span class="sc">~</span> lon <span class="sc">*</span> lat, <span class="at">data=</span>grid, <span class="at">at=</span>cutpts, <span class="at">cuts=</span><span class="dv">11</span>, <span class="at">pretty=</span>T, </span>
<span id="cb85-6"><a href="#cb85-6" tabindex="-1"></a>  <span class="at">col.regions=</span>(<span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>,<span class="st">&quot;RdBu&quot;</span>))), <span class="at">main=</span><span class="st">&quot;Mean July Temperature (C)&quot;</span>)</span>
<span id="cb85-7"><a href="#cb85-7" tabindex="-1"></a><span class="fu">levelplot</span>(mtwa_array3 <span class="sc">~</span> lon <span class="sc">*</span> lat, <span class="at">data=</span>grid, <span class="at">at=</span>cutpts, <span class="at">cuts=</span><span class="dv">11</span>, <span class="at">pretty=</span>T, </span>
<span id="cb85-8"><a href="#cb85-8" tabindex="-1"></a>  <span class="at">col.regions=</span>(<span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>,<span class="st">&quot;RdBu&quot;</span>))), <span class="at">main=</span><span class="st">&quot;MTWA (C)&quot;</span>)</span>
<span id="cb85-9"><a href="#cb85-9" tabindex="-1"></a><span class="fu">levelplot</span>(mtco_array3 <span class="sc">~</span> lon <span class="sc">*</span> lat, <span class="at">data=</span>grid, <span class="at">at=</span>cutpts, <span class="at">cuts=</span><span class="dv">11</span>, <span class="at">pretty=</span>T, </span>
<span id="cb85-10"><a href="#cb85-10" tabindex="-1"></a>  <span class="at">col.regions=</span>(<span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>,<span class="st">&quot;RdBu&quot;</span>))), <span class="at">main=</span><span class="st">&quot;MTCO (C)&quot;</span>)</span>
<span id="cb85-11"><a href="#cb85-11" tabindex="-1"></a><span class="fu">levelplot</span>(mat_array3 <span class="sc">~</span> lon <span class="sc">*</span> lat, <span class="at">data=</span>grid, <span class="at">at=</span>cutpts, <span class="at">cuts=</span><span class="dv">11</span>, <span class="at">pretty=</span>T, </span>
<span id="cb85-12"><a href="#cb85-12" tabindex="-1"></a>  <span class="at">col.regions=</span>(<span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="dv">10</span>,<span class="st">&quot;RdBu&quot;</span>))), <span class="at">main=</span><span class="st">&quot;MAT (C)&quot;</span>)</span></code></pre></div>
<p><img src="netCDF_files/figure-html/array16-1.png" width="75%" height="75%" /><img src="netCDF_files/figure-html/array16-2.png" width="75%" height="75%" /><img src="netCDF_files/figure-html/array16-3.png" width="75%" height="75%" /><img src="netCDF_files/figure-html/array16-4.png" width="75%" height="75%" /></p>
<p>Check what’s in the current workspace:</p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" tabindex="-1"></a><span class="fu">ls</span>()</span></code></pre></div>
<pre><code>##   [1] &quot;BF&quot;               &quot;carp&quot;             &quot;carp_heat&quot;        &quot;col&quot;              &quot;Conventions&quot;     
##   [6] &quot;csv_file&quot;         &quot;csv_name&quot;         &quot;csv_path&quot;         &quot;csvfile&quot;          &quot;csvname&quot;         
##  [11] &quot;csvpath&quot;          &quot;cutpts&quot;           &quot;datasource&quot;       &quot;dlname&quot;           &quot;dname&quot;           
##  [16] &quot;dunits&quot;           &quot;fillvalue&quot;        &quot;gf&quot;               &quot;grid&quot;             &quot;h1&quot;              
##  [21] &quot;h2&quot;               &quot;hdf_file&quot;         &quot;hdf_name&quot;         &quot;hdf_path&quot;         &quot;history&quot;         
##  [26] &quot;institution&quot;      &quot;j2&quot;               &quot;k2&quot;               &quot;l&quot;                &quot;lat&quot;             
##  [31] &quot;lat_array&quot;        &quot;lat2&quot;             &quot;lat3&quot;             &quot;lat4&quot;             &quot;lon&quot;             
##  [36] &quot;lon_array&quot;        &quot;lon2&quot;             &quot;lon3&quot;             &quot;lon4&quot;             &quot;lonlat&quot;          
##  [41] &quot;m&quot;                &quot;mapTheme&quot;         &quot;mat_array2&quot;       &quot;mat_array3&quot;       &quot;mb2&quot;             
##  [46] &quot;megabiome_name&quot;   &quot;mtco_array2&quot;      &quot;mtco_array3&quot;      &quot;mtwa_array2&quot;      &quot;mtwa_array3&quot;     
##  [51] &quot;n&quot;                &quot;napol&quot;            &quot;ncfname&quot;          &quot;ncin&quot;             &quot;ncname&quot;          
##  [56] &quot;ncpath&quot;           &quot;nlat&quot;             &quot;nlat2&quot;            &quot;nlat3&quot;            &quot;nlat4&quot;           
##  [61] &quot;nlon&quot;             &quot;nlon2&quot;            &quot;nlon3&quot;            &quot;nlon4&quot;            &quot;nm&quot;              
##  [66] &quot;nobs&quot;             &quot;nt&quot;               &quot;nt2&quot;              &quot;nt3&quot;              &quot;nt4&quot;             
##  [71] &quot;ny&quot;               &quot;orstationc&quot;       &quot;orstationc_file&quot;  &quot;orstationc_name&quot;  &quot;orstationc_path&quot; 
##  [76] &quot;plt&quot;              &quot;pngfile&quot;          &quot;ptm&quot;              &quot;references&quot;       &quot;shp_file&quot;        
##  [81] &quot;shp_name&quot;         &quot;shp_path&quot;         &quot;sort_cols&quot;        &quot;stcarp&quot;           &quot;tas_anm_ann&quot;     
##  [86] &quot;tas2&quot;             &quot;tcarp&quot;            &quot;tday&quot;             &quot;tdstr&quot;            &quot;temp_array&quot;      
##  [91] &quot;time&quot;             &quot;time2&quot;            &quot;time3&quot;            &quot;time4&quot;            &quot;title&quot;           
##  [96] &quot;tmonth&quot;           &quot;tmp_array&quot;        &quot;tmp_array2&quot;       &quot;tmp_array3&quot;       &quot;tmp_df01&quot;        
## [101] &quot;tmp_df02&quot;         &quot;tmp_df03&quot;         &quot;tmp_mat&quot;          &quot;tmp_mat2&quot;         &quot;tmp_slice&quot;       
## [106] &quot;tmp_vec&quot;          &quot;tmp_vec_long&quot;     &quot;tr21dec_file&quot;     &quot;tr21dec_name&quot;     &quot;tr21dec_path&quot;    
## [111] &quot;tunits&quot;           &quot;tunits2&quot;          &quot;tunits3&quot;          &quot;tunits4&quot;          &quot;tustr&quot;           
## [116] &quot;tyear&quot;            &quot;UVI&quot;              &quot;vegtype_name&quot;     &quot;world_outline&quot;    &quot;world_outline_gg&quot;
## [121] &quot;world_shp&quot;</code></pre>
<p><a href="netCDF.html">[Back to top]</a></p>
</div>
</div>
</div>
<div id="create-and-write-a-netcdf-file" class="section level1"
number="5">
<h1><span class="header-section-number">5</span> Create and write a
netCDF file</h1>
<p>In this example, the arrays created above are written out using the
<code>ncdf4</code> package. Creating and writing (new) netCDF files
involves first defining or “laying out” the dimensions and coordiate
variables and the individual variables, and the attrributes of each, and
then creating the file and “putting” the data into the file, along with
additional attributes or metadata.</p>
<p>First, create the netCDF filename:</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" tabindex="-1"></a><span class="co"># path and file name, set dname</span></span>
<span id="cb88-2"><a href="#cb88-2" tabindex="-1"></a>ncpath <span class="ot">&lt;-</span> <span class="st">&quot;/Users/bartlein/Projects/ESSD/data/nc_files/&quot;</span></span>
<span id="cb88-3"><a href="#cb88-3" tabindex="-1"></a>ncname <span class="ot">&lt;-</span> <span class="st">&quot;cru10min30_ncdf4&quot;</span>  </span>
<span id="cb88-4"><a href="#cb88-4" tabindex="-1"></a>ncfname <span class="ot">&lt;-</span> <span class="fu">paste</span>(ncpath, ncname, <span class="st">&quot;.nc&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb88-5"><a href="#cb88-5" tabindex="-1"></a>dname <span class="ot">&lt;-</span> <span class="st">&quot;tmp&quot;</span>  <span class="co"># note: tmp means temperature (not temporary)</span></span></code></pre></div>
<p>Then define the contents of the file:</p>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" tabindex="-1"></a><span class="co"># create and write the netCDF file -- ncdf4 version</span></span>
<span id="cb89-2"><a href="#cb89-2" tabindex="-1"></a><span class="co"># define dimensions</span></span>
<span id="cb89-3"><a href="#cb89-3" tabindex="-1"></a>londim <span class="ot">&lt;-</span> <span class="fu">ncdim_def</span>(<span class="st">&quot;lon&quot;</span>,<span class="st">&quot;degrees_east&quot;</span>,<span class="fu">as.double</span>(lon3)) </span>
<span id="cb89-4"><a href="#cb89-4" tabindex="-1"></a>latdim <span class="ot">&lt;-</span> <span class="fu">ncdim_def</span>(<span class="st">&quot;lat&quot;</span>,<span class="st">&quot;degrees_north&quot;</span>,<span class="fu">as.double</span>(lat3)) </span>
<span id="cb89-5"><a href="#cb89-5" tabindex="-1"></a>timedim <span class="ot">&lt;-</span> <span class="fu">ncdim_def</span>(<span class="st">&quot;time&quot;</span>,tunits3,<span class="fu">as.double</span>(time3))</span>
<span id="cb89-6"><a href="#cb89-6" tabindex="-1"></a></span>
<span id="cb89-7"><a href="#cb89-7" tabindex="-1"></a><span class="co"># define variables</span></span>
<span id="cb89-8"><a href="#cb89-8" tabindex="-1"></a>fillvalue <span class="ot">&lt;-</span> <span class="fl">1e32</span></span>
<span id="cb89-9"><a href="#cb89-9" tabindex="-1"></a>dlname <span class="ot">&lt;-</span> <span class="st">&quot;2m air temperature&quot;</span></span>
<span id="cb89-10"><a href="#cb89-10" tabindex="-1"></a>tmp_def <span class="ot">&lt;-</span> <span class="fu">ncvar_def</span>(<span class="st">&quot;tmp&quot;</span>,<span class="st">&quot;deg_C&quot;</span>,<span class="fu">list</span>(londim,latdim,timedim),fillvalue,dlname,<span class="at">prec=</span><span class="st">&quot;single&quot;</span>)</span>
<span id="cb89-11"><a href="#cb89-11" tabindex="-1"></a>dlname <span class="ot">&lt;-</span> <span class="st">&quot;mean_temperture_warmest_month&quot;</span></span>
<span id="cb89-12"><a href="#cb89-12" tabindex="-1"></a>mtwa.def <span class="ot">&lt;-</span> <span class="fu">ncvar_def</span>(<span class="st">&quot;mtwa&quot;</span>,<span class="st">&quot;deg_C&quot;</span>,<span class="fu">list</span>(londim,latdim),fillvalue,dlname,<span class="at">prec=</span><span class="st">&quot;single&quot;</span>)</span>
<span id="cb89-13"><a href="#cb89-13" tabindex="-1"></a>dlname <span class="ot">&lt;-</span> <span class="st">&quot;mean_temperature_coldest_month&quot;</span></span>
<span id="cb89-14"><a href="#cb89-14" tabindex="-1"></a>mtco.def <span class="ot">&lt;-</span> <span class="fu">ncvar_def</span>(<span class="st">&quot;mtco&quot;</span>,<span class="st">&quot;deg_C&quot;</span>,<span class="fu">list</span>(londim,latdim),fillvalue,dlname,<span class="at">prec=</span><span class="st">&quot;single&quot;</span>)</span>
<span id="cb89-15"><a href="#cb89-15" tabindex="-1"></a>dlname <span class="ot">&lt;-</span> <span class="st">&quot;mean_annual_temperature&quot;</span></span>
<span id="cb89-16"><a href="#cb89-16" tabindex="-1"></a>mat.def <span class="ot">&lt;-</span> <span class="fu">ncvar_def</span>(<span class="st">&quot;mat&quot;</span>,<span class="st">&quot;deg_C&quot;</span>,<span class="fu">list</span>(londim,latdim),fillvalue,dlname,<span class="at">prec=</span><span class="st">&quot;single&quot;</span>)</span></code></pre></div>
<p>Next, create the file, and put the variables into it, along with
additional variable and “global” attributes (those that apply to the
whole file). Note that the attributes are of key importance to the
self-documenting properties of netCDF files.</p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" tabindex="-1"></a><span class="co"># create netCDF file and put arrays</span></span>
<span id="cb90-2"><a href="#cb90-2" tabindex="-1"></a>ncout <span class="ot">&lt;-</span> <span class="fu">nc_create</span>(ncfname,<span class="fu">list</span>(tmp_def,mtco.def,mtwa.def,mat.def),<span class="at">force_v4=</span><span class="cn">TRUE</span>)</span>
<span id="cb90-3"><a href="#cb90-3" tabindex="-1"></a></span>
<span id="cb90-4"><a href="#cb90-4" tabindex="-1"></a><span class="co"># put variables</span></span>
<span id="cb90-5"><a href="#cb90-5" tabindex="-1"></a><span class="fu">ncvar_put</span>(ncout,tmp_def,tmp_array3)</span>
<span id="cb90-6"><a href="#cb90-6" tabindex="-1"></a><span class="fu">ncvar_put</span>(ncout,mtwa.def,mtwa_array3)</span>
<span id="cb90-7"><a href="#cb90-7" tabindex="-1"></a><span class="fu">ncvar_put</span>(ncout,mtco.def,mtco_array3)</span>
<span id="cb90-8"><a href="#cb90-8" tabindex="-1"></a><span class="fu">ncvar_put</span>(ncout,mat.def,mat_array3)</span>
<span id="cb90-9"><a href="#cb90-9" tabindex="-1"></a></span>
<span id="cb90-10"><a href="#cb90-10" tabindex="-1"></a><span class="co"># put additional attributes into dimension and data variables</span></span>
<span id="cb90-11"><a href="#cb90-11" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;lon&quot;</span>,<span class="st">&quot;axis&quot;</span>,<span class="st">&quot;X&quot;</span>) <span class="co">#,verbose=FALSE) #,definemode=FALSE)</span></span>
<span id="cb90-12"><a href="#cb90-12" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;lat&quot;</span>,<span class="st">&quot;axis&quot;</span>,<span class="st">&quot;Y&quot;</span>)</span>
<span id="cb90-13"><a href="#cb90-13" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;time&quot;</span>,<span class="st">&quot;axis&quot;</span>,<span class="st">&quot;T&quot;</span>)</span>
<span id="cb90-14"><a href="#cb90-14" tabindex="-1"></a></span>
<span id="cb90-15"><a href="#cb90-15" tabindex="-1"></a><span class="co"># add global attributes</span></span>
<span id="cb90-16"><a href="#cb90-16" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="dv">0</span>,<span class="st">&quot;title&quot;</span>,title<span class="sc">$</span>value)</span>
<span id="cb90-17"><a href="#cb90-17" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="dv">0</span>,<span class="st">&quot;institution&quot;</span>,institution<span class="sc">$</span>value)</span>
<span id="cb90-18"><a href="#cb90-18" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="dv">0</span>,<span class="st">&quot;source&quot;</span>,datasource<span class="sc">$</span>value)</span>
<span id="cb90-19"><a href="#cb90-19" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="dv">0</span>,<span class="st">&quot;references&quot;</span>,references<span class="sc">$</span>value)</span>
<span id="cb90-20"><a href="#cb90-20" tabindex="-1"></a>history <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;P.J. Bartlein&quot;</span>, <span class="fu">date</span>(), <span class="at">sep=</span><span class="st">&quot;, &quot;</span>)</span>
<span id="cb90-21"><a href="#cb90-21" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="dv">0</span>,<span class="st">&quot;history&quot;</span>,history)</span>
<span id="cb90-22"><a href="#cb90-22" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="dv">0</span>,<span class="st">&quot;Conventions&quot;</span>,Conventions<span class="sc">$</span>value)</span>
<span id="cb90-23"><a href="#cb90-23" tabindex="-1"></a></span>
<span id="cb90-24"><a href="#cb90-24" tabindex="-1"></a><span class="co"># Get a summary of the created file:</span></span>
<span id="cb90-25"><a href="#cb90-25" tabindex="-1"></a>ncout</span></code></pre></div>
<pre><code>## File /Users/bartlein/Projects/ESSD/data/nc_files/cru10min30_ncdf4.nc (NC_FORMAT_NETCDF4):
## 
##      4 variables (excluding dimension variables):
##         float tmp[lon,lat,time]   (Contiguous storage)  
##             units: deg_C
##             _FillValue: 1.00000003318135e+32
##             long_name: 2m air temperature
##         float mtco[lon,lat]   (Contiguous storage)  
##             units: deg_C
##             _FillValue: 1.00000003318135e+32
##             long_name: mean_temperature_coldest_month
##         float mtwa[lon,lat]   (Contiguous storage)  
##             units: deg_C
##             _FillValue: 1.00000003318135e+32
##             long_name: mean_temperture_warmest_month
##         float mat[lon,lat]   (Contiguous storage)  
##             units: deg_C
##             _FillValue: 1.00000003318135e+32
##             long_name: mean_annual_temperature
## 
##      3 dimensions:
##         lon  Size:720 
##             units: degrees_east
##             long_name: lon
##             axis: X
##         lat  Size:360 
##             units: degrees_north
##             long_name: lat
##             axis: Y
##         time  Size:12 
##             units: days since 1900-01-01 00:00:00.0 -0:00
##             long_name: time
##             axis: T
## 
##     6 global attributes:
##         title: CRU CL 2.0 -- 10min grid sampled every 0.5 degree
##         institution: http://www.cru.uea.ac.uk/
##         source: http://www.cru.uea.ac.uk/~markn/cru05/cru05_intro.html
##         references: New et al. (2002) Climate Res 21:1-25
##         history: P.J. Bartlein, Wed Nov  8 23:55:52 2023
##         Conventions: CF-1.0</code></pre>
<p>Finally, close the file, which writes the data to disk.</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" tabindex="-1"></a><span class="co"># close the file, writing data to disk</span></span>
<span id="cb92-2"><a href="#cb92-2" tabindex="-1"></a><span class="fu">nc_close</span>(ncout)</span></code></pre></div>
<p><a href="netCDF.html">[Back to top]</a></p>
</div>
<div id="reading-and-writing-a-projected-netcdf-file"
class="section level1" number="6">
<h1><span class="header-section-number">6</span> Reading and writing a
projected netCDF file</h1>
<p>NetCDF files aren’t restricted to unprojected or
longitude-by-latitude files, but may also contain projected grids, or
those indexed by x- and y-coordinates (as opposed to longitude and
latitude). An example data set is provided by output from the North
American Regional Reanalysis (NARR), which uses a Lambert Conformal
Conic projection. The example here uses file of long-term mean
soil-moisture data.</p>
<p>The data are available on <code>ClimateLab.uoregon.edu</code> (see
File transfer on the Tasks tab), in the <code>/nc_files</code> folder,
with the file name <code>NARR_soilm.mon.ltm.nc</code>. Download the
netCDF file to a convenient folder.</p>
<div id="open-the-netcdf-file-1" class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> Open the netCDF
file</h2>
<p>Set up the path to a local copy of the NARR (North American Regional
Reanalysis) soil-moisture file</p>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" tabindex="-1"></a><span class="co"># set path and filename</span></span>
<span id="cb93-2"><a href="#cb93-2" tabindex="-1"></a>ncpath <span class="ot">&lt;-</span> <span class="st">&quot;/Users/bartlein/Projects/ESSD/data/nc_files/&quot;</span></span>
<span id="cb93-3"><a href="#cb93-3" tabindex="-1"></a>ncname <span class="ot">&lt;-</span> <span class="st">&quot;NARR_soilm.mon.ltm.nc&quot;</span>  </span>
<span id="cb93-4"><a href="#cb93-4" tabindex="-1"></a>ncfname <span class="ot">&lt;-</span> <span class="fu">paste</span>(ncpath, ncname, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb93-5"><a href="#cb93-5" tabindex="-1"></a>dname <span class="ot">&lt;-</span> <span class="st">&quot;soilm&quot;</span>  <span class="co"># soil moisture 1 kg m-2 = 1 mm</span></span></code></pre></div>
<p>Open the NetCDF data set, and print some basic information. The
<code>print()</code> function applied to the <code>ncin</code> object
produces information similar to that produced by the command-line
utility <code>ncdump</code>.</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" tabindex="-1"></a><span class="co"># open a netCDF file</span></span>
<span id="cb94-2"><a href="#cb94-2" tabindex="-1"></a>ncin <span class="ot">&lt;-</span> <span class="fu">nc_open</span>(ncfname)</span>
<span id="cb94-3"><a href="#cb94-3" tabindex="-1"></a><span class="fu">print</span>(ncin)</span></code></pre></div>
<pre><code>## File /Users/bartlein/Projects/ESSD/data/nc_files/NARR_soilm.mon.ltm.nc (NC_FORMAT_NETCDF4_CLASSIC):
## 
##      5 variables (excluding dimension variables):
##         double climatological_bnds[nbnds,time]   (Chunking: [2,1])  
##             long_name: Time Boundaries
##         float lat[x,y]   (Contiguous storage)  
##             long_name: latitude coordinate
##             units: degrees_north
##             axis: Y
##             coordinate_defines: point
##             standard_name: latitude
##         float lon[x,y]   (Contiguous storage)  
##             units: degrees_east
##             long_name: longitude coordinate
##             axis: X
##             coordinate_defines: point
##             standard_name: longitude
##         int Lambert_Conformal[]   (Contiguous storage)  
##             grid_mapping_name: lambert_conformal_conic
##             standard_parallel: 50
##              standard_parallel: 50
##             longitude_of_central_meridian: -107
##             latitude_of_projection_origin: 50
##             false_easting: 5632642.22547
##             false_northing: 4612545.65137
##         float soilm[x,y,time]   (Chunking: [349,277,1])  (Compression: shuffle,level 2)
##             units: kg/m^2
##             long_name: Long Term Monthly Mean Soil Moisture Content                                                                                        
##             precision: 1
##             GRIB_name: SOILM
##             GRIB_id: 86
##             var_desc: Soil Moisture Content
##             standard_name: soil_moisture_content
##             level_desc: 0-200cm down
##             dataset: NARR Monthly Averages
##             statistic: Long Term Mean
##             parent_stat: Mean
##             grid_mapping: Lambert_Conformal
##             coordinates: lat lon
##             cell_methods: time: mean within months time: mean over months time: mean over years                                                               
##             missing_value: -9.96920996838687e+36
##             actual_range: 59.900146484375
##              actual_range: 927.800048828125
##             valid_range: 0
##              valid_range: 1200
## 
##      4 dimensions:
##         y  Size:277 
##             long_name: northward distance from southwest corner of domain in projection coordinates
##             units: m
##             standard_name: projection_y_coordinate
##         x  Size:349 
##             long_name: eastward distance from southwest corner of domain in projection coordinates
##             units: m
##             standard_name: projection_x_coordinate
##         time  Size:12   *** is unlimited *** 
##             units: hours since 1800-1-1 00:00:0.0
##             long_name: Time
##             axis: T
##             standard_name: time
##             coordinate_defines: start
##             delta_t: 0000-01-00 00:00:00
##             avg_period: 0025-00-00 00:00:00
##             ltm_range: 1586616
##              ltm_range: 1849560
##             prev_avg_period: 0000-00-01 00:00:00
##             actual_range: -15769752
##              actual_range: -15761736
##         nbnds  Size:2 (no dimvar)
## 
##     14 global attributes:
##         standardpar1: 50
##         standardpar2: 50.000001
##         centerlon: -107
##         centerlat: 50
##         latcorners: 1.00000095367432
##          latcorners: 0.897944986820221
##          latcorners: 46.3544006347656
##          latcorners: 46.6343307495117
##         loncorners: -145.5
##          loncorners: -68.3200531005859
##          loncorners: -2.56989097595215
##          loncorners: 148.641799926758
##         stream: s1
##         title: Monthly NARR
##         Conventions: CF-1.0
##         history: created 2006/07 by data management at NOAA/ESRL/PSD
##         institution: National Centers for Environmental Prediction
##         platform: Model
##         References: http://wwwt.emc.ncep.noaa.gov/mmb/rreanl/index.html
## http://www.esrl.noaa.gov/psd/data/gridded/data.narr.html
##         dataset_title: NCEP North American Regional Reanalysis (NARR)</code></pre>
<p>Note the presence of the variable <code>Lambert_Conformal</code>. The
attributes of this variable describe the particular projection that this
file uses. Also note that longitude and latitude are 2-D variables in
this data set, and are simple variables as opposed to dimension
variables. In this data set, <code>x</code> and <code>y</code> are the
dimension variables, and together define a rectangular grid.</p>
</div>
<div id="get-coordinate-including-time-variables-1"
class="section level2" number="6.2">
<h2><span class="header-section-number">6.2</span> Get coordinate
(including time) variables</h2>
<p>Next, get the coordinate variables <code>x</code> and <code>y</code>
are read using the <code>ncvar_get()</code> function, and the first few
values of each are listed using the <code>head()</code> and
<code>tail()</code> functions. The number of <code>x</code>’s (columns)
and <code>y</code>’s (rows) values can be verified using the
<code>dim()</code> function:</p>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" tabindex="-1"></a><span class="co"># get x&#39;s and y&#39;s</span></span>
<span id="cb96-2"><a href="#cb96-2" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">ncvar_get</span>(ncin,<span class="st">&quot;x&quot;</span>)</span>
<span id="cb96-3"><a href="#cb96-3" tabindex="-1"></a>xlname <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,<span class="st">&quot;x&quot;</span>,<span class="st">&quot;long_name&quot;</span>)</span>
<span id="cb96-4"><a href="#cb96-4" tabindex="-1"></a>xunits <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,<span class="st">&quot;x&quot;</span>,<span class="st">&quot;units&quot;</span>)</span>
<span id="cb96-5"><a href="#cb96-5" tabindex="-1"></a>nx <span class="ot">&lt;-</span> <span class="fu">dim</span>(x)</span>
<span id="cb96-6"><a href="#cb96-6" tabindex="-1"></a><span class="fu">head</span>(x)</span></code></pre></div>
<pre><code>## [1]      0  32463  64926  97389 129852 162315</code></pre>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="fu">ncvar_get</span>(ncin,<span class="st">&quot;y&quot;</span>)</span>
<span id="cb98-2"><a href="#cb98-2" tabindex="-1"></a>ylname <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,<span class="st">&quot;y&quot;</span>,<span class="st">&quot;long_name&quot;</span>)</span>
<span id="cb98-3"><a href="#cb98-3" tabindex="-1"></a>yunits <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,<span class="st">&quot;y&quot;</span>,<span class="st">&quot;units&quot;</span>)</span>
<span id="cb98-4"><a href="#cb98-4" tabindex="-1"></a>ny <span class="ot">&lt;-</span> <span class="fu">dim</span>(y)</span>
<span id="cb98-5"><a href="#cb98-5" tabindex="-1"></a><span class="fu">head</span>(y)</span></code></pre></div>
<pre><code>## [1]      0  32463  64926  97389 129852 162315</code></pre>
<div class="sourceCode" id="cb100"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">c</span>(nx,ny))</span></code></pre></div>
<pre><code>## [1] 349 277</code></pre>
<p>Get the time variable and its attributes using the
<code>ncvar_get()</code> and <code>ncatt_get()</code> functions, and
list them, and also get the number of time steps using the
<code>dim()</code> function.</p>
<div class="sourceCode" id="cb102"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" tabindex="-1"></a><span class="co"># get time</span></span>
<span id="cb102-2"><a href="#cb102-2" tabindex="-1"></a>time <span class="ot">&lt;-</span> <span class="fu">ncvar_get</span>(ncin,<span class="st">&quot;time&quot;</span>)</span>
<span id="cb102-3"><a href="#cb102-3" tabindex="-1"></a>time</span></code></pre></div>
<pre><code>##  [1] -15769752 -15769008 -15768336 -15767592 -15766872 -15766128 -15765408 -15764664 -15763920 -15763200
## [11] -15762456 -15761736</code></pre>
<div class="sourceCode" id="cb104"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" tabindex="-1"></a>tunits <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,<span class="st">&quot;time&quot;</span>,<span class="st">&quot;units&quot;</span>)</span>
<span id="cb104-2"><a href="#cb104-2" tabindex="-1"></a>nt <span class="ot">&lt;-</span> <span class="fu">dim</span>(time)</span>
<span id="cb104-3"><a href="#cb104-3" tabindex="-1"></a>nt</span></code></pre></div>
<pre><code>## [1] 12</code></pre>
<p>Note that these are particularly bizzare time values, and were
constructed just to represent 12 monthly mean values.</p>
</div>
<div id="get-a-variable-1" class="section level2" number="6.3">
<h2><span class="header-section-number">6.3</span> Get a variable</h2>
<p>Get the the variable (<code>soilm</code>) and its attributes, and
verify the size of the array.</p>
<div class="sourceCode" id="cb106"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" tabindex="-1"></a><span class="co"># get soil moisture</span></span>
<span id="cb106-2"><a href="#cb106-2" tabindex="-1"></a>soilm_array <span class="ot">&lt;-</span> <span class="fu">ncvar_get</span>(ncin,dname)</span>
<span id="cb106-3"><a href="#cb106-3" tabindex="-1"></a>dlname <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,dname,<span class="st">&quot;long_name&quot;</span>)</span>
<span id="cb106-4"><a href="#cb106-4" tabindex="-1"></a>dunits <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,dname,<span class="st">&quot;units&quot;</span>)</span>
<span id="cb106-5"><a href="#cb106-5" tabindex="-1"></a>fillvalue <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin,dname,<span class="st">&quot;_FillValue&quot;</span>)</span>
<span id="cb106-6"><a href="#cb106-6" tabindex="-1"></a><span class="fu">dim</span>(soilm_array)</span></code></pre></div>
<pre><code>## [1] 349 277  12</code></pre>
<p>Note that in a projected file, each grid point has a unique latitude
and longitude value. We can read those in just like any other 2-d
variable. Keep in mind, however, that these values aren’t used for
plotting the data.</p>
<div class="sourceCode" id="cb108"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" tabindex="-1"></a>lon <span class="ot">&lt;-</span> <span class="fu">ncvar_get</span>(ncin, <span class="st">&quot;lon&quot;</span>)</span>
<span id="cb108-2"><a href="#cb108-2" tabindex="-1"></a><span class="fu">dim</span>(lon)</span></code></pre></div>
<pre><code>## [1] 349 277</code></pre>
<div class="sourceCode" id="cb110"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" tabindex="-1"></a>lat <span class="ot">&lt;-</span> <span class="fu">ncvar_get</span>(ncin, <span class="st">&quot;lat&quot;</span>)</span>
<span id="cb110-2"><a href="#cb110-2" tabindex="-1"></a><span class="fu">dim</span>(lat)</span></code></pre></div>
<pre><code>## [1] 349 277</code></pre>
<p>The last thing to get is the coordinate reference system information,
which is provided as the attributes of the variable
<code>Lambert Conformal</code>.</p>
<div class="sourceCode" id="cb112"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" tabindex="-1"></a>grid_mapping_name <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin, <span class="st">&quot;Lambert_Conformal&quot;</span>, <span class="st">&quot;grid_mappping_name&quot;</span>)</span>
<span id="cb112-2"><a href="#cb112-2" tabindex="-1"></a>standard_parallel <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin, <span class="st">&quot;Lambert_Conformal&quot;</span>, <span class="st">&quot;standard_parallel&quot;</span>)</span>
<span id="cb112-3"><a href="#cb112-3" tabindex="-1"></a>longitude_of_central_meridian <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin, <span class="st">&quot;Lambert_Conformal&quot;</span>, <span class="st">&quot;longitude_of_central_meridian&quot;</span>)</span>
<span id="cb112-4"><a href="#cb112-4" tabindex="-1"></a>latitude_of_projection_origin <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin, <span class="st">&quot;Lambert_Conformal&quot;</span>, <span class="st">&quot;latitude_of_projection_origin&quot;</span>)</span>
<span id="cb112-5"><a href="#cb112-5" tabindex="-1"></a>false_easting <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin, <span class="st">&quot;Lambert_Conformal&quot;</span>, <span class="st">&quot;false_easting&quot;</span>)</span>
<span id="cb112-6"><a href="#cb112-6" tabindex="-1"></a>false_northing <span class="ot">&lt;-</span> <span class="fu">ncatt_get</span>(ncin, <span class="st">&quot;Lambert_Conformal&quot;</span>, <span class="st">&quot;false_northing&quot;</span>)</span></code></pre></div>
<p>Close the netCDF file using the <code>nc_close()</code> function.</p>
<p><a href="netCDF.html">[Back to top]</a></p>
</div>
</div>
<div id="map-the-data" class="section level1" number="7">
<h1><span class="header-section-number">7</span> Map the data</h1>
<div id="get-a-single-slice-of-data-1" class="section level2"
number="7.1">
<h2><span class="header-section-number">7.1</span> Get a single slice of
data</h2>
<div class="sourceCode" id="cb113"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" tabindex="-1"></a><span class="co"># get a single slice or layer (e.g the January long-term mean)</span></span>
<span id="cb113-2"><a href="#cb113-2" tabindex="-1"></a>m <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb113-3"><a href="#cb113-3" tabindex="-1"></a>soilm_slice <span class="ot">&lt;-</span> soilm_array[,,m]</span></code></pre></div>
<p>The dimensions of <code>soilm_slice</code>, e.g. 349, 277, can be
verified using the <code>dim()</code> function.</p>
</div>
<div id="maps" class="section level2" number="7.2">
<h2><span class="header-section-number">7.2</span> Maps</h2>
<p>A quick look (map) of the extracted slice of data can be gotten using
the <code>image()</code> function. Here is an <code>image()</code>
function map.</p>
<div class="sourceCode" id="cb114"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" tabindex="-1"></a><span class="co"># quick map</span></span>
<span id="cb114-2"><a href="#cb114-2" tabindex="-1"></a><span class="fu">image</span>(x,y,soilm_slice, <span class="at">col=</span><span class="fu">c</span>(<span class="fu">rgb</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>),<span class="fu">brewer.pal</span>(<span class="dv">9</span>,<span class="st">&quot;Blues&quot;</span>)) )</span></code></pre></div>
<p><img src="netCDF_files/figure-html/image2-1.png" width="75%" height="75%" /></p>
<p>A better map can be obtained using the <code>levelplot()</code>
function from the <code>lattice</code> package. The
<code>expand.grid()</code> function is used to create a set of 349, 277
by 349, 277 pairs of latitude and longitude values (with latitudes
varying most rapidly), one for each element in the
<code>soilm_slice</code> array. Specific values of the cutpoints of soil
moisture categories are defined to cover the range of soil moisture
values here.</p>
<div class="sourceCode" id="cb115"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" tabindex="-1"></a><span class="co"># levelplot of the slice</span></span>
<span id="cb115-2"><a href="#cb115-2" tabindex="-1"></a>grid <span class="ot">&lt;-</span> <span class="fu">expand.grid</span>(<span class="at">x=</span>x, <span class="at">y=</span>y)</span>
<span id="cb115-3"><a href="#cb115-3" tabindex="-1"></a>cutpts <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">100</span>, <span class="dv">200</span>, <span class="dv">300</span>, <span class="dv">400</span>, <span class="dv">500</span>, <span class="dv">600</span>, <span class="dv">700</span>, <span class="dv">800</span>, <span class="dv">900</span>, <span class="dv">1000</span>)</span>
<span id="cb115-4"><a href="#cb115-4" tabindex="-1"></a><span class="fu">levelplot</span>(soilm_slice <span class="sc">~</span> x <span class="sc">*</span> y, <span class="at">data=</span>grid, <span class="at">at=</span>cutpts, <span class="at">cuts=</span><span class="dv">11</span>, <span class="at">pretty=</span>T, </span>
<span id="cb115-5"><a href="#cb115-5" tabindex="-1"></a>  <span class="at">col.regions=</span><span class="fu">c</span>(<span class="fu">rgb</span>(<span class="dv">1</span>,<span class="dv">1</span>,<span class="dv">1</span>),<span class="fu">brewer.pal</span>(<span class="dv">9</span>,<span class="st">&quot;Blues&quot;</span>)) )</span></code></pre></div>
<p><img src="netCDF_files/figure-html/levelplot2-1.png" width="75%" height="75%" /></p>
<p><a href="netCDF.html">[Back to top]</a></p>
</div>
</div>
<div id="create-and-write-a-projected-netcdf-file"
class="section level1" number="8">
<h1><span class="header-section-number">8</span> Create and write a
projected netCDF file</h1>
<p>In this example, the array read in above is written out using the
<code>ncdf4</code> package. Creating and writing (new) netCDF files
involves first defining or “laying out” the dimensions and coordiate
variables and the individual variables, and the attrributes of each, and
then creating the file and “putting” the data into the file, along with
additional attributes or metadata.</p>
<p>First, create the netCDF filename:</p>
<div class="sourceCode" id="cb116"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" tabindex="-1"></a><span class="co"># path and file name, set dname</span></span>
<span id="cb116-2"><a href="#cb116-2" tabindex="-1"></a>ncpath <span class="ot">&lt;-</span> <span class="st">&quot;/Users/bartlein/Projects/ESSD/data/nc_files/&quot;</span></span>
<span id="cb116-3"><a href="#cb116-3" tabindex="-1"></a>ncname <span class="ot">&lt;-</span> <span class="st">&quot;soilm.mon.ltm_test&quot;</span>  </span>
<span id="cb116-4"><a href="#cb116-4" tabindex="-1"></a>ncfname <span class="ot">&lt;-</span> <span class="fu">paste</span>(ncpath, ncname, <span class="st">&quot;.nc&quot;</span>, <span class="at">sep=</span><span class="st">&quot;&quot;</span>)</span>
<span id="cb116-5"><a href="#cb116-5" tabindex="-1"></a>dname <span class="ot">&lt;-</span> <span class="st">&quot;soilm&quot;</span>  </span></code></pre></div>
<p>Then create the file:</p>
<div class="sourceCode" id="cb117"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" tabindex="-1"></a><span class="co"># create and write the netCDF file -- ncdf4 version</span></span>
<span id="cb117-2"><a href="#cb117-2" tabindex="-1"></a><span class="co"># define dimensions</span></span>
<span id="cb117-3"><a href="#cb117-3" tabindex="-1"></a>xdim <span class="ot">&lt;-</span> <span class="fu">ncdim_def</span>(<span class="st">&quot;x&quot;</span>,<span class="at">units=</span><span class="st">&quot;m&quot;</span>,</span>
<span id="cb117-4"><a href="#cb117-4" tabindex="-1"></a>  <span class="at">longname=</span><span class="st">&quot;eastward distance from southwest corner of domain in projection coordinates&quot;</span>,<span class="fu">as.double</span>(x))</span>
<span id="cb117-5"><a href="#cb117-5" tabindex="-1"></a>ydim <span class="ot">&lt;-</span> <span class="fu">ncdim_def</span>(<span class="st">&quot;y&quot;</span>,<span class="at">units=</span><span class="st">&quot;m&quot;</span>,</span>
<span id="cb117-6"><a href="#cb117-6" tabindex="-1"></a>  <span class="at">longname=</span><span class="st">&quot;northward distance from southwest corner of domain in projection coordinates&quot;</span>,<span class="fu">as.double</span>(y))</span>
<span id="cb117-7"><a href="#cb117-7" tabindex="-1"></a>timedim <span class="ot">&lt;-</span> <span class="fu">ncdim_def</span>(<span class="st">&quot;time&quot;</span>,tunits<span class="sc">$</span>value,<span class="fu">as.double</span>(time))</span>
<span id="cb117-8"><a href="#cb117-8" tabindex="-1"></a></span>
<span id="cb117-9"><a href="#cb117-9" tabindex="-1"></a><span class="co"># define variables also include longitude and latitude and the CRS variable</span></span>
<span id="cb117-10"><a href="#cb117-10" tabindex="-1"></a>fillvalue <span class="ot">&lt;-</span> <span class="fl">1e32</span></span>
<span id="cb117-11"><a href="#cb117-11" tabindex="-1"></a>dlname <span class="ot">&lt;-</span> <span class="st">&quot;soil moisture&quot;</span></span>
<span id="cb117-12"><a href="#cb117-12" tabindex="-1"></a>soilm_def <span class="ot">&lt;-</span> <span class="fu">ncvar_def</span>(<span class="st">&quot;soilm&quot;</span>,dunits<span class="sc">$</span>value,<span class="fu">list</span>(xdim,ydim,timedim),fillvalue,dlname,<span class="at">prec=</span><span class="st">&quot;single&quot;</span>)</span>
<span id="cb117-13"><a href="#cb117-13" tabindex="-1"></a>dlname <span class="ot">&lt;-</span> <span class="st">&quot;Longitude of cell center&quot;</span></span>
<span id="cb117-14"><a href="#cb117-14" tabindex="-1"></a>lon_def <span class="ot">&lt;-</span> <span class="fu">ncvar_def</span>(<span class="st">&quot;lon&quot;</span>,<span class="st">&quot;degrees_east&quot;</span>,<span class="fu">list</span>(xdim,ydim),<span class="cn">NULL</span>,dlname,<span class="at">prec=</span><span class="st">&quot;double&quot;</span>)</span>
<span id="cb117-15"><a href="#cb117-15" tabindex="-1"></a>dlname <span class="ot">&lt;-</span> <span class="st">&quot;Latitude of cell center&quot;</span></span>
<span id="cb117-16"><a href="#cb117-16" tabindex="-1"></a>lat_def <span class="ot">&lt;-</span> <span class="fu">ncvar_def</span>(<span class="st">&quot;lat&quot;</span>,<span class="st">&quot;degrees_north&quot;</span>,<span class="fu">list</span>(xdim,ydim),<span class="cn">NULL</span>,dlname,<span class="at">prec=</span><span class="st">&quot;double&quot;</span>)</span>
<span id="cb117-17"><a href="#cb117-17" tabindex="-1"></a>dlname <span class="ot">&lt;-</span> <span class="st">&quot;Lambert_Conformal&quot;</span></span>
<span id="cb117-18"><a href="#cb117-18" tabindex="-1"></a>proj_def <span class="ot">&lt;-</span> <span class="fu">ncvar_def</span>(<span class="st">&quot;Lambert_Conformal&quot;</span>,<span class="st">&quot;1&quot;</span>,<span class="cn">NULL</span>,<span class="cn">NULL</span>,<span class="at">longname=</span>dlname,<span class="at">prec=</span><span class="st">&quot;char&quot;</span>)</span></code></pre></div>
<p>Next, create the file, and put the variables into it, along with
additional variable and “global” attributes (those that apply to the
whole file). Note that the attributes are of key importance to the
self-documenting properties of netCDF files.</p>
<div class="sourceCode" id="cb118"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" tabindex="-1"></a><span class="co"># create netCDF file and put arrays</span></span>
<span id="cb118-2"><a href="#cb118-2" tabindex="-1"></a>ncout <span class="ot">&lt;-</span> <span class="fu">nc_create</span>(ncfname,<span class="fu">list</span>(soilm_def,lon_def,lat_def,proj_def),<span class="at">force_v4=</span><span class="cn">TRUE</span>)</span></code></pre></div>
<div class="sourceCode" id="cb119"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" tabindex="-1"></a><span class="co"># put variables</span></span>
<span id="cb119-2"><a href="#cb119-2" tabindex="-1"></a><span class="fu">ncvar_put</span>(ncout,soilm_def,soilm_array)</span>
<span id="cb119-3"><a href="#cb119-3" tabindex="-1"></a><span class="fu">ncvar_put</span>(ncout,lon_def,lon)</span>
<span id="cb119-4"><a href="#cb119-4" tabindex="-1"></a><span class="fu">ncvar_put</span>(ncout,lat_def,lat)</span>
<span id="cb119-5"><a href="#cb119-5" tabindex="-1"></a></span>
<span id="cb119-6"><a href="#cb119-6" tabindex="-1"></a><span class="co"># put additional attributes into dimension and data variables</span></span>
<span id="cb119-7"><a href="#cb119-7" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;x&quot;</span>,<span class="st">&quot;axis&quot;</span>,<span class="st">&quot;X&quot;</span>)</span>
<span id="cb119-8"><a href="#cb119-8" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;x&quot;</span>,<span class="st">&quot;standard_name&quot;</span>,<span class="st">&quot;projection_x_coordinate&quot;</span>)</span>
<span id="cb119-9"><a href="#cb119-9" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;x&quot;</span>,<span class="st">&quot;_CoordinateAxisType&quot;</span>,<span class="st">&quot;GeoX&quot;</span>)</span>
<span id="cb119-10"><a href="#cb119-10" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;y&quot;</span>,<span class="st">&quot;axis&quot;</span>,<span class="st">&quot;Y&quot;</span>)</span>
<span id="cb119-11"><a href="#cb119-11" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;y&quot;</span>,<span class="st">&quot;standard_name&quot;</span>,<span class="st">&quot;projection_y_coordinate&quot;</span>)</span>
<span id="cb119-12"><a href="#cb119-12" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;y&quot;</span>,<span class="st">&quot;_CoordinateAxisType&quot;</span>,<span class="st">&quot;GeoY&quot;</span>)</span>
<span id="cb119-13"><a href="#cb119-13" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;soilm&quot;</span>,<span class="st">&quot;grid_mapping&quot;</span>, <span class="st">&quot;Lambert_Conformal&quot;</span>)</span>
<span id="cb119-14"><a href="#cb119-14" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;soilm&quot;</span>,<span class="st">&quot;coordinates&quot;</span>, <span class="st">&quot;lat lon&quot;</span>)</span>
<span id="cb119-15"><a href="#cb119-15" tabindex="-1"></a></span>
<span id="cb119-16"><a href="#cb119-16" tabindex="-1"></a><span class="co"># put the CRS attributes</span></span>
<span id="cb119-17"><a href="#cb119-17" tabindex="-1"></a>projname <span class="ot">&lt;-</span> <span class="st">&quot;lambert_conformal_conic&quot;</span></span>
<span id="cb119-18"><a href="#cb119-18" tabindex="-1"></a>false_easting <span class="ot">&lt;-</span> <span class="fl">5632642.22547</span></span>
<span id="cb119-19"><a href="#cb119-19" tabindex="-1"></a>false_northing <span class="ot">&lt;-</span> <span class="fl">4612545.65137</span></span>
<span id="cb119-20"><a href="#cb119-20" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;Lambert_Conformal&quot;</span>,<span class="st">&quot;name&quot;</span>,projname)</span>
<span id="cb119-21"><a href="#cb119-21" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;Lambert_Conformal&quot;</span>,<span class="st">&quot;long_name&quot;</span>,projname)</span>
<span id="cb119-22"><a href="#cb119-22" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;Lambert_Conformal&quot;</span>,<span class="st">&quot;grid_mapping_name&quot;</span>,projname)</span>
<span id="cb119-23"><a href="#cb119-23" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;Lambert_Conformal&quot;</span>,<span class="st">&quot;longitude_of_central_meridian&quot;</span>, <span class="fu">as.double</span>(longitude_of_central_meridian<span class="sc">$</span>value))</span>
<span id="cb119-24"><a href="#cb119-24" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;Lambert_Conformal&quot;</span>,<span class="st">&quot;latitude_of_projection_origin&quot;</span>, <span class="fu">as.double</span>(latitude_of_projection_origin<span class="sc">$</span>value))</span>
<span id="cb119-25"><a href="#cb119-25" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;Lambert_Conformal&quot;</span>,<span class="st">&quot;standard_parallel&quot;</span>, <span class="fu">c</span>(<span class="fl">50.0</span>, <span class="fl">50.0</span>))</span>
<span id="cb119-26"><a href="#cb119-26" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;Lambert_Conformal&quot;</span>,<span class="st">&quot;false_easting&quot;</span>,false_easting)</span>
<span id="cb119-27"><a href="#cb119-27" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;Lambert_Conformal&quot;</span>,<span class="st">&quot;false_northing&quot;</span>,false_northing)</span>
<span id="cb119-28"><a href="#cb119-28" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;Lambert_Conformal&quot;</span>,<span class="st">&quot;_CoordinateTransformType&quot;</span>,<span class="st">&quot;Projection&quot;</span>)</span>
<span id="cb119-29"><a href="#cb119-29" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="st">&quot;Lambert_Conformal&quot;</span>,<span class="st">&quot;_CoordinateAxisTypes&quot;</span>,<span class="st">&quot;GeoX GeoY&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb120"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" tabindex="-1"></a><span class="co"># add global attributes</span></span>
<span id="cb120-2"><a href="#cb120-2" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="dv">0</span>,<span class="st">&quot;title&quot;</span>,<span class="st">&quot;test output of projected data&quot;</span>)</span>
<span id="cb120-3"><a href="#cb120-3" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="dv">0</span>,<span class="st">&quot;institution&quot;</span>,<span class="st">&quot;NOAA ESRL PSD&quot;</span>)</span>
<span id="cb120-4"><a href="#cb120-4" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="dv">0</span>,<span class="st">&quot;source&quot;</span>,<span class="st">&quot;soilm.mon.ltm.nc&quot;</span>)</span>
<span id="cb120-5"><a href="#cb120-5" tabindex="-1"></a>history <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">&quot;P.J. Bartlein&quot;</span>, <span class="fu">date</span>(), <span class="at">sep=</span><span class="st">&quot;, &quot;</span>)</span>
<span id="cb120-6"><a href="#cb120-6" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="dv">0</span>,<span class="st">&quot;history&quot;</span>,history)</span>
<span id="cb120-7"><a href="#cb120-7" tabindex="-1"></a><span class="fu">ncatt_put</span>(ncout,<span class="dv">0</span>,<span class="st">&quot;Conventions&quot;</span>,<span class="st">&quot;CF=1.6&quot;</span>)</span>
<span id="cb120-8"><a href="#cb120-8" tabindex="-1"></a></span>
<span id="cb120-9"><a href="#cb120-9" tabindex="-1"></a><span class="co"># Get a summary of the created file:</span></span>
<span id="cb120-10"><a href="#cb120-10" tabindex="-1"></a>ncout</span></code></pre></div>
<pre><code>## File /Users/bartlein/Projects/ESSD/data/nc_files/soilm.mon.ltm_test.nc (NC_FORMAT_NETCDF4):
## 
##      4 variables (excluding dimension variables):
##         float soilm[x,y,time]   (Contiguous storage)  
##             units: kg/m^2
##             _FillValue: 1.00000003318135e+32
##             long_name: soil moisture
##             grid_mapping: Lambert_Conformal
##             coordinates: lat lon
##         double lon[x,y]   (Contiguous storage)  
##             units: degrees_east
##             long_name: Longitude of cell center
##         double lat[x,y]   (Contiguous storage)  
##             units: degrees_north
##             long_name: Latitude of cell center
##         char Lambert_Conformal[]   (Contiguous storage)  
##             units: 1
##             name: lambert_conformal_conic
##             long_name: lambert_conformal_conic
##             grid_mapping_name: lambert_conformal_conic
##             longitude_of_central_meridian: -107
##             latitude_of_projection_origin: 50
##             standard_parallel: 50
##              standard_parallel: 50
##             false_easting: 5632642.22547
##             false_northing: 4612545.65137
##             _CoordinateTransformType: Projection
##             _CoordinateAxisTypes: GeoX GeoY
## 
##      3 dimensions:
##         x  Size:349 
##             units: m
##             long_name: eastward distance from southwest corner of domain in projection coordinates
##             axis: X
##             standard_name: projection_x_coordinate
##             _CoordinateAxisType: GeoX
##         y  Size:277 
##             units: m
##             long_name: northward distance from southwest corner of domain in projection coordinates
##             axis: Y
##             standard_name: projection_y_coordinate
##             _CoordinateAxisType: GeoY
##         time  Size:12 
##             units: hours since 1800-1-1 00:00:0.0
##             long_name: time
## 
##     5 global attributes:
##         title: test output of projected data
##         institution: NOAA ESRL PSD
##         source: soilm.mon.ltm.nc
##         history: P.J. Bartlein, Wed Nov  8 23:55:53 2023
##         Conventions: CF=1.6</code></pre>
<p>Finally, close the file, which writes the data to disk.</p>
<div class="sourceCode" id="cb122"><pre
class="sourceCode r"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" tabindex="-1"></a><span class="co"># close the file, writing data to disk</span></span>
<span id="cb122-2"><a href="#cb122-2" tabindex="-1"></a><span class="fu">nc_close</span>(ncout)</span></code></pre></div>
<p><a href="netCDF.html">[Back to top]</a></p>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
