<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>GitHub Project Repository and Web Page</title>

<script src="site_libs/header-attrs-2.25/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/clipboard-1.7.1/clipboard.min.js"></script>
<link href="site_libs/primer-tooltips-1.4.0/build.css" rel="stylesheet" />
<link href="site_libs/klippy-0.0.0.9500/css/klippy.min.css" rel="stylesheet" />
<script src="site_libs/klippy-0.0.0.9500/js/klippy.min.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; } /* Alert */
code span.an { color: #008000; } /* Annotation */
code span.at { } /* Attribute */
code span.bu { } /* BuiltIn */
code span.cf { color: #0000ff; } /* ControlFlow */
code span.ch { color: #008080; } /* Char */
code span.cn { } /* Constant */
code span.co { color: #008000; } /* Comment */
code span.cv { color: #008000; } /* CommentVar */
code span.do { color: #008000; } /* Documentation */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.im { } /* Import */
code span.in { color: #008000; } /* Information */
code span.kw { color: #0000ff; } /* Keyword */
code span.op { } /* Operator */
code span.ot { color: #ff4000; } /* Other */
code span.pp { color: #ff4000; } /* Preprocessor */
code span.sc { color: #008080; } /* SpecialChar */
code span.ss { color: #008080; } /* SpecialString */
code span.st { color: #008080; } /* String */
code span.va { } /* Variable */
code span.vs { color: #008080; } /* VerbatimString */
code span.wa { color: #008000; font-weight: bold; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link rel="stylesheet" href="html-md-01.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">R for Earth-System Science</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="overview.html">Overview</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Topics
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li class="dropdown-header">Lecture topics</li>
    <li>
      <a href="intro.html">Introduction: R, RStudio, and other infrastructure</a>
    </li>
    <li>
      <a href="usingR.html">Using R</a>
    </li>
    <li>
      <a href="ESSdata.html">Earth-system science data</a>
    </li>
    <li>
      <a href="netCDF.html">netCDF in R</a>
    </li>
    <li>
      <a href="raster_intro.html">raster and netCDF in R</a>
    </li>
    <li>
      <a href="hdf5_intro.html">HDF in R</a>
    </li>
    <li>
      <a href="geospatial.html">Geospatial analyses</a>
    </li>
    <li>
      <a href="RMaps.html">Maps in R</a>
    </li>
    <li>
      <a href="RMaps2.html">Maps in R (2)</a>
    </li>
    <li>
      <a href="rasterVis01.html">Visualization:  raster and rasterVis</a>
    </li>
    <li>
      <a href="vis_ggplot01.html">Visualization for explanation</a>
    </li>
    <li>
      <a href="highdim.html">Visualization of high-dimensional data</a>
    </li>
    <li>
      <a href="PCA.html">Principal components analysis</a>
    </li>
    <li>
      <a href="multivariate.html">Other multivariate analyses</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tasks
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="geog495_ex01.html">Install R and RStudio</a>
    </li>
    <li>
      <a href="packages-and-data.html">Install packages and data</a>
    </li>
    <li>
      <a href="GitHub.html">Git and GitHub</a>
    </li>
    <li>
      <a href="Ex_UsingR.html">Using R exercise</a>
    </li>
    <li>
      <a href="markdown.html">Markdown and RMarkdown</a>
    </li>
    <li>
      <a href="install_netCDF.html">Install netCDF</a>
    </li>
    <li>
      <a href="install_Panoply.html">Install Panoply</a>
    </li>
    <li>
      <a href="transfer.html">File transfer</a>
    </li>
    <li>
      <a href="project.html">GitHub project repository and web page</a>
    </li>
    <li>
      <a href="ltms-and-anomalies.html">Long-term means and anomalies</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Resources
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="readings.html">Readings</a>
    </li>
    <li>
      <a href="websites.html">Old course web pages</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="about.html">About</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">GitHub Project Repository and Web Page</h1>

</div>

<div id="TOC">
<ul>
<li><a href="#introduction" id="toc-introduction"><span
class="toc-section-number">1</span> Introduction</a></li>
<li><a href="#creating-a-repository"
id="toc-creating-a-repository"><span class="toc-section-number">2</span>
Creating a repository</a>
<ul>
<li><a href="#set-up" id="toc-set-up"><span
class="toc-section-number">2.1</span> Set up</a></li>
<li><a href="#web-site-setup" id="toc-web-site-setup"><span
class="toc-section-number">2.2</span> Web site setup</a></li>
</ul></li>
<li><a href="#some-examples-of-project-web-pages"
id="toc-some-examples-of-project-web-pages"><span
class="toc-section-number">3</span> Some examples of project web
pages</a>
<ul>
<li><a href="#a-simple-script-file-based-web-page-example-1"
id="toc-a-simple-script-file-based-web-page-example-1"><span
class="toc-section-number">3.1</span> A simple script-file based web
page (Example 1)</a></li>
<li><a href="#an-r-markdown-notebook-based-web-page-example-2"
id="toc-an-r-markdown-notebook-based-web-page-example-2"><span
class="toc-section-number">3.2</span> An R Markdown Notebook based web
page (Example 2)</a></li>
<li><a href="#single-page-r-markdown-based-web-page-example-3"
id="toc-single-page-r-markdown-based-web-page-example-3"><span
class="toc-section-number">3.3</span> Single-page R Markdown-based web
page (Example 3)</a></li>
<li><a href="#a-multi-page-r-markdown-generated-website-example-4"
id="toc-a-multi-page-r-markdown-generated-website-example-4"><span
class="toc-section-number">3.4</span> A multi-page R Markdown-generated
website (Example 4)</a></li>
</ul></li>
</ul>
</div>

<script>
  addClassKlippyTo("pre.r, pre.markdown");
  addKlippy('right', 'top', 'auto', '1', 'Copy code', 'Copied!');
</script>
<div id="introduction" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>There is a hierarchy of output formats that could be used for
completing the data analysis project, ranging from a simple R script, to
a multi-page web site. These formats include</p>
<ul>
<li>an R script file (<code>*.R</code>), where the output would be
produced by downloading the script file, and executing it locally;</li>
<li>an R Markdown “Notebook” file (<code>*.Rmd</code>) that combines
code and output in a single file, which can be saved as a self-contained
<code>.html</code> file;</li>
<li>an R Markdown file (<code>*.Rmd</code>) that creates a single-page
<code>.html</code> file (or a single-file Word (<code>*.docx</code>) or
PDF (<code>*.pdf</code>) file, with some additional installations
required) that combines text, code, and output, and also contains
internal navigation (i.e. a table of contents) which could be posted on
a web page;</li>
<li>a multiple-page <code>*.html</code> web site, that includes several
<code>Rmd</code> files, a “site YAML” file (<code>_site.yml</code>), and
possibly a style sheet (<code>*.css</code>).</li>
</ul>
<p>The single-page <code>*.Rmd</code> file that produces a stand-alone
<code>*.html</code> file that is hosted on GitHub will probably be the
ultimate choice for most people, but the multiple-page web site might
also be worth attempting.</p>
<p>Implementing these approaches will be described sequentially below.
Various files and folders will be created in each example, and some of
these will have to be deleted before moving on to the next example. See
the “CLEAN UP” steps at the beginning of the second, third and fourth
examples</p>
</div>
<div id="creating-a-repository" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Creating a
repository</h1>
<p>The key steps include:</p>
<ol style="list-style-type: decimal">
<li>registering a (free) GitHub account at <a
href="github.com">[github.com]</a> (see Section 4 of <a
href="https://happygitwithr.com/index.html">[<em>Happy Git and GitHub
for the useR (HappyGitWithR)</em>]</a>) (See also the <a
href="https://education.github.com">[GitHub Education]</a> page.)</li>
<li>installing Git (Chapter 6)</li>
<li>installing a Git client (Chapter 8)</li>
<li>configuring RStudio and GitHub (Chapters 9-14, and see Chapter 10 in
particular about cacheing credentials)</li>
<li>starting a new project (Chapter 15)</li>
</ol>
<div id="set-up" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Set up</h2>
<p>The basic idea is to create a (mostly) empty new repository on
GitHub, and then “clone” to a local machine. The steps include:</p>
<p>At GitHub:</p>
<ol start="6" style="list-style-type: decimal">
<li>create a new repository (e.g. <code>geog490</code> (or
<code>geog590</code>);</li>
<li>create and edit its <code>README.md</code> file.</li>
<li>copy or write down its URL.</li>
</ol>
<p>On the local machine:</p>
<ol start="8" style="list-style-type: decimal">
<li>create a folder,
e.g. <code>/Users/bartlein/Projects/geog490</code></li>
</ol>
<p>In RStudio:</p>
<ol start="9" style="list-style-type: decimal">
<li>create a project in an existing folder (File &gt; New Project… &gt;
Version Control &gt; Git<br />
</li>
<li>Repository URL (e.g.):
<code>https://github.com/pjbartlein/geog490</code><br />
</li>
<li>Project directory name: <code>geog490</code><br />
</li>
<li>Browse to folder: <code>/Users/Projects/</code></li>
<li>click on “Create Project”</li>
</ol>
<p>(note that <code>.gitignore</code>, <code>README.md</code> and
<code>REarthSysSci</code> will have turned up in the working
directory)</p>
<p>Then, on the local machine:</p>
<ol start="14" style="list-style-type: decimal">
<li>Edit <code>README.md</code> and save (perhaps using a Markdown
editor, or in RStudio)</li>
<li>Click on <code>Tools &gt; Version Control &gt; Commit</code><br />
</li>
<li>Click on checkboxes to “stage” files</li>
<li>Click on the <code>Push</code> button to synchronize the local
repository with the remote GitHub repository.</li>
</ol>
<p>At this point, the project folder should contain the following
folders and files:</p>
<pre><code>/.git
/.Rproj.user
.gitignore
geog490.Rproj
README.md</code></pre>
</div>
<div id="web-site-setup" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Web site setup</h2>
<p>Go back to the repository
(e.g. <code>https://github.com/pjbartlein/geog490</code>, if it’s not
still open in your web browser). Check to see that you’re still logged
in. Then:</p>
<ol style="list-style-type: decimal">
<li>Click on “Create new file”;</li>
<li>In the “Name your file…” box, type (exactly):
<code>docs/</code></li>
<li>The path will expand to (e.g.) <code>geog490/docs/</code> and in the
“Name your file…” box, enter <code>index.html</code>.</li>
<li>In the <code>&lt;&gt; Edit new file</code> panel, type something
like “GEOG490 temporary page”</li>
<li>Scroll down and click on the “Commit new file” button.</li>
</ol>
<p>If you back up to the top-level <code>geog490</code> page, you’ll see
a new folder. Then click on “Settings” and scroll down to “GitHub
Pages”.</p>
<ol style="list-style-type: decimal">
<li>Click on the drop-down box below “Source”.</li>
<li>Click on the “master branch/docs folder”</li>
</ol>
<p>Sometimes it takes a little while, but you should see a the message
“Your site is ready to be published at
<code>https://pjbartlein.github.io/geog490</code>” (but with your GitHub
user name and repository name instead), just below the top of “GitHub
Pages”. If you open a new tab or window on the browser and type or paste
in that URL, you should see a new web page with the text “GEOG490
temporary page” (or whatever you typed in above). Notice that in the
projects folder you’ll see a new folder <code>/docs</code>.</p>
<p>Note that there are two web pages associated with the project
now:</p>
<ul>
<li><a
href="https://github.com/pjbartlein/geog490/"><code>https://github.com/pjbartlein/geog490/</code></a>
(the GitHub repository page); and</li>
<li><a
href="https://pjbartlein.github.io/geog490/index.html"><code>https://pjbartlein.github.io/geog490/index.html</code></a>
(the projects web page).</li>
</ul>
</div>
</div>
<div id="some-examples-of-project-web-pages" class="section level1"
number="3">
<h1><span class="header-section-number">3</span> Some examples of
project web pages</h1>
<p>R Markdown is implemented by two packages (and their dependencies),
<code>rmarkdown</code> and <code>knitr</code>. The packages can be
installed as follows</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;rmarkdown&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;knitr&quot;</span>)</span></code></pre></div>
<p>The following examples can be reproduced using this data set:
<code>cru10min30_bio.nc</code>, and the shapefile components of
<code>ne_110m_admin_0_countries.shp</code>. These can be found on the
course <code>SFTP</code> site.</p>
<p>Transfer the files to a convenient folder(s), and modify the paths in
the scripts below. Remember to get all of the shapefile components, not
just <code>ne_110m_admin_0_countries.shp</code>.</p>
<div id="a-simple-script-file-based-web-page-example-1"
class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> A simple script-file
based web page (Example 1)</h2>
<p>To demonstrate how this approach works, a small script file that
reads a shapefile and a dataset from a <code>netCDF</code> file and
makes a map will be run, and the results combined in a Markdown file
(using a Markdown editor) and then rendered as an <code>.html</code>
file. This file will then be pushed to the GitHub repository. (Make sure
the <code>netCDF</code> and shapefile listed above are available
first.)</p>
<p>To begin,</p>
<ol style="list-style-type: decimal">
<li>Create a new R Script file in the projects folder (File &gt; New
File &gt; R Script);</li>
<li>Copy the contents of <a
href="https://pjbartlein.github.io/REarthSysSci/plot_alpha.html#plot_alpha_rscript.r-example-1">[plot_alpha_RScript.R]</a>
into it;</li>
<li>Edit the paths in the script file as appropriate;</li>
<li>Save it as <code>plot_alpha_RScript.R</code> in the projects
folder.</li>
</ol>
<p>Then,</p>
<ol start="2" style="list-style-type: decimal">
<li>Run the script file and save the map of alpha as a PNG file
(e.g. <code>alpha.png</code>);</li>
<li>Create a Markdown file named <code>index.md</code>;</li>
<li>Include some text, (e.g. the summary listing of alpha, copies from
the Console pane), and a link to the image
(e.g. <code>![](alpha.png)</code>);</li>
<li>Save the <code>index.md</code> file, and also export it as an
<code>.html</code> file (e.g. <code>index.html</code>).</li>
</ol>
<p>The <code>index.md</code> Markdown file might look something like
this:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Alpha (AE/PE)  </span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="fu">Alpha</span> (AE<span class="sc">/</span>PE) was read from <span class="st">`</span><span class="at">cru10min30_bio.nc</span><span class="st">`</span> and contains<span class="sc">:</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>class       <span class="sc">:</span> RasterLayer  </span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>dimensions  <span class="sc">:</span> <span class="dv">360</span>, <span class="dv">720</span>, <span class="dv">259200</span>  (nrow, ncol, ncell)  </span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>resolution  <span class="sc">:</span> <span class="fl">0.5</span>, <span class="fl">0.5</span>  (x, y) </span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>extent      <span class="sc">:</span> <span class="sc">-</span><span class="dv">180</span>, <span class="dv">180</span>, <span class="sc">-</span><span class="dv">90</span>, <span class="dv">90</span>  (xmin, xmax, ymin, ymax)  </span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>coord. ref. <span class="sc">:</span> <span class="sc">+</span>proj<span class="ot">=</span>longlat <span class="sc">+</span>datum<span class="ot">=</span>WGS84 <span class="sc">+</span>ellps<span class="ot">=</span>WGS84 <span class="sc">+</span>towgs84<span class="ot">=</span><span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>  </span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>data source <span class="sc">:</span> <span class="er">/</span>Users<span class="sc">/</span>bartlein<span class="sc">/</span>Projects<span class="sc">/</span>ESSD<span class="sc">/</span><span class="er">/</span>data<span class="sc">/</span>nc_files<span class="sc">/</span>cru10min30_bio.nc  </span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>names       <span class="sc">:</span> Priestley.Taylor..alpha..parameter..AE.PE.  </span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>zvar        <span class="sc">:</span> mipt  </span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a><span class="sc">!</span>[](alpha.png)</span></code></pre></div>
<p>At this point, there should be three new files in the projects
folder: <code>index.md</code>, <code>index.html</code>, and
<code>alpha.png</code>. Move <code>index.html</code> and
<code>alpha.png</code> into the <code>/docs</code> folder, overwriting
<code>index.html</code>. Then</p>
<ol style="list-style-type: decimal">
<li>On the Git pane in RStudio, select all of the files, and click on
“Commit”;</li>
<li>In the Commit message pane, type something like “edited
index.html”;</li>
<li>Click on “Commit”; and</li>
<li>Click on “Push”.</li>
</ol>
<p>If you check the project’s repository page (e.g. <a
href="https://github.com/pjbartlein/geog490"><code>https://github.com/pjbartlein/geog490</code></a>)
you should see the recent commits, and if you check the project’s
<em>web</em> page (e.g. <a
href="https://pjbartlein.github.io/geog490/"><code>https://pjbartlein.github.io/geog490/</code></a>)
you should see the web page you rendered from the Markdown file created
above.</p>
</div>
<div id="an-r-markdown-notebook-based-web-page-example-2"
class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> An R Markdown
Notebook based web page (Example 2)</h2>
<p><strong>CLEAN UP FIRST:</strong> Delete (or move somewhere else) the
<code>index.md</code> file created above, and also delete the files in
the <code>/docs</code> folder (but leave that folder in place).</p>
<p>Then,</p>
<ol style="list-style-type: decimal">
<li>Create a new R Markdown file in the projects folder (File &gt; New
File &gt; R Notebook);</li>
<li>Delete the sample code in the new page. (New <code>*.Rmd</code>
files come populated with code that creates a demonstration page);</li>
<li>Copy the contents of <a
href="https://pjbartlein.github.io/REarthSysSci/plot_alpha.html#plot_alpha_rmarkdown_notebook.rmd-example-2">[plot_alpha_RMarkdown_Notebook.Rmd]</a>
into it;</li>
<li>Edit the paths in the script file as appropriate;</li>
<li>Save it as <code>plot\_alpha\_RMarkdown\_Notebook.Rmd</code> in the
projects folder; and</li>
<li>“Knit” the file by clicking on the “Knit” tab at the top of the
script-editing window.</li>
</ol>
<p>Knitting the <code>.Rmd</code> file will create the file
<code>plot_alpha_RNotebook.html</code> that contains the text, code, and
output that you would see in RStudio if you stepped through the R
Markdown Notebook file, code chunk-by-code chunk (using the little green
arrows).</p>
<ol start="3" style="list-style-type: decimal">
<li>Move <code>plot_alpha_RNotebook.html</code> to the
<code>/docs</code> folder, and ename The file as
<code>index.html</code>, and</li>
<li>Use RStudio to commit and push the files to GitHub.</li>
</ol>
<p>Check the project’s repository page (e.g. <a
href="https://github.com/pjbartlein/geog490"><code>https://github.com/pjbartlein/geog490</code></a>)
to see the recent commits, and the project’s <em>web</em> page (e.g. <a
href="https://pjbartlein.github.io/geog490/"><code>https://pjbartlein.github.io/geog490/</code></a>)
to see the Notebook file.</p>
</div>
<div id="single-page-r-markdown-based-web-page-example-3"
class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Single-page R
Markdown-based web page (Example 3)</h2>
<p><strong>CLEAN UP FIRST:</strong> Delete or move the R Markdown
Notebook file <code>plot_alpha_RNotebook.Rmd</code>, delete the folder
<code>/plot_alpha_Notebook_files</code> (which is probably empty, and
delete <code>index.html</code> in the <code>/docs</code> folder (but
leave that folder in place).</p>
<p>A single-page R Markdown <code>.html</code> page provides a more
elegant approach for combining text, code, and output that an R Markdown
Notebook page. It can include page-navigation tools (i.e. a fixed or
“floating” table of contents) and it can also employ a <code>.css</code>
style sheet to change the basic format of the page. With additional
tools installed, target output formats also include Word documents
(<code>*.docx</code>) and PDF files (<code>.pdf</code>).</p>
<p>Create an R Markdown file that will produce a single-page
<code>.html</code> file as output:</p>
<ol style="list-style-type: decimal">
<li>Create a new R Markdown file in the projects folder (File &gt; New
File &gt; R Markdown);</li>
<li>Delete the sample code in the new page;</li>
<li>Copy the contents of <a
href="https://pjbartlein.github.io/REarthSysSci/plot_alpha.html#plot_alpha_rmarkdown_1page.rmd-example-3">[plot_alpha_RMarkdown_1Page.Rmd]</a>
into it;</li>
<li>Edit the paths in the script file as appropriate;</li>
<li>Save it as <code>plot\_alpha\_RMarkdown\_1Page.Rmd</code> in the
projects folder;</li>
</ol>
<p>Also download a <code>.css</code> file that implements the appearance
of the course web pages: <a
href="https://pjbartlein.github.io/REarthSysSci/html-md-01.css">[html-md-01.css]</a>.
Save this in the projects folder.</p>
<p>The <code>.Rmd</code> file looks much like the R Markdown Notebook
file, but has two additional blocks of code at the beginning of the
file. The first block is a “YAML header” that governs how the file is
rendered</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>  <span class="sc">---</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>  title<span class="sc">:</span>  Raster mapping</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  output<span class="sc">:</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>    html_document<span class="sc">:</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>      css<span class="sc">:</span> html<span class="sc">-</span>md<span class="fl">-01.</span>css</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>      fig_caption<span class="sc">:</span> yes</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>      highlight<span class="sc">:</span> haddock</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>      number_sections<span class="sc">:</span> yes</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>      theme<span class="sc">:</span> spacelab</span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>      toc<span class="sc">:</span> yes</span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a>      toc_float<span class="sc">:</span> true</span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>      collapsed<span class="sc">:</span> no</span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a>  <span class="sc">---</span></span></code></pre></div>
<p>Most of the entries are obvious: “<code>title</code>” is used at the
top of the page, “<code>output</code>” species that an
<code>.html</code> file is desired (as opposed to a Word document or
PDF); “<code>highlight: haddock</code>” specifies a particular style of
syntax highlighting (<code>haddock</code>);
“<code>number_sections: yes</code>” numbers the headings, which are
reflected in the table of contents; “<code>theme: spacelab</code>”
specifies a particular “Bootswatch” theme (the course pages use
“cosmos”; “<code>toc: yes</code>” turns on the table of contents, and
“<code>toc_float: true</code>” keeps the table of contents visible on
the left-hand side of the window; and “<code>collapsed: no</code>”
specifies that the table of contents should be visible. See Chapter 3 of
<a href="https://bookdown.org/yihui/rmarkdown/documents.html"><em>R
Markdown — The Definitive Guide</em></a> for more discussion of the YAML
headers.</p>
<p>The second block of code includes some “global” code-chunk options
(that apply to all “code chunks”, or blocks of code set off by backtics
(```) (Remove <code>##</code>’s before using this code fragment)</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="do">## ```{r set-options, echo=FALSE}</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="do">## options(width = 105)</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="do">## knitr::opts_chunk$set(dev=&#39;png&#39;, dpi=300, cache=TRUE)</span></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="do">## pdf.options(useDingbats = TRUE)</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="do">## ```</span></span></code></pre></div>
<p>As before,</p>
<ol style="list-style-type: decimal">
<li>Open the <code>.Rmd</code>
(<code>plot\_alpha\_RMarkdown\_1Page.Rmd</code>) file if it’s not
already open;</li>
<li>“Knit” the file by clicking on the “Knit” tab at the top of the
script-editing window.</li>
</ol>
<p>Knitting the <code>.Rmd</code> file will create the file
<code>plot_alpha_RMarkdown_1page.html</code> that contains the text,
code, and output (maps and text).</p>
<ul>
<li>Move <code>plot_alpha_RMarkdown_1page.html</code> to the
<code>/docs</code> folder and rename the file as
<code>index.html</code>;</li>
<li>Delete the folders <code>/plot_alpha_RMarkdown_1page_cache</code>
and <code>/plot_alpha_RMarkdown_1page_files</code>;</li>
<li>Use RStudio to commit and push the files to GitHub.</li>
</ul>
<p>Knitting the <code>.Rmd</code> file will create the file
<code>plot_alpha_RMarkdown_1page.html</code> that contains the text,
code, and output (maps and text).</p>
<ul>
<li>Move <code>plot_alpha_RMarkdown_1page.html</code> to the
<code>/docs</code> folder and rename the file as
<code>index.html</code>;</li>
<li>Delete the folders <code>/plot_alpha_RMarkdown_1page_cache</code>
and <code>/plot_alpha_RMarkdown_1page_files</code>;</li>
<li>Use RStudio to commit and push the files to GitHub.</li>
</ul>
</div>
<div id="a-multi-page-r-markdown-generated-website-example-4"
class="section level2" number="3.4">
<h2><span class="header-section-number">3.4</span> A multi-page R
Markdown-generated website (Example 4)</h2>
<p><strong>CLEAN UP FIRST:</strong> Delete or move the R Markdown file
<code>plot_alpha_RMarkdown_1Page.Rmd</code>, delete the folders
<code>/plot_alpha_RMarkdown_1Page_files</code> and
<code>/plot_alpha_RMarkdown_1Page_cache</code>, and delete the files in
the <code>/docs</code> folder (but leave that folder in place)</p>
<p>A multi-page R Markdown website, as the name suggests, includes a
collection of <code>*.Rmd</code> files (that are each knit to create
<code>*.html</code> files, plus a support file (<code>_site.yml</code>)
that controls the navigation bar layout. The steps below will create a
minimalist multi-page web site. The site will include an
<code>index.html</code> page with a brief overview, a “Topics” tab with
two pages, one an introduction, and the other being the “Raster mapping”
page, and an “About” tab, implemented by four <code>*.Rmd</code> files
(plus the <code>_site.yml</code> file). Although R Studio contains great
script-editing features, it’s a little clunky as a text editor compared
to Markdown Pad or MacDown. In practice, if there is a lot of text to
compose, edit, and format (as is the case with this page), it may be
more efficient to do that in a Markdown editor, and then include the
contents of the <code>*.md</code> file created by the editor in a
<code>*.Rmd</code> file using the Knitr “child document” chunk option
(e.g. <code>{r child="index.md"}</code>). That will be done here for the
<code>index.Rmd</code>, <code>intro.Rmd</code>, and
<code>about.Rmd</code> files, while a standard <code>*.Rmd</code> file
will be used for the main results page.</p>
<p>Start by creating an R Markdown file that will produce a single-page
<code>.html</code> file as output, which will be incorporated into the
web site later:</p>
<ol style="list-style-type: decimal">
<li>Create a new R Markdown file in the projects folder (File &gt; New
File &gt; R Markdown);</li>
<li>Delete the sample code in the new page;</li>
<li>Copy the contents of <a
href="https://pjbartlein.github.io/REarthSysSci/plot_alpha.html#plot_alpha_rmarkdown_site.rmd-example-4">[plot_alpha_RMarkdown_Site.Rmd]</a>
into it;</li>
<li>Edit the paths in the script file as appropriate;</li>
<li>Save it as <code>plot\_alpha\_RMarkdown\_Site.Rmd</code> in the
projects folder; and</li>
<li>Knit the file.</li>
</ol>
<p>Next, create a set of three <code>*.md</code> files as follows:</p>
<p><code>index.Rmd</code>: (Remove <code>##</code>’s before attempting
to run.)</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="do">## ---</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="do">## title: &quot;GEOG4/590: R for Earth-System Science&quot;</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="do">## output:</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="do">##   html_document:</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a><span class="do">##     css: html-md-01.css</span></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="do">##     fig_caption: no</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a><span class="do">##     number_sections: no</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a><span class="do">##     toc: no</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a><span class="do">##     toc_float: false</span></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="do">##     collapsed: no</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a><span class="do">## ---</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a><span class="do">## ```{r set-options, echo=FALSE}</span></span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a><span class="do">## options(width = 105)</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a><span class="do">## knitr::opts_chunk$set(dev=&#39;png&#39;, dpi=300, cache=TRUE)</span></span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a><span class="do">## pdf.options(useDingbats = TRUE)</span></span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a><span class="do">## ```</span></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="do">## ## Project web page for GEOG 4/595 *R for Earth-System Science* ##</span></span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a><span class="do">## Topics covered here include:</span></span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a><span class="do">## - An introduction to the project</span></span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a><span class="do">## - A simple example of plotting a raster slice</span></span>
<span id="cb6-25"><a href="#cb6-25" tabindex="-1"></a><span class="do">## - (more to come…)</span></span></code></pre></div>
<p><code>intro.Rmd</code>: (Remove <code>##</code>’s before attempting
to run.)</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="do">## ---</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="do">## title: &quot;Introduction to the project&quot;</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="do">## output:</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="do">##   html_document:</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="do">##     css: html-md-01.css</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="do">##     fig_caption: no</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="do">##     number_sections: no</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a><span class="do">##     toc: no</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a><span class="do">##     toc_float: false</span></span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a><span class="do">##     collapsed: no</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="do">## ---</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a><span class="do">## ```{r set-options, echo=FALSE}</span></span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a><span class="do">## options(width = 105)</span></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a><span class="do">## knitr::opts_chunk$set(dev=&#39;png&#39;, dpi=300, cache=TRUE)</span></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a><span class="do">## pdf.options(useDingbats = TRUE)</span></span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a><span class="do">## ```</span></span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a><span class="do">## The visualization developed here uses the `raster` and `rasterVis` packages to map the global</span></span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a><span class="do">## pattern of alpha, calculated using the CRU CL2 climate data set.  The maps also include a</span></span>
<span id="cb7-21"><a href="#cb7-21" tabindex="-1"></a><span class="do">## shapefile of world coastlines and countries.</span></span></code></pre></div>
<p><code>about.Rmd</code>: (Remove <code>##</code>’s before attempting
to run.)</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="do">## ---</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="do">## title: &quot;GEOG 4/590:  R for Earth-System Science&quot;</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="do">## output:</span></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="do">##   html_document:</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="do">##     css: html-md-01.css</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a><span class="do">##     fig_caption: no</span></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="do">##     number_sections: no</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="do">##     toc: no</span></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="do">##     toc_float: false</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="do">##     collapsed: no</span></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="do">## ---</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="do">## ```{r set-options, echo=FALSE}</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="do">## options(width = 105)</span></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a><span class="do">## knitr::opts_chunk$set(dev=&#39;png&#39;, dpi=300, cache=TRUE)</span></span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a><span class="do">## pdf.options(useDingbats = TRUE)</span></span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a><span class="do">## ```</span></span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a><span class="do">## P.J. Bartlein</span></span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a><span class="do">## Dept. Geography</span></span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a><span class="do">## Univ. Oregon</span></span>
<span id="cb8-21"><a href="#cb8-21" tabindex="-1"></a><span class="do">## bartlein@uoregon.edu</span></span>
<span id="cb8-22"><a href="#cb8-22" tabindex="-1"></a><span class="do">##</span></span>
<span id="cb8-23"><a href="#cb8-23" tabindex="-1"></a><span class="do">## The GitHub repository for this web site is at:</span></span>
<span id="cb8-24"><a href="#cb8-24" tabindex="-1"></a><span class="do">## [https://github.com/pjbartlein/geog490](https://github.com/pjbartlein/geog490)</span></span></code></pre></div>
<p>(The above files are just examples.)</p>
<p>Knit the three files.</p>
<p>Finally, create a <code>_site.yml</code> file (File &gt; New File
&gt; Text File in RStudio, or use a text editor):</p>
<p><code>_site.yml</code>:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a>name<span class="sc">:</span> GEOG <span class="dv">4</span><span class="sc">/</span><span class="dv">590</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>navbar<span class="sc">:</span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>  title<span class="sc">:</span> <span class="st">&quot;GEOG 4/590:  R for Earth-System Science&quot;</span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>  left<span class="sc">:</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>  <span class="sc">-</span> text<span class="sc">:</span> Topics</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>    menu<span class="sc">:</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>    <span class="sc">-</span> text<span class="sc">:</span> Introduction</span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>      href<span class="sc">:</span> intro.html</span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>    <span class="sc">-</span> text<span class="sc">:</span> Raster mapping example</span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a>      href<span class="sc">:</span> plot_alpha_RMarkdown_Site.html</span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>  right<span class="sc">:</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a>  <span class="sc">-</span> text<span class="sc">:</span> About</span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a>    href<span class="sc">:</span> about.html</span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a>output<span class="sc">:</span></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a>  html_document<span class="sc">:</span></span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a>    theme<span class="sc">:</span> spacelab</span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a>    highlight<span class="sc">:</span> haddock</span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a>    css<span class="sc">:</span> html<span class="sc">-</span>md<span class="fl">-01.</span>css</span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a>    fig_caption<span class="sc">:</span> no</span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a>    number_sections<span class="sc">:</span> yes</span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a>    toc<span class="sc">:</span> yes</span>
<span id="cb9-22"><a href="#cb9-22" tabindex="-1"></a>    toc_float<span class="sc">:</span> no</span>
<span id="cb9-23"><a href="#cb9-23" tabindex="-1"></a>    collapsed<span class="sc">:</span> no</span>
<span id="cb9-24"><a href="#cb9-24" tabindex="-1"></a>    lib_dir<span class="sc">:</span> site_libs</span>
<span id="cb9-25"><a href="#cb9-25" tabindex="-1"></a>    self_contained<span class="sc">:</span> no</span>
<span id="cb9-26"><a href="#cb9-26" tabindex="-1"></a>output_dir<span class="sc">:</span> docs</span></code></pre></div>
<p>The <code>knitr()</code> package is quite fussy about formatting of
the <code>_site.yml</code> file so check it first in the case of later
errors.</p>
<p>In RStudio, the “build tools” needed to create the web site will need
to be configured. On the RStudio menu, click on Build. If there is a
menu choice that says “Install Tools” do that first, otherwise click on
“Configure Build Tools…” A “Project Options” dialog will open. On the
“Project build tools:” dropdown box, select “Website” The initial
choices are probably fine, with the “Site directory” set to “(Project
Root)” and the “Preview” and “Re-knit” checkboxes checked. Upon clicking
OK, a “Build” tab should appear in the upper-right pane. If it doesn’t,
restart R Studio (Session &gt; Restart R). Select the build tab, and
click on “Build Website”.</p>
<p>At this point, the project folder should contain:</p>
<pre><code>/.git
/.Rproj.user
/docs
/plot_alpha_RMarkdown_Site_cache
/plot_alpha_RMarkdown_Site_files

html-md-01.css

.gitignore
.RData
.Rhistory

about.Rmd
index.Rmd
intro.Rmd
plot_alpha_RMarkdown_Site.Rmd

geog490.Rproj

README.md
_site.yml</code></pre>
<p>… and the docs folder,</p>
<pre><code>/site_libs
/plot_alpha_RMarkdown_Site_files

html-md-01.css

about.html
index.html
intro.html
plot_alpha_RMarkdown_Site.html
README.html</code></pre>
<p>The website should be viewable at: <a
href="https://pjbartlein.github.io/geog490/index.html"><code>https://pjbartlein.github.io/geog490/index.html</code></a>
(the project’s web page)</p>
</div>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
